
<!DOCTYPE html>

<html>
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>ALLCools.clustering</title>
    
  <link href="../../../_static/css/theme.css" rel="stylesheet">
  <link href="../../../_static/css/index.ff1ffe594081f20da1ef19478df9384b.css" rel="stylesheet">

    
  <link rel="stylesheet"
    href="../../../_static/vendor/fontawesome/5.13.0/css/all.min.css">
  <link rel="preload" as="font" type="font/woff2" crossorigin
    href="../../../_static/vendor/fontawesome/5.13.0/webfonts/fa-solid-900.woff2">
  <link rel="preload" as="font" type="font/woff2" crossorigin
    href="../../../_static/vendor/fontawesome/5.13.0/webfonts/fa-brands-400.woff2">

    
      

    
    <link rel="stylesheet" type="text/css" href="../../../_static/pygments.css" />
    <link rel="stylesheet" type="text/css" href="../../../_static/sphinx-book-theme.css?digest=c3fdc42140077d1ad13ad2f1588a4309" />
    <link rel="stylesheet" type="text/css" href="../../../_static/togglebutton.css" />
    <link rel="stylesheet" type="text/css" href="../../../_static/copybutton.css" />
    <link rel="stylesheet" type="text/css" href="../../../_static/mystnb.css" />
    <link rel="stylesheet" type="text/css" href="../../../_static/sphinx-thebe.css" />
    <link rel="stylesheet" type="text/css" href="../../../_static/graphviz.css" />
    <link rel="stylesheet" type="text/css" href="../../../_static/sphinx-codeautolink.css" />
    <link rel="stylesheet" type="text/css" href="../../../_static/panels-main.c949a650a448cc0ae9fd3441c0e17fb0.css" />
    <link rel="stylesheet" type="text/css" href="../../../_static/panels-variables.06eb56fa6e07937060861dad626602ad.css" />
    
  <link rel="preload" as="script" href="../../../_static/js/index.be7d3bbb2ef33a8344ce.js">

    <script data-url_root="../../../" id="documentation_options" src="../../../_static/documentation_options.js"></script>
    <script src="../../../_static/jquery.js"></script>
    <script src="../../../_static/underscore.js"></script>
    <script src="../../../_static/doctools.js"></script>
    <script src="../../../_static/togglebutton.js"></script>
    <script src="../../../_static/clipboard.min.js"></script>
    <script src="../../../_static/copybutton.js"></script>
    <script>var togglebuttonSelector = '.toggle, .admonition.dropdown, .tag_hide_input div.cell_input, .tag_hide-input div.cell_input, .tag_hide_output div.cell_output, .tag_hide-output div.cell_output, .tag_hide_cell.cell, .tag_hide-cell.cell';</script>
    <script src="../../../_static/sphinx-book-theme.d59cb220de22ca1c485ebbdc042f0030.js"></script>
    <script async="async" src="https://unpkg.com/thebe@0.5.1/lib/index.js"></script>
    <script>
        const thebe_selector = ".thebe"
        const thebe_selector_input = "pre"
        const thebe_selector_output = ".output"
    </script>
    <script async="async" src="../../../_static/sphinx-thebe.js"></script>
    <script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    <link rel="shortcut icon" href="../../../_static/allcools_favicon.png"/>
    <link rel="index" title="Index" href="../../../genindex.html" />
    <link rel="search" title="Search" href="../../../search.html" />
    <link rel="next" title="ALLCools.clustering.doublets" href="doublets/index.html" />
    <link rel="prev" title="ALLCools.abc.abc_score" href="../abc/abc_score/index.html" />
    <meta name="viewport" content="width=device-width, initial-scale=1" />
    <meta name="docsearch:language" content="None">
    

    <!-- Google Analytics -->
    
<script async="" src="https://www.googletagmanager.com/gtag/js?id=G-4F8CRF0GK9"></script>
<script>
                    window.dataLayer = window.dataLayer || [];
                    function gtag(){ dataLayer.push(arguments); }
                    gtag('js', new Date());
                    gtag('config', 'G-4F8CRF0GK9');
                </script>

  </head>
  <body data-spy="scroll" data-target="#bd-toc-nav" data-offset="80">
    
    <div class="container-fluid" id="banner"></div>

    

    <div class="container-xl">
      <div class="row">
          
<div class="col-12 col-md-3 bd-sidebar site-navigation show" id="site-navigation">
    
        <div class="navbar-brand-box">
    <a class="navbar-brand text-wrap" href="../../../index.html">
      
        <!-- `logo` is deprecated in Sphinx 4.0, so remove this when we stop supporting 3 -->
        
      
      
      <img src="../../../_static/allcools_logo.png" class="logo" alt="logo">
      
      
    </a>
</div><form class="bd-search d-flex align-items-center" action="../../../search.html" method="get">
  <i class="icon fas fa-search"></i>
  <input type="search" class="form-control" name="q" id="search-input" placeholder="Search this book..." aria-label="Search this book..." autocomplete="off" >
</form><nav class="bd-links" id="bd-docs-nav" aria-label="Main">
    <div class="bd-toc-item active">
        <ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../../../intro.html">
   ALLCools: ALL methyl-Cytosine tools
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  GET STARTED
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../../../start/installation.html">
   Installation
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../../start/input_files.html">
   Input Files
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../../start/analysis_steps.html">
   Analysis Steps
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../../start/interactive.html">
   Examples In This Document
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  CELLULAR ANALYSIS
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../../../cell_level/basic/intro_basic_clustering.html">
   Basic Clustering Walk-through
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-1" name="toctree-checkbox-1" type="checkbox"/>
  <label for="toctree-checkbox-1">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../../../cell_level/basic/mch_mcg_100k_basic.html">
     Basic Cell Clustering Using 100Kb Bins
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../../cell_level/basic/mcg_5kb_basic.html">
     Basic Cell Clustering Using mCG-5Kb Bins
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../../../cell_level/step_by_step/intro_step_by_step_clustering.html">
   Step-by-step Clustering Analysis
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-2" name="toctree-checkbox-2" type="checkbox"/>
  <label for="toctree-checkbox-2">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2 has-children">
    <a class="reference internal" href="../../../cell_level/step_by_step/100kb/intro_100kb.html">
     Step-by-step Clustering Using 100Kb bins
    </a>
    <input class="toctree-checkbox" id="toctree-checkbox-3" name="toctree-checkbox-3" type="checkbox"/>
    <label for="toctree-checkbox-3">
     <i class="fas fa-chevron-down">
     </i>
    </label>
    <ul>
     <li class="toctree-l3">
      <a class="reference internal" href="../../../cell_level/step_by_step/100kb/01-CellBasicFiltering.html">
       Cell Basic Filtering
      </a>
     </li>
     <li class="toctree-l3">
      <a class="reference internal" href="../../../cell_level/step_by_step/100kb/02-BasicFeatureFiltering.html">
       Feature Basic Filtering
      </a>
     </li>
     <li class="toctree-l3">
      <a class="reference internal" href="../../../cell_level/step_by_step/100kb/03-HighlyVariableFeatureSelection.html">
       Calculate Highly Variable Features And Get mC Fraction AnnData
      </a>
     </li>
     <li class="toctree-l3">
      <a class="reference internal" href="../../../cell_level/step_by_step/100kb/04a-PreclusteringAndClusterEnrichedFeatures-mCH.html">
       Preclustering and Cluster Enriched Features
      </a>
     </li>
     <li class="toctree-l3">
      <a class="reference internal" href="../../../cell_level/step_by_step/100kb/04b-PreclusteringAndClusterEnrichedFeatures-mCG.html">
       Preclustering and Cluster Enriched Features
      </a>
     </li>
     <li class="toctree-l3">
      <a class="reference internal" href="../../../cell_level/step_by_step/100kb/05-DecompositionAndEmbedding.html">
       Decomposition And Embedding
      </a>
     </li>
     <li class="toctree-l3">
      <a class="reference internal" href="../../../cell_level/step_by_step/100kb/06-Clustering.html">
       Consensus Clustering
      </a>
     </li>
    </ul>
   </li>
   <li class="toctree-l2 has-children">
    <a class="reference internal" href="../../../cell_level/step_by_step/5kb/intro_5kb.html">
     Step-by-step Clustering using mCG-5Kb Input
    </a>
    <input class="toctree-checkbox" id="toctree-checkbox-4" name="toctree-checkbox-4" type="checkbox"/>
    <label for="toctree-checkbox-4">
     <i class="fas fa-chevron-down">
     </i>
    </label>
    <ul>
     <li class="toctree-l3">
      <a class="reference internal" href="../../../cell_level/step_by_step/5kb/01-CellBasicFiltering.html">
       Cell Basic Filtering
      </a>
     </li>
     <li class="toctree-l3">
      <a class="reference internal" href="../../../cell_level/step_by_step/5kb/02-DecompositionAndEmbedding.html">
       Decomposition Using mCG-5Kb Bins
      </a>
     </li>
     <li class="toctree-l3">
      <a class="reference internal" href="../../../cell_level/step_by_step/5kb/03-Clustering.html">
       Consensus Clustering
      </a>
     </li>
    </ul>
   </li>
  </ul>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../../../cell_level/dmg/intro_dmg.html">
   Differential Methylated Genes Analysis
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-5" name="toctree-checkbox-5" type="checkbox"/>
  <label for="toctree-checkbox-5">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../../../cell_level/dmg/01-AddGenemCFractions.html">
     Calculate Gene mC Fractions
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../../cell_level/dmg/02-DifferentiallyMethylatedGenes.html">
     Cluster Differentially Methylated Genes
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../../cell_level/dmg/03-PlotClusterDMGs.html">
     Plot Cluster DMGs
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../../cell_level/dmg/04-PairwiseDMG.html">
     Differential Methylated Genes - Pairwise
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../../../cell_level/integration/intro_integration.html">
   Data Integration
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-6" name="toctree-checkbox-6" type="checkbox"/>
  <label for="toctree-checkbox-6">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../../../cell_level/integration/multi_modalities/Neurons/01-preprocess_10x.html">
     Preprocess 10X snRNA-seq
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../../cell_level/integration/multi_modalities/Neurons/02-preprocess_SMART.html">
     Preprocess SMART-seq
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../../cell_level/integration/multi_modalities/Neurons/03-preprocess_snATAC.html">
     Preprocess snATAC-seq
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../../cell_level/integration/multi_modalities/Neurons/04-integration.html">
     Perform Cross-Modality Integration
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../../../cell_level/doublets/intro_doublets.html">
   Doublets Identification
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-7" name="toctree-checkbox-7" type="checkbox"/>
  <label for="toctree-checkbox-7">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../../../cell_level/doublets/doublet_score_chrom100k.html">
     Calculate Cell Doublet Score
    </a>
   </li>
  </ul>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  GENOMIC ANALYSIS
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../../../cluster_level/intro.html">
   Post-clustering Genomic Analysis
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-8" name="toctree-checkbox-8" type="checkbox"/>
  <label for="toctree-checkbox-8">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../../../cluster_level/RegionDS/01a.call_dmr.html">
     DMR Calling
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../../cluster_level/RegionDS/01b.other_option_to_init_region_ds.html">
     Alternative Way to Create RegionDS
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../../cluster_level/RegionDS/02.annotation.html">
     Annotate RegionDS
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../../cluster_level/RegionDS/03.filter_dmr.html">
     Filter DMR
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../../../cluster_level/RegionDS/intro_motif.html">
   Motif Scan and Analysis
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-9" name="toctree-checkbox-9" type="checkbox"/>
  <label for="toctree-checkbox-9">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../../../cluster_level/RegionDS/04.motif_scan.html">
     Motif Scan
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../../cluster_level/RegionDS/05.motif_enrichment.html">
     Motif Enrichment Analysis
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../../../cluster_level/Correlation/intro_corr.html">
   Region-Region Correlation Analysis
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-10" name="toctree-checkbox-10" type="checkbox"/>
  <label for="toctree-checkbox-10">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../../../cluster_level/Correlation/01.mc_cluster_profile.html">
     Prepare Cluster-by-Gene mC Profile
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../../cluster_level/Correlation/02.dmr_gene_corr.html">
     DMR-Gene Correlation
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../../../cluster_level/REPTILE/intro_reptile.html">
   REPTILE
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-11" name="toctree-checkbox-11" type="checkbox"/>
  <label for="toctree-checkbox-11">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../../../cluster_level/REPTILE/01.prepare_input.html">
     Prepare REPTILE
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../../cluster_level/REPTILE/02.run_reptile.html">
     Predict Enhancer with REPTILE Algorithm
    </a>
   </li>
  </ul>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  COMMAND LINE TOOLS
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../../../command_line/allcools.html">
   Entry Point
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../../command_line/allcools_allc.html">
   <code class="docutils literal notranslate">
    <span class="pre">
     allcools
    </span>
    <span class="pre">
     allc
    </span>
   </code>
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../../command_line/allcools_standard.html">
   <code class="docutils literal notranslate">
    <span class="pre">
     allcools
    </span>
    <span class="pre">
     standard
    </span>
   </code>
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../../command_line/allcools_profile.html">
   <code class="docutils literal notranslate">
    <span class="pre">
     allcools
    </span>
    <span class="pre">
     profile
    </span>
   </code>
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../../command_line/allcools_tbi.html">
   <code class="docutils literal notranslate">
    <span class="pre">
     allcools
    </span>
    <span class="pre">
     tbi
    </span>
   </code>
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../../command_line/allcools_merge.html">
   <code class="docutils literal notranslate">
    <span class="pre">
     allcools
    </span>
    <span class="pre">
     merge
    </span>
   </code>
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../../command_line/allcools_extract.html">
   <code class="docutils literal notranslate">
    <span class="pre">
     allcools
    </span>
    <span class="pre">
     extract
    </span>
   </code>
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../../command_line/allcools_bw.html">
   <code class="docutils literal notranslate">
    <span class="pre">
     allcools
    </span>
    <span class="pre">
     bw
    </span>
   </code>
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../../command_line/allcools_region.html">
   <code class="docutils literal notranslate">
    <span class="pre">
     allcools
    </span>
    <span class="pre">
     region
    </span>
   </code>
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../../command_line/allcools_dataset.html">
   <code class="docutils literal notranslate">
    <span class="pre">
     allcools
    </span>
    <span class="pre">
     generate-dataset
    </span>
   </code>
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../../command_line/allcools_mcds.html">
   <code class="docutils literal notranslate">
    <span class="pre">
     allcools
    </span>
    <span class="pre">
     mcds
    </span>
   </code>
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../../command_line/allcools_mcad.html">
   <code class="docutils literal notranslate">
    <span class="pre">
     allcools
    </span>
    <span class="pre">
     mcad
    </span>
   </code>
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../../command_line/allcools_table_to_allc.html">
   <code class="docutils literal notranslate">
    <span class="pre">
     allcools
    </span>
    <span class="pre">
     tbi
    </span>
   </code>
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  API
 </span>
</p>
<ul class="current nav bd-sidenav">
 <li class="toctree-l1 current active has-children">
  <a class="reference internal" href="../index.html">
   <code class="xref py py-mod docutils literal notranslate">
    <span class="pre">
     ALLCools
    </span>
   </code>
  </a>
  <input checked="" class="toctree-checkbox" id="toctree-checkbox-12" name="toctree-checkbox-12" type="checkbox"/>
  <label for="toctree-checkbox-12">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul class="current">
   <li class="toctree-l2 has-children">
    <a class="reference internal" href="../abc/index.html">
     <code class="xref py py-mod docutils literal notranslate">
      <span class="pre">
       ALLCools.abc
      </span>
     </code>
    </a>
    <input class="toctree-checkbox" id="toctree-checkbox-13" name="toctree-checkbox-13" type="checkbox"/>
    <label for="toctree-checkbox-13">
     <i class="fas fa-chevron-down">
     </i>
    </label>
    <ul>
     <li class="toctree-l3">
      <a class="reference internal" href="../abc/abc_score/index.html">
       <code class="xref py py-mod docutils literal notranslate">
        <span class="pre">
         ALLCools.abc.abc_score
        </span>
       </code>
      </a>
     </li>
    </ul>
   </li>
   <li class="toctree-l2 current active has-children">
    <a class="current reference internal" href="#">
     <code class="xref py py-mod docutils literal notranslate">
      <span class="pre">
       ALLCools.clustering
      </span>
     </code>
    </a>
    <input checked="" class="toctree-checkbox" id="toctree-checkbox-14" name="toctree-checkbox-14" type="checkbox"/>
    <label for="toctree-checkbox-14">
     <i class="fas fa-chevron-down">
     </i>
    </label>
    <ul>
     <li class="toctree-l3 has-children">
      <a class="reference internal" href="doublets/index.html">
       <code class="xref py py-mod docutils literal notranslate">
        <span class="pre">
         ALLCools.clustering.doublets
        </span>
       </code>
      </a>
      <input class="toctree-checkbox" id="toctree-checkbox-15" name="toctree-checkbox-15" type="checkbox"/>
      <label for="toctree-checkbox-15">
       <i class="fas fa-chevron-down">
       </i>
      </label>
      <ul>
       <li class="toctree-l4">
        <a class="reference internal" href="doublets/coverage_doublets/index.html">
         <code class="xref py py-mod docutils literal notranslate">
          <span class="pre">
           ALLCools.clustering.doublets.coverage_doublets
          </span>
         </code>
        </a>
       </li>
       <li class="toctree-l4">
        <a class="reference internal" href="doublets/scrublet/index.html">
         <code class="xref py py-mod docutils literal notranslate">
          <span class="pre">
           ALLCools.clustering.doublets.scrublet
          </span>
         </code>
        </a>
       </li>
      </ul>
     </li>
     <li class="toctree-l3 has-children">
      <a class="reference internal" href="feature_selection/index.html">
       <code class="xref py py-mod docutils literal notranslate">
        <span class="pre">
         ALLCools.clustering.feature_selection
        </span>
       </code>
      </a>
      <input class="toctree-checkbox" id="toctree-checkbox-16" name="toctree-checkbox-16" type="checkbox"/>
      <label for="toctree-checkbox-16">
       <i class="fas fa-chevron-down">
       </i>
      </label>
      <ul>
       <li class="toctree-l4">
        <a class="reference internal" href="feature_selection/feature_enrichment/index.html">
         <code class="xref py py-mod docutils literal notranslate">
          <span class="pre">
           ALLCools.clustering.feature_selection.feature_enrichment
          </span>
         </code>
        </a>
       </li>
       <li class="toctree-l4">
        <a class="reference internal" href="feature_selection/gene_panel_design/index.html">
         <code class="xref py py-mod docutils literal notranslate">
          <span class="pre">
           ALLCools.clustering.feature_selection.gene_panel_design
          </span>
         </code>
        </a>
       </li>
      </ul>
     </li>
     <li class="toctree-l3">
      <a class="reference internal" href="ClusterMerging/index.html">
       <code class="xref py py-mod docutils literal notranslate">
        <span class="pre">
         ALLCools.clustering.ClusterMerging
        </span>
       </code>
      </a>
     </li>
     <li class="toctree-l3">
      <a class="reference internal" href="ConsensusClustering/index.html">
       <code class="xref py py-mod docutils literal notranslate">
        <span class="pre">
         ALLCools.clustering.ConsensusClustering
        </span>
       </code>
      </a>
     </li>
     <li class="toctree-l3">
      <a class="reference internal" href="art_of_tsne/index.html">
       <code class="xref py py-mod docutils literal notranslate">
        <span class="pre">
         ALLCools.clustering.art_of_tsne
        </span>
       </code>
      </a>
     </li>
     <li class="toctree-l3">
      <a class="reference internal" href="balanced_pca/index.html">
       <code class="xref py py-mod docutils literal notranslate">
        <span class="pre">
         ALLCools.clustering.balanced_pca
        </span>
       </code>
      </a>
     </li>
     <li class="toctree-l3">
      <a class="reference internal" href="dmg/index.html">
       <code class="xref py py-mod docutils literal notranslate">
        <span class="pre">
         ALLCools.clustering.dmg
        </span>
       </code>
      </a>
     </li>
     <li class="toctree-l3">
      <a class="reference internal" href="incremental_pca/index.html">
       <code class="xref py py-mod docutils literal notranslate">
        <span class="pre">
         ALLCools.clustering.incremental_pca
        </span>
       </code>
      </a>
     </li>
     <li class="toctree-l3">
      <a class="reference internal" href="mcad/index.html">
       <code class="xref py py-mod docutils literal notranslate">
        <span class="pre">
         ALLCools.clustering.mcad
        </span>
       </code>
      </a>
     </li>
     <li class="toctree-l3">
      <a class="reference internal" href="pvclust/index.html">
       <code class="xref py py-mod docutils literal notranslate">
        <span class="pre">
         ALLCools.clustering.pvclust
        </span>
       </code>
      </a>
     </li>
    </ul>
   </li>
   <li class="toctree-l2 has-children">
    <a class="reference internal" href="../count_matrix/index.html">
     <code class="xref py py-mod docutils literal notranslate">
      <span class="pre">
       ALLCools.count_matrix
      </span>
     </code>
    </a>
    <input class="toctree-checkbox" id="toctree-checkbox-17" name="toctree-checkbox-17" type="checkbox"/>
    <label for="toctree-checkbox-17">
     <i class="fas fa-chevron-down">
     </i>
    </label>
    <ul>
     <li class="toctree-l3">
      <a class="reference internal" href="../count_matrix/atac/index.html">
       <code class="xref py py-mod docutils literal notranslate">
        <span class="pre">
         ALLCools.count_matrix.atac
        </span>
       </code>
      </a>
     </li>
     <li class="toctree-l3">
      <a class="reference internal" href="../count_matrix/dataset/index.html">
       <code class="xref py py-mod docutils literal notranslate">
        <span class="pre">
         ALLCools.count_matrix.dataset
        </span>
       </code>
      </a>
     </li>
     <li class="toctree-l3">
      <a class="reference internal" href="../count_matrix/h5ad/index.html">
       <code class="xref py py-mod docutils literal notranslate">
        <span class="pre">
         ALLCools.count_matrix.h5ad
        </span>
       </code>
      </a>
     </li>
     <li class="toctree-l3">
      <a class="reference internal" href="../count_matrix/mcad/index.html">
       <code class="xref py py-mod docutils literal notranslate">
        <span class="pre">
         ALLCools.count_matrix.mcad
        </span>
       </code>
      </a>
     </li>
     <li class="toctree-l3">
      <a class="reference internal" href="../count_matrix/mcds/index.html">
       <code class="xref py py-mod docutils literal notranslate">
        <span class="pre">
         ALLCools.count_matrix.mcds
        </span>
       </code>
      </a>
     </li>
     <li class="toctree-l3">
      <a class="reference internal" href="../count_matrix/snap/index.html">
       <code class="xref py py-mod docutils literal notranslate">
        <span class="pre">
         ALLCools.count_matrix.snap
        </span>
       </code>
      </a>
     </li>
     <li class="toctree-l3">
      <a class="reference internal" href="../count_matrix/zarr/index.html">
       <code class="xref py py-mod docutils literal notranslate">
        <span class="pre">
         ALLCools.count_matrix.zarr
        </span>
       </code>
      </a>
     </li>
    </ul>
   </li>
   <li class="toctree-l2 has-children">
    <a class="reference internal" href="../dmr/index.html">
     <code class="xref py py-mod docutils literal notranslate">
      <span class="pre">
       ALLCools.dmr
      </span>
     </code>
    </a>
    <input class="toctree-checkbox" id="toctree-checkbox-18" name="toctree-checkbox-18" type="checkbox"/>
    <label for="toctree-checkbox-18">
     <i class="fas fa-chevron-down">
     </i>
    </label>
    <ul>
     <li class="toctree-l3">
      <a class="reference internal" href="../dmr/call_dmr/index.html">
       <code class="xref py py-mod docutils literal notranslate">
        <span class="pre">
         ALLCools.dmr.call_dmr
        </span>
       </code>
      </a>
     </li>
     <li class="toctree-l3">
      <a class="reference internal" href="../dmr/call_dms/index.html">
       <code class="xref py py-mod docutils literal notranslate">
        <span class="pre">
         ALLCools.dmr.call_dms
        </span>
       </code>
      </a>
     </li>
     <li class="toctree-l3">
      <a class="reference internal" href="../dmr/parse_methylpy/index.html">
       <code class="xref py py-mod docutils literal notranslate">
        <span class="pre">
         ALLCools.dmr.parse_methylpy
        </span>
       </code>
      </a>
     </li>
     <li class="toctree-l3">
      <a class="reference internal" href="../dmr/rms_test/index.html">
       <code class="xref py py-mod docutils literal notranslate">
        <span class="pre">
         ALLCools.dmr.rms_test
        </span>
       </code>
      </a>
     </li>
    </ul>
   </li>
   <li class="toctree-l2 has-children">
    <a class="reference internal" href="../gtf/index.html">
     <code class="xref py py-mod docutils literal notranslate">
      <span class="pre">
       ALLCools.gtf
      </span>
     </code>
    </a>
    <input class="toctree-checkbox" id="toctree-checkbox-19" name="toctree-checkbox-19" type="checkbox"/>
    <label for="toctree-checkbox-19">
     <i class="fas fa-chevron-down">
     </i>
    </label>
    <ul>
     <li class="toctree-l3">
      <a class="reference internal" href="../gtf/Gtf/index.html">
       <code class="xref py py-mod docutils literal notranslate">
        <span class="pre">
         ALLCools.gtf.Gtf
        </span>
       </code>
      </a>
     </li>
     <li class="toctree-l3">
      <a class="reference internal" href="../gtf/utilities/index.html">
       <code class="xref py py-mod docutils literal notranslate">
        <span class="pre">
         ALLCools.gtf.utilities
        </span>
       </code>
      </a>
     </li>
    </ul>
   </li>
   <li class="toctree-l2 has-children">
    <a class="reference internal" href="../integration/index.html">
     <code class="xref py py-mod docutils literal notranslate">
      <span class="pre">
       ALLCools.integration
      </span>
     </code>
    </a>
    <input class="toctree-checkbox" id="toctree-checkbox-20" name="toctree-checkbox-20" type="checkbox"/>
    <label for="toctree-checkbox-20">
     <i class="fas fa-chevron-down">
     </i>
    </label>
    <ul>
     <li class="toctree-l3">
      <a class="reference internal" href="../integration/cca/index.html">
       <code class="xref py py-mod docutils literal notranslate">
        <span class="pre">
         ALLCools.integration.cca
        </span>
       </code>
      </a>
     </li>
     <li class="toctree-l3">
      <a class="reference internal" href="../integration/confusion/index.html">
       <code class="xref py py-mod docutils literal notranslate">
        <span class="pre">
         ALLCools.integration.confusion
        </span>
       </code>
      </a>
     </li>
     <li class="toctree-l3">
      <a class="reference internal" href="../integration/harmony/index.html">
       <code class="xref py py-mod docutils literal notranslate">
        <span class="pre">
         ALLCools.integration.harmony
        </span>
       </code>
      </a>
     </li>
    </ul>
   </li>
   <li class="toctree-l2 has-children">
    <a class="reference internal" href="../mcds/index.html">
     <code class="xref py py-mod docutils literal notranslate">
      <span class="pre">
       ALLCools.mcds
      </span>
     </code>
    </a>
    <input class="toctree-checkbox" id="toctree-checkbox-21" name="toctree-checkbox-21" type="checkbox"/>
    <label for="toctree-checkbox-21">
     <i class="fas fa-chevron-down">
     </i>
    </label>
    <ul>
     <li class="toctree-l3">
      <a class="reference internal" href="../mcds/correlation/index.html">
       <code class="xref py py-mod docutils literal notranslate">
        <span class="pre">
         ALLCools.mcds.correlation
        </span>
       </code>
      </a>
     </li>
     <li class="toctree-l3">
      <a class="reference internal" href="../mcds/mcds/index.html">
       <code class="xref py py-mod docutils literal notranslate">
        <span class="pre">
         ALLCools.mcds.mcds
        </span>
       </code>
      </a>
     </li>
     <li class="toctree-l3">
      <a class="reference internal" href="../mcds/region_ds/index.html">
       <code class="xref py py-mod docutils literal notranslate">
        <span class="pre">
         ALLCools.mcds.region_ds
        </span>
       </code>
      </a>
     </li>
     <li class="toctree-l3">
      <a class="reference internal" href="../mcds/utilities/index.html">
       <code class="xref py py-mod docutils literal notranslate">
        <span class="pre">
         ALLCools.mcds.utilities
        </span>
       </code>
      </a>
     </li>
    </ul>
   </li>
   <li class="toctree-l2 has-children">
    <a class="reference internal" href="../motif/index.html">
     <code class="xref py py-mod docutils literal notranslate">
      <span class="pre">
       ALLCools.motif
      </span>
     </code>
    </a>
    <input class="toctree-checkbox" id="toctree-checkbox-22" name="toctree-checkbox-22" type="checkbox"/>
    <label for="toctree-checkbox-22">
     <i class="fas fa-chevron-down">
     </i>
    </label>
    <ul>
     <li class="toctree-l3">
      <a class="reference internal" href="../motif/default_motif_set/index.html">
       <code class="xref py py-mod docutils literal notranslate">
        <span class="pre">
         ALLCools.motif.default_motif_set
        </span>
       </code>
      </a>
     </li>
     <li class="toctree-l3">
      <a class="reference internal" href="../motif/motifs/index.html">
       <code class="xref py py-mod docutils literal notranslate">
        <span class="pre">
         ALLCools.motif.motifs
        </span>
       </code>
      </a>
     </li>
     <li class="toctree-l3">
      <a class="reference internal" href="../motif/parse_meme/index.html">
       <code class="xref py py-mod docutils literal notranslate">
        <span class="pre">
         ALLCools.motif.parse_meme
        </span>
       </code>
      </a>
     </li>
     <li class="toctree-l3">
      <a class="reference internal" href="../motif/snakemake/index.html">
       <code class="xref py py-mod docutils literal notranslate">
        <span class="pre">
         ALLCools.motif.snakemake
        </span>
       </code>
      </a>
     </li>
    </ul>
   </li>
   <li class="toctree-l2 has-children">
    <a class="reference internal" href="../plot/index.html">
     <code class="xref py py-mod docutils literal notranslate">
      <span class="pre">
       ALLCools.plot
      </span>
     </code>
    </a>
    <input class="toctree-checkbox" id="toctree-checkbox-23" name="toctree-checkbox-23" type="checkbox"/>
    <label for="toctree-checkbox-23">
     <i class="fas fa-chevron-down">
     </i>
    </label>
    <ul>
     <li class="toctree-l3 has-children">
      <a class="reference internal" href="../plot/genome_track/index.html">
       <code class="xref py py-mod docutils literal notranslate">
        <span class="pre">
         ALLCools.plot.genome_track
        </span>
       </code>
      </a>
      <input class="toctree-checkbox" id="toctree-checkbox-24" name="toctree-checkbox-24" type="checkbox"/>
      <label for="toctree-checkbox-24">
       <i class="fas fa-chevron-down">
       </i>
      </label>
      <ul>
       <li class="toctree-l4">
        <a class="reference internal" href="../plot/genome_track/GtfTrack/index.html">
         <code class="xref py py-mod docutils literal notranslate">
          <span class="pre">
           ALLCools.plot.genome_track.GtfTrack
          </span>
         </code>
        </a>
       </li>
       <li class="toctree-l4">
        <a class="reference internal" href="../plot/genome_track/HiCMatrixCoolTrack/index.html">
         <code class="xref py py-mod docutils literal notranslate">
          <span class="pre">
           ALLCools.plot.genome_track.HiCMatrixCoolTrack
          </span>
         </code>
        </a>
       </li>
       <li class="toctree-l4">
        <a class="reference internal" href="../plot/genome_track/_pygenometrack/index.html">
         <code class="xref py py-mod docutils literal notranslate">
          <span class="pre">
           ALLCools.plot.genome_track._pygenometrack
          </span>
         </code>
        </a>
       </li>
      </ul>
     </li>
     <li class="toctree-l3">
      <a class="reference internal" href="../plot/categorical_scatter/index.html">
       <code class="xref py py-mod docutils literal notranslate">
        <span class="pre">
         ALLCools.plot.categorical_scatter
        </span>
       </code>
      </a>
     </li>
     <li class="toctree-l3">
      <a class="reference internal" href="../plot/color/index.html">
       <code class="xref py py-mod docutils literal notranslate">
        <span class="pre">
         ALLCools.plot.color
        </span>
       </code>
      </a>
     </li>
     <li class="toctree-l3">
      <a class="reference internal" href="../plot/continuous_scatter/index.html">
       <code class="xref py py-mod docutils literal notranslate">
        <span class="pre">
         ALLCools.plot.continuous_scatter
        </span>
       </code>
      </a>
     </li>
     <li class="toctree-l3">
      <a class="reference internal" href="../plot/contour/index.html">
       <code class="xref py py-mod docutils literal notranslate">
        <span class="pre">
         ALLCools.plot.contour
        </span>
       </code>
      </a>
     </li>
     <li class="toctree-l3">
      <a class="reference internal" href="../plot/decomposition/index.html">
       <code class="xref py py-mod docutils literal notranslate">
        <span class="pre">
         ALLCools.plot.decomposition
        </span>
       </code>
      </a>
     </li>
     <li class="toctree-l3">
      <a class="reference internal" href="../plot/dendro/index.html">
       <code class="xref py py-mod docutils literal notranslate">
        <span class="pre">
         ALLCools.plot.dendro
        </span>
       </code>
      </a>
     </li>
     <li class="toctree-l3">
      <a class="reference internal" href="../plot/interactive_scatter/index.html">
       <code class="xref py py-mod docutils literal notranslate">
        <span class="pre">
         ALLCools.plot.interactive_scatter
        </span>
       </code>
      </a>
     </li>
     <li class="toctree-l3">
      <a class="reference internal" href="../plot/qc_plots/index.html">
       <code class="xref py py-mod docutils literal notranslate">
        <span class="pre">
         ALLCools.plot.qc_plots
        </span>
       </code>
      </a>
     </li>
     <li class="toctree-l3">
      <a class="reference internal" href="../plot/size/index.html">
       <code class="xref py py-mod docutils literal notranslate">
        <span class="pre">
         ALLCools.plot.size
        </span>
       </code>
      </a>
     </li>
     <li class="toctree-l3">
      <a class="reference internal" href="../plot/sunburst/index.html">
       <code class="xref py py-mod docutils literal notranslate">
        <span class="pre">
         ALLCools.plot.sunburst
        </span>
       </code>
      </a>
     </li>
     <li class="toctree-l3">
      <a class="reference internal" href="../plot/text_anno_scatter/index.html">
       <code class="xref py py-mod docutils literal notranslate">
        <span class="pre">
         ALLCools.plot.text_anno_scatter
        </span>
       </code>
      </a>
     </li>
     <li class="toctree-l3">
      <a class="reference internal" href="../plot/utilities/index.html">
       <code class="xref py py-mod docutils literal notranslate">
        <span class="pre">
         ALLCools.plot.utilities
        </span>
       </code>
      </a>
     </li>
    </ul>
   </li>
   <li class="toctree-l2 has-children">
    <a class="reference internal" href="../pseudo_cell/index.html">
     <code class="xref py py-mod docutils literal notranslate">
      <span class="pre">
       ALLCools.pseudo_cell
      </span>
     </code>
    </a>
    <input class="toctree-checkbox" id="toctree-checkbox-25" name="toctree-checkbox-25" type="checkbox"/>
    <label for="toctree-checkbox-25">
     <i class="fas fa-chevron-down">
     </i>
    </label>
    <ul>
     <li class="toctree-l3">
      <a class="reference internal" href="../pseudo_cell/pseudo_cell_kmeans/index.html">
       <code class="xref py py-mod docutils literal notranslate">
        <span class="pre">
         ALLCools.pseudo_cell.pseudo_cell_kmeans
        </span>
       </code>
      </a>
     </li>
     <li class="toctree-l3">
      <a class="reference internal" href="../pseudo_cell/pseudo_cell_knn/index.html">
       <code class="xref py py-mod docutils literal notranslate">
        <span class="pre">
         ALLCools.pseudo_cell.pseudo_cell_knn
        </span>
       </code>
      </a>
     </li>
    </ul>
   </li>
   <li class="toctree-l2 has-children">
    <a class="reference internal" href="../reptile/index.html">
     <code class="xref py py-mod docutils literal notranslate">
      <span class="pre">
       ALLCools.reptile
      </span>
     </code>
    </a>
    <input class="toctree-checkbox" id="toctree-checkbox-26" name="toctree-checkbox-26" type="checkbox"/>
    <label for="toctree-checkbox-26">
     <i class="fas fa-chevron-down">
     </i>
    </label>
    <ul>
     <li class="toctree-l3">
      <a class="reference internal" href="../reptile/reptile/index.html">
       <code class="xref py py-mod docutils literal notranslate">
        <span class="pre">
         ALLCools.reptile.reptile
        </span>
       </code>
      </a>
     </li>
    </ul>
   </li>
   <li class="toctree-l2 has-children">
    <a class="reference internal" href="../sandbox/index.html">
     <code class="xref py py-mod docutils literal notranslate">
      <span class="pre">
       ALLCools.sandbox
      </span>
     </code>
    </a>
    <input class="toctree-checkbox" id="toctree-checkbox-27" name="toctree-checkbox-27" type="checkbox"/>
    <label for="toctree-checkbox-27">
     <i class="fas fa-chevron-down">
     </i>
    </label>
    <ul>
     <li class="toctree-l3 has-children">
      <a class="reference internal" href="../sandbox/igv/index.html">
       <code class="xref py py-mod docutils literal notranslate">
        <span class="pre">
         ALLCools.sandbox.igv
        </span>
       </code>
      </a>
      <input class="toctree-checkbox" id="toctree-checkbox-28" name="toctree-checkbox-28" type="checkbox"/>
      <label for="toctree-checkbox-28">
       <i class="fas fa-chevron-down">
       </i>
      </label>
      <ul>
       <li class="toctree-l4">
        <a class="reference internal" href="../sandbox/igv/BaseClass/index.html">
         <code class="xref py py-mod docutils literal notranslate">
          <span class="pre">
           ALLCools.sandbox.igv.BaseClass
          </span>
         </code>
        </a>
       </li>
       <li class="toctree-l4">
        <a class="reference internal" href="../sandbox/igv/Session/index.html">
         <code class="xref py py-mod docutils literal notranslate">
          <span class="pre">
           ALLCools.sandbox.igv.Session
          </span>
         </code>
        </a>
       </li>
       <li class="toctree-l4">
        <a class="reference internal" href="../sandbox/igv/Track/index.html">
         <code class="xref py py-mod docutils literal notranslate">
          <span class="pre">
           ALLCools.sandbox.igv.Track
          </span>
         </code>
        </a>
       </li>
       <li class="toctree-l4">
        <a class="reference internal" href="../sandbox/igv/defaults/index.html">
         <code class="xref py py-mod docutils literal notranslate">
          <span class="pre">
           ALLCools.sandbox.igv.defaults
          </span>
         </code>
        </a>
       </li>
       <li class="toctree-l4">
        <a class="reference internal" href="../sandbox/igv/make_session/index.html">
         <code class="xref py py-mod docutils literal notranslate">
          <span class="pre">
           ALLCools.sandbox.igv.make_session
          </span>
         </code>
        </a>
       </li>
      </ul>
     </li>
     <li class="toctree-l3 has-children">
      <a class="reference internal" href="../sandbox/motif/index.html">
       <code class="xref py py-mod docutils literal notranslate">
        <span class="pre">
         ALLCools.sandbox.motif
        </span>
       </code>
      </a>
      <input class="toctree-checkbox" id="toctree-checkbox-29" name="toctree-checkbox-29" type="checkbox"/>
      <label for="toctree-checkbox-29">
       <i class="fas fa-chevron-down">
       </i>
      </label>
      <ul>
       <li class="toctree-l4">
        <a class="reference internal" href="../sandbox/motif/ame/index.html">
         <code class="xref py py-mod docutils literal notranslate">
          <span class="pre">
           ALLCools.sandbox.motif.ame
          </span>
         </code>
        </a>
       </li>
       <li class="toctree-l4">
        <a class="reference internal" href="../sandbox/motif/cmotif/index.html">
         <code class="xref py py-mod docutils literal notranslate">
          <span class="pre">
           ALLCools.sandbox.motif.cmotif
          </span>
         </code>
        </a>
       </li>
       <li class="toctree-l4">
        <a class="reference internal" href="../sandbox/motif/fimo/index.html">
         <code class="xref py py-mod docutils literal notranslate">
          <span class="pre">
           ALLCools.sandbox.motif.fimo
          </span>
         </code>
        </a>
       </li>
       <li class="toctree-l4">
        <a class="reference internal" href="../sandbox/motif/motif_scan/index.html">
         <code class="xref py py-mod docutils literal notranslate">
          <span class="pre">
           ALLCools.sandbox.motif.motif_scan
          </span>
         </code>
        </a>
       </li>
       <li class="toctree-l4">
        <a class="reference internal" href="../sandbox/motif/utilities/index.html">
         <code class="xref py py-mod docutils literal notranslate">
          <span class="pre">
           ALLCools.sandbox.motif.utilities
          </span>
         </code>
        </a>
       </li>
      </ul>
     </li>
     <li class="toctree-l3">
      <a class="reference internal" href="../sandbox/balanced_knn/index.html">
       <code class="xref py py-mod docutils literal notranslate">
        <span class="pre">
         ALLCools.sandbox.balanced_knn
        </span>
       </code>
      </a>
     </li>
     <li class="toctree-l3">
      <a class="reference internal" href="../sandbox/scanorama/index.html">
       <code class="xref py py-mod docutils literal notranslate">
        <span class="pre">
         ALLCools.sandbox.scanorama
        </span>
       </code>
      </a>
     </li>
     <li class="toctree-l3">
      <a class="reference internal" href="../sandbox/sccaf/index.html">
       <code class="xref py py-mod docutils literal notranslate">
        <span class="pre">
         ALLCools.sandbox.sccaf
        </span>
       </code>
      </a>
     </li>
    </ul>
   </li>
   <li class="toctree-l2 has-children">
    <a class="reference internal" href="../schema/index.html">
     <code class="xref py py-mod docutils literal notranslate">
      <span class="pre">
       ALLCools.schema
      </span>
     </code>
    </a>
    <input class="toctree-checkbox" id="toctree-checkbox-30" name="toctree-checkbox-30" type="checkbox"/>
    <label for="toctree-checkbox-30">
     <i class="fas fa-chevron-down">
     </i>
    </label>
    <ul>
     <li class="toctree-l3">
      <a class="reference internal" href="../schema/mcds_schema/index.html">
       <code class="xref py py-mod docutils literal notranslate">
        <span class="pre">
         ALLCools.schema.mcds_schema
        </span>
       </code>
      </a>
     </li>
    </ul>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../__main__/index.html">
     <code class="xref py py-mod docutils literal notranslate">
      <span class="pre">
       ALLCools.__main__
      </span>
     </code>
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../_allc_to_bigwig/index.html">
     <code class="xref py py-mod docutils literal notranslate">
      <span class="pre">
       ALLCools._allc_to_bigwig
      </span>
     </code>
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../_allc_to_bigwig_new/index.html">
     <code class="xref py py-mod docutils literal notranslate">
      <span class="pre">
       ALLCools._allc_to_bigwig_new
      </span>
     </code>
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../_allc_to_region_count/index.html">
     <code class="xref py py-mod docutils literal notranslate">
      <span class="pre">
       ALLCools._allc_to_region_count
      </span>
     </code>
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../_bam_to_allc/index.html">
     <code class="xref py py-mod docutils literal notranslate">
      <span class="pre">
       ALLCools._bam_to_allc
      </span>
     </code>
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../_doc/index.html">
     <code class="xref py py-mod docutils literal notranslate">
      <span class="pre">
       ALLCools._doc
      </span>
     </code>
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../_extract_allc/index.html">
     <code class="xref py py-mod docutils literal notranslate">
      <span class="pre">
       ALLCools._extract_allc
      </span>
     </code>
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../_merge_allc/index.html">
     <code class="xref py py-mod docutils literal notranslate">
      <span class="pre">
       ALLCools._merge_allc
      </span>
     </code>
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../_open/index.html">
     <code class="xref py py-mod docutils literal notranslate">
      <span class="pre">
       ALLCools._open
      </span>
     </code>
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../_version/index.html">
     <code class="xref py py-mod docutils literal notranslate">
      <span class="pre">
       ALLCools._version
      </span>
     </code>
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../api/index.html">
     <code class="xref py py-mod docutils literal notranslate">
      <span class="pre">
       ALLCools.api
      </span>
     </code>
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../table_to_allc/index.html">
     <code class="xref py py-mod docutils literal notranslate">
      <span class="pre">
       ALLCools.table_to_allc
      </span>
     </code>
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../utilities/index.html">
     <code class="xref py py-mod docutils literal notranslate">
      <span class="pre">
       ALLCools.utilities
      </span>
     </code>
    </a>
   </li>
  </ul>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  PROJECT INFO
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../../../project_info/citation.html">
   Citation And Reference
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference external" href="https://github.com/lhqing/ALLCools/releases">
   Changelog
  </a>
 </li>
</ul>

    </div>
</nav> <!-- To handle the deprecated key -->

<div class="navbar_extra_footer">
  Powered by <a href="https://jupyterbook.org">Jupyter Book</a>
</div>

</div>


          


          
<main class="col py-md-3 pl-md-4 bd-content overflow-auto" role="main">
    
    <div class="topbar container-xl fixed-top">
    <div class="topbar-contents row">
        <div class="col-12 col-md-3 bd-topbar-whitespace site-navigation show"></div>
        <div class="col pl-md-4 topbar-main">
            
            <button id="navbar-toggler" class="navbar-toggler ml-0" type="button" data-toggle="collapse"
                data-toggle="tooltip" data-placement="bottom" data-target=".site-navigation" aria-controls="navbar-menu"
                aria-expanded="true" aria-label="Toggle navigation" aria-controls="site-navigation"
                title="Toggle navigation" data-toggle="tooltip" data-placement="left">
                <i class="fas fa-bars"></i>
                <i class="fas fa-arrow-left"></i>
                <i class="fas fa-arrow-up"></i>
            </button>
            
            
<div class="dropdown-buttons-trigger">
    <button id="dropdown-buttons-trigger" class="btn btn-secondary topbarbtn" aria-label="Download this page"><i
            class="fas fa-download"></i></button>

    <div class="dropdown-buttons">
        <!-- ipynb file if we had a myst markdown file -->
        
        <!-- Download raw file -->
        <a class="dropdown-buttons" href="../../../_sources/api/ALLCools/clustering/index.rst"><button type="button"
                class="btn btn-secondary topbarbtn" title="Download source file" data-toggle="tooltip"
                data-placement="left">.rst</button></a>
        <!-- Download PDF via print -->
        <button type="button" id="download-print" class="btn btn-secondary topbarbtn" title="Print to PDF"
                onclick="printPdf(this)" data-toggle="tooltip" data-placement="left">.pdf</button>
    </div>
</div>

            <!-- Source interaction buttons -->

<div class="dropdown-buttons-trigger">
    <button id="dropdown-buttons-trigger" class="btn btn-secondary topbarbtn"
        aria-label="Connect with source repository"><i class="fab fa-github"></i></button>
    <div class="dropdown-buttons sourcebuttons">
        <a class="repository-button"
            href="https://github.com/lhqing/ALLCools"><button type="button" class="btn btn-secondary topbarbtn"
                data-toggle="tooltip" data-placement="left" title="Source repository"><i
                    class="fab fa-github"></i>repository</button></a>
        <a class="issues-button"
            href="https://github.com/lhqing/ALLCools/issues/new?title=Issue%20on%20page%20%2Fapi/ALLCools/clustering/index.html&body=Your%20issue%20content%20here."><button
                type="button" class="btn btn-secondary topbarbtn" data-toggle="tooltip" data-placement="left"
                title="Open an issue"><i class="fas fa-lightbulb"></i>open issue</button></a>
        <a class="edit-button" href="https://github.com/lhqing/ALLCools/edit/master/docs/allcools/api/ALLCools/clustering/index.rst"><button
                type="button" class="btn btn-secondary topbarbtn" data-toggle="tooltip" data-placement="left"
                title="Edit this page"><i class="fas fa-pencil-alt"></i>suggest edit</button></a>
    </div>
</div>

            <!-- Full screen (wrap in <a> to have style consistency -->

<a class="full-screen-button"><button type="button" class="btn btn-secondary topbarbtn" data-toggle="tooltip"
        data-placement="bottom" onclick="toggleFullScreen()" aria-label="Fullscreen mode"
        title="Fullscreen mode"><i
            class="fas fa-expand"></i></button></a>

            <!-- Launch buttons -->

        </div>

        <!-- Table of contents -->
        <div class="d-none d-md-block col-md-2 bd-toc show noprint">
            
            <div class="tocsection onthispage pt-5 pb-3">
                <i class="fas fa-list"></i> Contents
            </div>
            <nav id="bd-toc-nav" aria-label="Page">
                <ul class="visible nav section-nav flex-column">
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#subpackages">
   Subpackages
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#submodules">
   Submodules
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#package-contents">
   Package Contents
  </a>
  <ul class="nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#ALLCools.clustering.ConsensusClustering">
     ConsensusClustering
    </a>
    <ul class="nav section-nav flex-column">
     <li class="toc-h4 nav-item toc-entry">
      <a class="reference internal nav-link" href="#ALLCools.clustering.ConsensusClustering.add_data">
       add_data
      </a>
     </li>
     <li class="toc-h4 nav-item toc-entry">
      <a class="reference internal nav-link" href="#ALLCools.clustering.ConsensusClustering.fit_predict">
       fit_predict
      </a>
     </li>
     <li class="toc-h4 nav-item toc-entry">
      <a class="reference internal nav-link" href="#ALLCools.clustering.ConsensusClustering.compute_neighbors">
       compute_neighbors
      </a>
     </li>
     <li class="toc-h4 nav-item toc-entry">
      <a class="reference internal nav-link" href="#ALLCools.clustering.ConsensusClustering.multi_leiden_clustering">
       multi_leiden_clustering
      </a>
     </li>
     <li class="toc-h4 nav-item toc-entry">
      <a class="reference internal nav-link" href="#ALLCools.clustering.ConsensusClustering._summarize_multi_leiden">
       _summarize_multi_leiden
      </a>
     </li>
     <li class="toc-h4 nav-item toc-entry">
      <a class="reference internal nav-link" href="#ALLCools.clustering.ConsensusClustering._create_model">
       _create_model
      </a>
     </li>
     <li class="toc-h4 nav-item toc-entry">
      <a class="reference internal nav-link" href="#ALLCools.clustering.ConsensusClustering.supervise_learning">
       supervise_learning
      </a>
     </li>
     <li class="toc-h4 nav-item toc-entry">
      <a class="reference internal nav-link" href="#ALLCools.clustering.ConsensusClustering.final_evaluation">
       final_evaluation
      </a>
     </li>
     <li class="toc-h4 nav-item toc-entry">
      <a class="reference internal nav-link" href="#ALLCools.clustering.ConsensusClustering.save">
       save
      </a>
     </li>
     <li class="toc-h4 nav-item toc-entry">
      <a class="reference internal nav-link" href="#ALLCools.clustering.ConsensusClustering.plot_leiden_cases">
       plot_leiden_cases
      </a>
     </li>
     <li class="toc-h4 nav-item toc-entry">
      <a class="reference internal nav-link" href="#ALLCools.clustering.ConsensusClustering.plot_before_after">
       plot_before_after
      </a>
     </li>
     <li class="toc-h4 nav-item toc-entry">
      <a class="reference internal nav-link" href="#ALLCools.clustering.ConsensusClustering.plot_steps">
       plot_steps
      </a>
     </li>
     <li class="toc-h4 nav-item toc-entry">
      <a class="reference internal nav-link" href="#ALLCools.clustering.ConsensusClustering.plot_merge_process">
       plot_merge_process
      </a>
     </li>
    </ul>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#ALLCools.clustering.select_confusion_pairs">
     select_confusion_pairs
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#ALLCools.clustering.tsne">
     tsne
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#ALLCools.clustering.balanced_pca">
     balanced_pca
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#ALLCools.clustering.significant_pc_test">
     significant_pc_test
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#ALLCools.clustering.log_scale">
     log_scale
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#ALLCools.clustering.get_pc_centers">
     get_pc_centers
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#ALLCools.clustering.ReproduciblePCA">
     ReproduciblePCA
    </a>
    <ul class="nav section-nav flex-column">
     <li class="toc-h4 nav-item toc-entry">
      <a class="reference internal nav-link" href="#ALLCools.clustering.ReproduciblePCA.mcds_to_adata">
       mcds_to_adata
      </a>
     </li>
     <li class="toc-h4 nav-item toc-entry">
      <a class="reference internal nav-link" href="#ALLCools.clustering.ReproduciblePCA.scale">
       scale
      </a>
     </li>
     <li class="toc-h4 nav-item toc-entry">
      <a class="reference internal nav-link" href="#ALLCools.clustering.ReproduciblePCA.pc_transform">
       pc_transform
      </a>
     </li>
     <li class="toc-h4 nav-item toc-entry">
      <a class="reference internal nav-link" href="#ALLCools.clustering.ReproduciblePCA.mcds_to_adata_with_pc">
       mcds_to_adata_with_pc
      </a>
     </li>
     <li class="toc-h4 nav-item toc-entry">
      <a class="reference internal nav-link" href="#ALLCools.clustering.ReproduciblePCA.dump">
       dump
      </a>
     </li>
    </ul>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#ALLCools.clustering.Dendrogram">
     Dendrogram
    </a>
    <ul class="nav section-nav flex-column">
     <li class="toc-h4 nav-item toc-entry">
      <a class="reference internal nav-link" href="#ALLCools.clustering.Dendrogram.fit">
       fit
      </a>
     </li>
     <li class="toc-h4 nav-item toc-entry">
      <a class="reference internal nav-link" href="#ALLCools.clustering.Dendrogram.save">
       save
      </a>
     </li>
    </ul>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#ALLCools.clustering.PairwiseDMG">
     PairwiseDMG
    </a>
    <ul class="nav section-nav flex-column">
     <li class="toc-h4 nav-item toc-entry">
      <a class="reference internal nav-link" href="#ALLCools.clustering.PairwiseDMG.fit_predict">
       fit_predict
      </a>
     </li>
     <li class="toc-h4 nav-item toc-entry">
      <a class="reference internal nav-link" href="#ALLCools.clustering.PairwiseDMG._save_cluster_adata">
       _save_cluster_adata
      </a>
     </li>
     <li class="toc-h4 nav-item toc-entry">
      <a class="reference internal nav-link" href="#ALLCools.clustering.PairwiseDMG._pairwise_dmg">
       _pairwise_dmg
      </a>
     </li>
     <li class="toc-h4 nav-item toc-entry">
      <a class="reference internal nav-link" href="#ALLCools.clustering.PairwiseDMG._cleanup">
       _cleanup
      </a>
     </li>
     <li class="toc-h4 nav-item toc-entry">
      <a class="reference internal nav-link" href="#ALLCools.clustering.PairwiseDMG.aggregate_pairwise_dmg">
       aggregate_pairwise_dmg
      </a>
     </li>
    </ul>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#ALLCools.clustering.one_vs_rest_dmg">
     one_vs_rest_dmg
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#ALLCools.clustering.cluster_enriched_features">
     cluster_enriched_features
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#ALLCools.clustering.filter_regions">
     filter_regions
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#ALLCools.clustering.lsi">
     lsi
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#ALLCools.clustering.binarize_matrix">
     binarize_matrix
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#ALLCools.clustering.remove_black_list_region">
     remove_black_list_region
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#ALLCools.clustering.ClusterMerge">
     ClusterMerge
    </a>
    <ul class="nav section-nav flex-column">
     <li class="toc-h4 nav-item toc-entry">
      <a class="reference internal nav-link" href="#ALLCools.clustering.ClusterMerge._construct_tree">
       _construct_tree
      </a>
     </li>
     <li class="toc-h4 nav-item toc-entry">
      <a class="reference internal nav-link" href="#ALLCools.clustering.ClusterMerge._traverse">
       _traverse
      </a>
     </li>
     <li class="toc-h4 nav-item toc-entry">
      <a class="reference internal nav-link" href="#ALLCools.clustering.ClusterMerge._merge_pair">
       _merge_pair
      </a>
     </li>
     <li class="toc-h4 nav-item toc-entry">
      <a class="reference internal nav-link" href="#ALLCools.clustering.ClusterMerge.fit_predict">
       fit_predict
      </a>
     </li>
    </ul>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#ALLCools.clustering.PairwiseDMGCriterion">
     PairwiseDMGCriterion
    </a>
    <ul class="nav section-nav flex-column">
     <li class="toc-h4 nav-item toc-entry">
      <a class="reference internal nav-link" href="#ALLCools.clustering.PairwiseDMGCriterion.predict">
       predict
      </a>
     </li>
    </ul>
   </li>
  </ul>
 </li>
</ul>

            </nav>
        </div>
    </div>
</div>
    <div id="main-content" class="row">
        <div class="col-12 col-md-9 pl-md-3 pr-md-0">
            <!-- Table of contents that is only displayed when printing the page -->
            <div id="jb-print-docs-body" class="onlyprint">
                <h1>ALLCools.clustering</h1>
                <!-- Table of contents -->
                <div id="print-main-content">
                    <div id="jb-print-toc">
                        
                        <div>
                            <h2> Contents </h2>
                        </div>
                        <nav aria-label="Page">
                            <ul class="visible nav section-nav flex-column">
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#subpackages">
   Subpackages
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#submodules">
   Submodules
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#package-contents">
   Package Contents
  </a>
  <ul class="nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#ALLCools.clustering.ConsensusClustering">
     ConsensusClustering
    </a>
    <ul class="nav section-nav flex-column">
     <li class="toc-h4 nav-item toc-entry">
      <a class="reference internal nav-link" href="#ALLCools.clustering.ConsensusClustering.add_data">
       add_data
      </a>
     </li>
     <li class="toc-h4 nav-item toc-entry">
      <a class="reference internal nav-link" href="#ALLCools.clustering.ConsensusClustering.fit_predict">
       fit_predict
      </a>
     </li>
     <li class="toc-h4 nav-item toc-entry">
      <a class="reference internal nav-link" href="#ALLCools.clustering.ConsensusClustering.compute_neighbors">
       compute_neighbors
      </a>
     </li>
     <li class="toc-h4 nav-item toc-entry">
      <a class="reference internal nav-link" href="#ALLCools.clustering.ConsensusClustering.multi_leiden_clustering">
       multi_leiden_clustering
      </a>
     </li>
     <li class="toc-h4 nav-item toc-entry">
      <a class="reference internal nav-link" href="#ALLCools.clustering.ConsensusClustering._summarize_multi_leiden">
       _summarize_multi_leiden
      </a>
     </li>
     <li class="toc-h4 nav-item toc-entry">
      <a class="reference internal nav-link" href="#ALLCools.clustering.ConsensusClustering._create_model">
       _create_model
      </a>
     </li>
     <li class="toc-h4 nav-item toc-entry">
      <a class="reference internal nav-link" href="#ALLCools.clustering.ConsensusClustering.supervise_learning">
       supervise_learning
      </a>
     </li>
     <li class="toc-h4 nav-item toc-entry">
      <a class="reference internal nav-link" href="#ALLCools.clustering.ConsensusClustering.final_evaluation">
       final_evaluation
      </a>
     </li>
     <li class="toc-h4 nav-item toc-entry">
      <a class="reference internal nav-link" href="#ALLCools.clustering.ConsensusClustering.save">
       save
      </a>
     </li>
     <li class="toc-h4 nav-item toc-entry">
      <a class="reference internal nav-link" href="#ALLCools.clustering.ConsensusClustering.plot_leiden_cases">
       plot_leiden_cases
      </a>
     </li>
     <li class="toc-h4 nav-item toc-entry">
      <a class="reference internal nav-link" href="#ALLCools.clustering.ConsensusClustering.plot_before_after">
       plot_before_after
      </a>
     </li>
     <li class="toc-h4 nav-item toc-entry">
      <a class="reference internal nav-link" href="#ALLCools.clustering.ConsensusClustering.plot_steps">
       plot_steps
      </a>
     </li>
     <li class="toc-h4 nav-item toc-entry">
      <a class="reference internal nav-link" href="#ALLCools.clustering.ConsensusClustering.plot_merge_process">
       plot_merge_process
      </a>
     </li>
    </ul>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#ALLCools.clustering.select_confusion_pairs">
     select_confusion_pairs
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#ALLCools.clustering.tsne">
     tsne
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#ALLCools.clustering.balanced_pca">
     balanced_pca
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#ALLCools.clustering.significant_pc_test">
     significant_pc_test
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#ALLCools.clustering.log_scale">
     log_scale
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#ALLCools.clustering.get_pc_centers">
     get_pc_centers
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#ALLCools.clustering.ReproduciblePCA">
     ReproduciblePCA
    </a>
    <ul class="nav section-nav flex-column">
     <li class="toc-h4 nav-item toc-entry">
      <a class="reference internal nav-link" href="#ALLCools.clustering.ReproduciblePCA.mcds_to_adata">
       mcds_to_adata
      </a>
     </li>
     <li class="toc-h4 nav-item toc-entry">
      <a class="reference internal nav-link" href="#ALLCools.clustering.ReproduciblePCA.scale">
       scale
      </a>
     </li>
     <li class="toc-h4 nav-item toc-entry">
      <a class="reference internal nav-link" href="#ALLCools.clustering.ReproduciblePCA.pc_transform">
       pc_transform
      </a>
     </li>
     <li class="toc-h4 nav-item toc-entry">
      <a class="reference internal nav-link" href="#ALLCools.clustering.ReproduciblePCA.mcds_to_adata_with_pc">
       mcds_to_adata_with_pc
      </a>
     </li>
     <li class="toc-h4 nav-item toc-entry">
      <a class="reference internal nav-link" href="#ALLCools.clustering.ReproduciblePCA.dump">
       dump
      </a>
     </li>
    </ul>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#ALLCools.clustering.Dendrogram">
     Dendrogram
    </a>
    <ul class="nav section-nav flex-column">
     <li class="toc-h4 nav-item toc-entry">
      <a class="reference internal nav-link" href="#ALLCools.clustering.Dendrogram.fit">
       fit
      </a>
     </li>
     <li class="toc-h4 nav-item toc-entry">
      <a class="reference internal nav-link" href="#ALLCools.clustering.Dendrogram.save">
       save
      </a>
     </li>
    </ul>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#ALLCools.clustering.PairwiseDMG">
     PairwiseDMG
    </a>
    <ul class="nav section-nav flex-column">
     <li class="toc-h4 nav-item toc-entry">
      <a class="reference internal nav-link" href="#ALLCools.clustering.PairwiseDMG.fit_predict">
       fit_predict
      </a>
     </li>
     <li class="toc-h4 nav-item toc-entry">
      <a class="reference internal nav-link" href="#ALLCools.clustering.PairwiseDMG._save_cluster_adata">
       _save_cluster_adata
      </a>
     </li>
     <li class="toc-h4 nav-item toc-entry">
      <a class="reference internal nav-link" href="#ALLCools.clustering.PairwiseDMG._pairwise_dmg">
       _pairwise_dmg
      </a>
     </li>
     <li class="toc-h4 nav-item toc-entry">
      <a class="reference internal nav-link" href="#ALLCools.clustering.PairwiseDMG._cleanup">
       _cleanup
      </a>
     </li>
     <li class="toc-h4 nav-item toc-entry">
      <a class="reference internal nav-link" href="#ALLCools.clustering.PairwiseDMG.aggregate_pairwise_dmg">
       aggregate_pairwise_dmg
      </a>
     </li>
    </ul>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#ALLCools.clustering.one_vs_rest_dmg">
     one_vs_rest_dmg
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#ALLCools.clustering.cluster_enriched_features">
     cluster_enriched_features
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#ALLCools.clustering.filter_regions">
     filter_regions
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#ALLCools.clustering.lsi">
     lsi
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#ALLCools.clustering.binarize_matrix">
     binarize_matrix
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#ALLCools.clustering.remove_black_list_region">
     remove_black_list_region
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#ALLCools.clustering.ClusterMerge">
     ClusterMerge
    </a>
    <ul class="nav section-nav flex-column">
     <li class="toc-h4 nav-item toc-entry">
      <a class="reference internal nav-link" href="#ALLCools.clustering.ClusterMerge._construct_tree">
       _construct_tree
      </a>
     </li>
     <li class="toc-h4 nav-item toc-entry">
      <a class="reference internal nav-link" href="#ALLCools.clustering.ClusterMerge._traverse">
       _traverse
      </a>
     </li>
     <li class="toc-h4 nav-item toc-entry">
      <a class="reference internal nav-link" href="#ALLCools.clustering.ClusterMerge._merge_pair">
       _merge_pair
      </a>
     </li>
     <li class="toc-h4 nav-item toc-entry">
      <a class="reference internal nav-link" href="#ALLCools.clustering.ClusterMerge.fit_predict">
       fit_predict
      </a>
     </li>
    </ul>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#ALLCools.clustering.PairwiseDMGCriterion">
     PairwiseDMGCriterion
    </a>
    <ul class="nav section-nav flex-column">
     <li class="toc-h4 nav-item toc-entry">
      <a class="reference internal nav-link" href="#ALLCools.clustering.PairwiseDMGCriterion.predict">
       predict
      </a>
     </li>
    </ul>
   </li>
  </ul>
 </li>
</ul>

                        </nav>
                    </div>
                </div>
            </div>
            
              <div>
                
  <div class="section" id="module-ALLCools.clustering">
<span id="allcools-clustering"></span><h1><a class="reference internal" href="#module-ALLCools.clustering" title="ALLCools.clustering"><code class="xref py py-mod docutils literal notranslate"><span class="pre">ALLCools.clustering</span></code></a><a class="headerlink" href="#module-ALLCools.clustering" title="Permalink to this headline"></a></h1>
<p>Basic Cellular Analysis Functions</p>
<div class="section" id="subpackages">
<h2>Subpackages<a class="headerlink" href="#subpackages" title="Permalink to this headline"></a></h2>
<div class="toctree-wrapper compound">
<ul>
<li class="toctree-l1"><a class="reference internal" href="doublets/index.html"><code class="xref py py-mod docutils literal notranslate"><span class="pre">ALLCools.clustering.doublets</span></code></a><ul>
<li class="toctree-l2"><a class="reference internal" href="doublets/coverage_doublets/index.html"><code class="xref py py-mod docutils literal notranslate"><span class="pre">ALLCools.clustering.doublets.coverage_doublets</span></code></a></li>
<li class="toctree-l2"><a class="reference internal" href="doublets/scrublet/index.html"><code class="xref py py-mod docutils literal notranslate"><span class="pre">ALLCools.clustering.doublets.scrublet</span></code></a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="feature_selection/index.html"><code class="xref py py-mod docutils literal notranslate"><span class="pre">ALLCools.clustering.feature_selection</span></code></a><ul>
<li class="toctree-l2"><a class="reference internal" href="feature_selection/feature_enrichment/index.html"><code class="xref py py-mod docutils literal notranslate"><span class="pre">ALLCools.clustering.feature_selection.feature_enrichment</span></code></a></li>
<li class="toctree-l2"><a class="reference internal" href="feature_selection/gene_panel_design/index.html"><code class="xref py py-mod docutils literal notranslate"><span class="pre">ALLCools.clustering.feature_selection.gene_panel_design</span></code></a></li>
</ul>
</li>
</ul>
</div>
</div>
<div class="section" id="submodules">
<h2>Submodules<a class="headerlink" href="#submodules" title="Permalink to this headline"></a></h2>
<div class="toctree-wrapper compound">
<ul>
<li class="toctree-l1"><a class="reference internal" href="ClusterMerging/index.html"><code class="xref py py-mod docutils literal notranslate"><span class="pre">ALLCools.clustering.ClusterMerging</span></code></a></li>
<li class="toctree-l1"><a class="reference internal" href="ConsensusClustering/index.html"><code class="xref py py-mod docutils literal notranslate"><span class="pre">ALLCools.clustering.ConsensusClustering</span></code></a></li>
<li class="toctree-l1"><a class="reference internal" href="art_of_tsne/index.html"><code class="xref py py-mod docutils literal notranslate"><span class="pre">ALLCools.clustering.art_of_tsne</span></code></a></li>
<li class="toctree-l1"><a class="reference internal" href="balanced_pca/index.html"><code class="xref py py-mod docutils literal notranslate"><span class="pre">ALLCools.clustering.balanced_pca</span></code></a></li>
<li class="toctree-l1"><a class="reference internal" href="dmg/index.html"><code class="xref py py-mod docutils literal notranslate"><span class="pre">ALLCools.clustering.dmg</span></code></a></li>
<li class="toctree-l1"><a class="reference internal" href="incremental_pca/index.html"><code class="xref py py-mod docutils literal notranslate"><span class="pre">ALLCools.clustering.incremental_pca</span></code></a></li>
<li class="toctree-l1"><a class="reference internal" href="mcad/index.html"><code class="xref py py-mod docutils literal notranslate"><span class="pre">ALLCools.clustering.mcad</span></code></a></li>
<li class="toctree-l1"><a class="reference internal" href="pvclust/index.html"><code class="xref py py-mod docutils literal notranslate"><span class="pre">ALLCools.clustering.pvclust</span></code></a></li>
</ul>
</div>
</div>
<div class="section" id="package-contents">
<h2>Package Contents<a class="headerlink" href="#package-contents" title="Permalink to this headline"></a></h2>
<dl class="py class">
<dt class="sig sig-object py" id="ALLCools.clustering.ConsensusClustering">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">ConsensusClustering</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">model</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n_neighbors</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">25</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">metric</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'euclidean'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">min_cluster_size</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">10</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">leiden_repeats</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">200</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">leiden_resolution</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">target_accuracy</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.95</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">consensus_rate</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.7</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">random_state</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">train_frac</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.5</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">train_max_n</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">500</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">max_iter</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">50</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n_jobs</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">-</span> <span class="pre">1</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#ALLCools.clustering.ConsensusClustering" title="Permalink to this definition"></a></dt>
<dd><dl class="py method">
<dt class="sig sig-object py" id="ALLCools.clustering.ConsensusClustering.add_data">
<span class="sig-name descname"><span class="pre">add_data</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">self</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">x</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#ALLCools.clustering.ConsensusClustering.add_data" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="ALLCools.clustering.ConsensusClustering.fit_predict">
<span class="sig-name descname"><span class="pre">fit_predict</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">self</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">x</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">leiden_kwds</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#ALLCools.clustering.ConsensusClustering.fit_predict" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="ALLCools.clustering.ConsensusClustering.compute_neighbors">
<span class="sig-name descname"><span class="pre">compute_neighbors</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">self</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#ALLCools.clustering.ConsensusClustering.compute_neighbors" title="Permalink to this definition"></a></dt>
<dd><p>Calculate KNN graph</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="ALLCools.clustering.ConsensusClustering.multi_leiden_clustering">
<span class="sig-name descname"><span class="pre">multi_leiden_clustering</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">self</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">partition_type</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">partition_kwargs</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">use_weights</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n_iterations</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">-</span> <span class="pre">1</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#ALLCools.clustering.ConsensusClustering.multi_leiden_clustering" title="Permalink to this definition"></a></dt>
<dd><p>Modified from scanpy, perform Leiden clustering multiple times with different random states</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="ALLCools.clustering.ConsensusClustering._summarize_multi_leiden">
<span class="sig-name descname"><span class="pre">_summarize_multi_leiden</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">self</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#ALLCools.clustering.ConsensusClustering._summarize_multi_leiden" title="Permalink to this definition"></a></dt>
<dd><p>Summarize the multi_leiden results,
generate a raw cluster version simply based on the hamming distance
between cells and split cluster with cutoff (consensus_rate)</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="ALLCools.clustering.ConsensusClustering._create_model">
<span class="sig-name descname"><span class="pre">_create_model</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">self</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n_estimators</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1000</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#ALLCools.clustering.ConsensusClustering._create_model" title="Permalink to this definition"></a></dt>
<dd><p>Init default model</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="ALLCools.clustering.ConsensusClustering.supervise_learning">
<span class="sig-name descname"><span class="pre">supervise_learning</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">self</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#ALLCools.clustering.ConsensusClustering.supervise_learning" title="Permalink to this definition"></a></dt>
<dd><p>Perform supervised learning and cluster merge process</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="ALLCools.clustering.ConsensusClustering.final_evaluation">
<span class="sig-name descname"><span class="pre">final_evaluation</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">self</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#ALLCools.clustering.ConsensusClustering.final_evaluation" title="Permalink to this definition"></a></dt>
<dd><p>Final evaluation of the model and assign outliers</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="ALLCools.clustering.ConsensusClustering.save">
<span class="sig-name descname"><span class="pre">save</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">self</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">output_path</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#ALLCools.clustering.ConsensusClustering.save" title="Permalink to this definition"></a></dt>
<dd><p>Save the model</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="ALLCools.clustering.ConsensusClustering.plot_leiden_cases">
<span class="sig-name descname"><span class="pre">plot_leiden_cases</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">self</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">coord_data</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">coord_base</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'umap'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">plot_size</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">3</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dpi</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">300</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">plot_n_cases</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">4</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">s</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">3</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#ALLCools.clustering.ConsensusClustering.plot_leiden_cases" title="Permalink to this definition"></a></dt>
<dd><p>Show some leiden runs with biggest different as measured by ARI</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="ALLCools.clustering.ConsensusClustering.plot_before_after">
<span class="sig-name descname"><span class="pre">plot_before_after</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">self</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">coord_data</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">coord_base</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'umap'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">plot_size</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">3</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dpi</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">300</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#ALLCools.clustering.ConsensusClustering.plot_before_after" title="Permalink to this definition"></a></dt>
<dd><p>Plot the raw clusters from multi-leiden and final clusters after merge</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="ALLCools.clustering.ConsensusClustering.plot_steps">
<span class="sig-name descname"><span class="pre">plot_steps</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">self</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">coord_data</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">coord_base</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'umap'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">plot_size</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">3</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dpi</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">300</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#ALLCools.clustering.ConsensusClustering.plot_steps" title="Permalink to this definition"></a></dt>
<dd><p>Plot the supervised learning and merge steps</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="ALLCools.clustering.ConsensusClustering.plot_merge_process">
<span class="sig-name descname"><span class="pre">plot_merge_process</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">self</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">plot_size</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">3</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#ALLCools.clustering.ConsensusClustering.plot_merge_process" title="Permalink to this definition"></a></dt>
<dd><p>Plot the change of accuracy during merge</p>
</dd></dl>

</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="ALLCools.clustering.select_confusion_pairs">
<span class="sig-name descname"><span class="pre">select_confusion_pairs</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">true_label</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">predicted_label</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ratio_cutoff</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.001</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/ALLCools/clustering/ConsensusClustering.html#select_confusion_pairs"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#ALLCools.clustering.select_confusion_pairs" title="Permalink to this definition"></a></dt>
<dd><p>Select cluster pairs that are confusing (ratio_cutoff) between true and predicted labels</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>true_label</strong> (<em>true cell labels</em>)  </p></li>
<li><p><strong>predicted_label</strong> (<em>predicted cell labels</em>)  </p></li>
<li><p><strong>ratio_cutoff</strong> (<em>ratio of clusters cutoff to define confusion</em>)  </p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>list of cluster pair tuples</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>confused_pairs</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="ALLCools.clustering.tsne">
<span class="sig-name descname"><span class="pre">tsne</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">adata</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">obsm</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'X_pca'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">metric</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Union</span><span class="p"><span class="pre">[</span></span><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.10)"><span class="pre">str</span></a><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">Callable</span><span class="p"><span class="pre">]</span></span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'euclidean'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">exaggeration</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.10)"><span class="pre">float</span></a></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">-</span> <span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">perplexity</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.10)"><span class="pre">int</span></a></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">30</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n_jobs</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.10)"><span class="pre">int</span></a></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">-</span> <span class="pre">1</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/ALLCools/clustering/art_of_tsne.html#tsne"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#ALLCools.clustering.tsne" title="Permalink to this definition"></a></dt>
<dd><p>Calculating T-SNE embedding with the openTSNE package <span id="id1">[<a class="reference internal" href="../../../project_info/citation.html#id12">Poliar <em>et al.</em>, 2019</a>]</span> and
parameter optimization strategy described in <span id="id2">[<a class="reference internal" href="../../../project_info/citation.html#id11">Kobak and Berens, 2019</a>]</span>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>adata</strong>  adata object with principle components or equivalent matrix stored in .obsm</p></li>
<li><p><strong>obsm</strong>  name of the matrix in .obsm that can be used as T-SNE input</p></li>
<li><p><strong>metric</strong>  Any metric allowed by PyNNDescent (default: euclidean)</p></li>
<li><p><strong>exaggeration</strong>  The exaggeration to use for the embedding</p></li>
<li><p><strong>perplexity</strong>  The perplexity to use for the embedding</p></li>
<li><p><strong>n_jobs</strong>  Number of CPUs to use</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p></p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>T-SNE embedding will be stored at adata.obsm[X_tsne]</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="ALLCools.clustering.balanced_pca">
<span class="sig-name descname"><span class="pre">balanced_pca</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">adata</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://anndata.readthedocs.io/en/stable/generated/anndata.AnnData.html#anndata.AnnData" title="(in anndata v0.8.0)"><span class="pre">anndata.AnnData</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">groups</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.10)"><span class="pre">str</span></a></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'pre_clusters'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">max_cell_prop</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n_comps</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">200</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">scale</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/ALLCools/clustering/balanced_pca.html#balanced_pca"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#ALLCools.clustering.balanced_pca" title="Permalink to this definition"></a></dt>
<dd><p>Given a categorical variable (e.g., a pre-clustering label), perform balanced PCA by downsample
cells in the large categories to make the overall population more balanced, so the PCs are expected
to represent more variance among small categories.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>adata</strong>  adata after preprocessing and feature selection steps</p></li>
<li><p><strong>groups</strong>  the name of the categorical variable in adata.obsm</p></li>
<li><p><strong>max_cell_prop</strong>  any single category with cells &gt; <cite>n_cell * max_cell_prop</cite> will be downsampled to this number.</p></li>
<li><p><strong>n_comps</strong>  Number of components in PCA</p></li>
<li><p><strong>scale</strong>  whether to scale the input matrix before PCA</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p></p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>adata with PC information stored in obsm, varm and uns like the <a class="reference external" href="https://scanpy.readthedocs.io/en/stable/generated/scanpy.tl.pca.html#scanpy.tl.pca" title="(in Scanpy v1.9.1)"><code class="xref py py-func docutils literal notranslate"><span class="pre">scanpy.tl.pca()</span></code></a> do.</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="ALLCools.clustering.significant_pc_test">
<span class="sig-name descname"><span class="pre">significant_pc_test</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">adata</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://anndata.readthedocs.io/en/stable/generated/anndata.AnnData.html#anndata.AnnData" title="(in anndata v0.8.0)"><span class="pre">anndata.AnnData</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">p_cutoff</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">update</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">obsm</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'X_pca'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">downsample</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">50000</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/ALLCools/clustering/balanced_pca.html#significant_pc_test"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#ALLCools.clustering.significant_pc_test" title="Permalink to this definition"></a></dt>
<dd><p>Perform two-sample Kolmogorov-Smirnov test for goodness of fit on two adjacent PCs,
select top PCs based on the <cite>p_cutoff</cite>. Top PCs have significantly different distributions, while
later PCs only capturing random noise will have larger p-values. An idea from <span id="id3">[<a class="reference internal" href="../../../project_info/citation.html#id10">Zeisel <em>et al.</em>, 2018</a>]</span>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>adata</strong>  adata with PC matrix calculated and stored in adata.obsm</p></li>
<li><p><strong>p_cutoff</strong>  the p-value cutoff to select top PCs</p></li>
<li><p><strong>update</strong>  Whether modify adata.obsm and only keep significant PCs</p></li>
<li><p><strong>obsm</strong>  name of the PC matrix in adata.obsm</p></li>
<li><p><strong>downsample</strong>  If the dataset is too large, downsample the cells before testing.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>number of PCs selected</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>n_components</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="ALLCools.clustering.log_scale">
<span class="sig-name descname"><span class="pre">log_scale</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">adata</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">method</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'standard'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">with_mean</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">with_std</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">max_value</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">10</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">scaler</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/ALLCools/clustering/balanced_pca.html#log_scale"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#ALLCools.clustering.log_scale" title="Permalink to this definition"></a></dt>
<dd><p>Perform log transform and then scale the cell-by-feature matrix</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>adata</strong>  adata with normalized, unscaled cell-by-feature matrix</p></li>
<li><p><strong>method</strong>  the type of scaler to use:
standard for <a class="reference external" href="https://scikit-learn.org/stable/modules/generated/sklearn.preprocessing.StandardScaler.html#sklearn.preprocessing.StandardScaler" title="(in scikit-learn v1.0)"><code class="xref py py-class docutils literal notranslate"><span class="pre">sklearn.preprocessing.StandardScaler</span></code></a>;
robust for <a class="reference external" href="https://scikit-learn.org/stable/modules/generated/sklearn.preprocessing.RobustScaler.html#sklearn.preprocessing.RobustScaler" title="(in scikit-learn v1.0)"><code class="xref py py-class docutils literal notranslate"><span class="pre">sklearn.preprocessing.RobustScaler</span></code></a>.</p></li>
<li><p><strong>with_mean</strong>  Whether scale with mean center</p></li>
<li><p><strong>with_std</strong>  Whether scale the std</p></li>
<li><p><strong>max_value</strong>  Whether clip large values after scale</p></li>
<li><p><strong>scaler</strong>  A fitted sklearn scaler, if provided, will only use it to transform the adata.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p></p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>adata.X is scaled in place, the fitted scaler object will be return if the <cite>scaler</cite> parameter is None.</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="ALLCools.clustering.get_pc_centers">
<span class="sig-name descname"><span class="pre">get_pc_centers</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">adata</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://anndata.readthedocs.io/en/stable/generated/anndata.AnnData.html#anndata.AnnData" title="(in anndata v0.8.0)"><span class="pre">anndata.AnnData</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">group</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.10)"><span class="pre">str</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">outlier_label</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">obsm</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'X_pca'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/ALLCools/clustering/balanced_pca.html#get_pc_centers"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#ALLCools.clustering.get_pc_centers" title="Permalink to this definition"></a></dt>
<dd><p>Get the cluster centroids of the PC matrix</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>adata</strong>  adata with cluster labels in obs and PC in obsm</p></li>
<li><p><strong>group</strong>  the name of cluster labels in adata.obs</p></li>
<li><p><strong>outlier_label</strong>  if there are outlier labels in obs, will exclude outliers before calculating centroids.</p></li>
<li><p><strong>obsm</strong>  the key of PC matrix in obsm</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>a dataframe for cluster centroids by PC</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>pc_center</p>
</dd>
</dl>
</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="ALLCools.clustering.ReproduciblePCA">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">ReproduciblePCA</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">scaler</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mc_type</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">adata</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pca_obj</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pc_loading</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">var_names</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">max_value</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">10</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/ALLCools/clustering/balanced_pca.html#ReproduciblePCA"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#ALLCools.clustering.ReproduciblePCA" title="Permalink to this definition"></a></dt>
<dd><dl class="py method">
<dt class="sig sig-object py" id="ALLCools.clustering.ReproduciblePCA.mcds_to_adata">
<span class="sig-name descname"><span class="pre">mcds_to_adata</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">self</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mcds</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#ALLCools.clustering.ReproduciblePCA.mcds_to_adata" title="Permalink to this definition"></a></dt>
<dd><p>Get adata from MCDS with only selected features</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>mcds</strong> (<em>Input raw count MCDS object</em>)  </p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Adata with per-cell normalized mC fraction and selected features</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>adata</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="ALLCools.clustering.ReproduciblePCA.scale">
<span class="sig-name descname"><span class="pre">scale</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">self</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">adata</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#ALLCools.clustering.ReproduciblePCA.scale" title="Permalink to this definition"></a></dt>
<dd><p>Perform {func}`log_scale &lt;ALLCools.clustering.balanced_pca.log_scale&gt;` with fitted scaler</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>adata</strong>  Adata with per-cell normalized mC fraction and selected features</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p></p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>adata.X is transformed in place</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="ALLCools.clustering.ReproduciblePCA.pc_transform">
<span class="sig-name descname"><span class="pre">pc_transform</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">self</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">adata</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#ALLCools.clustering.ReproduciblePCA.pc_transform" title="Permalink to this definition"></a></dt>
<dd><p>calculate the PC from adata.X and PC loading, store PCs in adata.obsm[X_pca] and
loadings in adata.varm[PCs]</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>adata</strong>  Adata with log_scale transformed mC fraction and selected features</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p></p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>PC information stored in adata.obsm and varm</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="ALLCools.clustering.ReproduciblePCA.mcds_to_adata_with_pc">
<span class="sig-name descname"><span class="pre">mcds_to_adata_with_pc</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">self</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mcds</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#ALLCools.clustering.ReproduciblePCA.mcds_to_adata_with_pc" title="Permalink to this definition"></a></dt>
<dd><p>From raw count MCDS to adata object with PCs using fitted scaler and PC loadings.
Steps include select features, calculate per-cell normalized mC fractions,
log_scale transform the data with fitted scaler, and finally add PC matrix.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>mcds</strong>  Raw count MCDS</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>anndata object with per-cell normalized, log scale transformed matrix in .X and PCs in
adata.obsm[X_pca] and PC loadings in adata.varm[PCs]. The scale and PC are done with fitted model.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>adata</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="ALLCools.clustering.ReproduciblePCA.dump">
<span class="sig-name descname"><span class="pre">dump</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">self</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">path</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#ALLCools.clustering.ReproduciblePCA.dump" title="Permalink to this definition"></a></dt>
<dd><p>Save the ReproduciblePCA to path</p>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="ALLCools.clustering.Dendrogram">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">Dendrogram</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">nboot</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1000</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">method_dist</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'correlation'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">method_hclust</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'average'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n_jobs</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">-</span> <span class="pre">1</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/ALLCools/clustering/pvclust.html#Dendrogram"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#ALLCools.clustering.Dendrogram" title="Permalink to this definition"></a></dt>
<dd><dl class="py method">
<dt class="sig sig-object py" id="ALLCools.clustering.Dendrogram.fit">
<span class="sig-name descname"><span class="pre">fit</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">self</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">data</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#ALLCools.clustering.Dendrogram.fit" title="Permalink to this definition"></a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>data</strong>  The data is in obs-by-var form, row is obs.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="ALLCools.clustering.Dendrogram.save">
<span class="sig-name descname"><span class="pre">save</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">self</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">output_path</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#ALLCools.clustering.Dendrogram.save" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="ALLCools.clustering.PairwiseDMG">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">PairwiseDMG</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">max_cell_per_group</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1000</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">top_n</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">10000</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">adj_p_cutoff</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.001</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">delta_rate_cutoff</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.3</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">auroc_cutoff</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.9</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">random_state</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n_jobs</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">verbose</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/ALLCools/clustering/dmg.html#PairwiseDMG"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#ALLCools.clustering.PairwiseDMG" title="Permalink to this definition"></a></dt>
<dd><dl class="py method">
<dt class="sig sig-object py" id="ALLCools.clustering.PairwiseDMG.fit_predict">
<span class="sig-name descname"><span class="pre">fit_predict</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">self</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">x</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">groups</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">var_dim</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">obs_dim</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'cell'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">outlier</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'Outlier'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cleanup</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">selected_pairs</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">List</span><span class="p"><span class="pre">[</span></span><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#tuple" title="(in Python v3.10)"><span class="pre">tuple</span></a><span class="p"><span class="pre">]</span></span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#ALLCools.clustering.PairwiseDMG.fit_predict" title="Permalink to this definition"></a></dt>
<dd><p>provide data and perform the pairwise DMG</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>x</strong>  2D cell-by-feature xarray.DataArray</p></li>
<li><p><strong>groups</strong>  cluster labels</p></li>
<li><p><strong>obs_dim</strong>  name of the cell dim</p></li>
<li><p><strong>var_dim</strong>  name of the feature dim</p></li>
<li><p><strong>outlier</strong>  name of the outlier group, if provided, will ignore this label</p></li>
<li><p><strong>cleanup</strong>  Whether to delete the group adata file</p></li>
<li><p><strong>selected_pairs</strong>  By default, pairwise DMG will calculate all possible pairs between all the groups, which might be very
time consuming if the group number is large. With this parameter, you may provide a list of cluster pairs</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="ALLCools.clustering.PairwiseDMG._save_cluster_adata">
<span class="sig-name descname"><span class="pre">_save_cluster_adata</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">self</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#ALLCools.clustering.PairwiseDMG._save_cluster_adata" title="Permalink to this definition"></a></dt>
<dd><p>Save each group into separate adata, this way reduce the memory during parallel</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="ALLCools.clustering.PairwiseDMG._pairwise_dmg">
<span class="sig-name descname"><span class="pre">_pairwise_dmg</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">self</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#ALLCools.clustering.PairwiseDMG._pairwise_dmg" title="Permalink to this definition"></a></dt>
<dd><p>pairwise DMG runner, result save to self.dmg_table</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="ALLCools.clustering.PairwiseDMG._cleanup">
<span class="sig-name descname"><span class="pre">_cleanup</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">self</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#ALLCools.clustering.PairwiseDMG._cleanup" title="Permalink to this definition"></a></dt>
<dd><p>Delete group adata files</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="ALLCools.clustering.PairwiseDMG.aggregate_pairwise_dmg">
<span class="sig-name descname"><span class="pre">aggregate_pairwise_dmg</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">self</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">adata</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">groupby</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">obsm</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'X_pca'</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#ALLCools.clustering.PairwiseDMG.aggregate_pairwise_dmg" title="Permalink to this definition"></a></dt>
<dd><p>Aggregate pairwise DMG results for each cluster, rank DMG for the cluster by the sum of
AUROC * cluster_pair_similarity
This way, the DMGs having large AUROC between similar clusters get more weights</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>adata</strong>  </p></li>
<li><p><strong>groupby</strong>  </p></li>
<li><p><strong>obsm</strong>  </p></li>
</ul>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="ALLCools.clustering.one_vs_rest_dmg">
<span class="sig-name descname"><span class="pre">one_vs_rest_dmg</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">cell_meta</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">group</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mcds</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mcds_paths</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">obs_dim</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'cell'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">var_dim</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'gene'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mc_type</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'CHN'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">top_n</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1000</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">adj_p_cutoff</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.01</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">fc_cutoff</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.8</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">auroc_cutoff</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.8</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">max_cluster_cells</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">2000</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">max_other_fold</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">5</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cpu</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">verbose</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/ALLCools/clustering/dmg.html#one_vs_rest_dmg"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#ALLCools.clustering.one_vs_rest_dmg" title="Permalink to this definition"></a></dt>
<dd><p>Calculating cluster marker genes using one-vs-rest strategy.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>cell_meta</strong>  cell metadata containing cluster labels</p></li>
<li><p><strong>group</strong>  the name of the cluster label column</p></li>
<li><p><strong>mcds</strong>  cell-by-gene MCDS object for calculating DMG. Provide either mcds_paths or mcds.</p></li>
<li><p><strong>mcds_paths</strong>  cell-by-gene MCDS paths for calculating DMG. Provide either mcds_paths or mcds.</p></li>
<li><p><strong>obs_dim</strong>  dimension name of the cells</p></li>
<li><p><strong>var_dim</strong>  dimension name of the features</p></li>
<li><p><strong>mc_type</strong>  value to select methylation type in the mc_type dimension</p></li>
<li><p><strong>top_n</strong>  report top N DMGs</p></li>
<li><p><strong>adj_p_cutoff</strong>  adjusted P value cutoff to report significant DMG</p></li>
<li><p><strong>fc_cutoff</strong>  mC fraction fold change cutoff to report significant DMG</p></li>
<li><p><strong>auroc_cutoff</strong>  AUROC cutoff to report significant DMG</p></li>
<li><p><strong>max_cluster_cells</strong>  The maximum number of cells from a group, downsample large group to this number</p></li>
<li><p><strong>max_other_fold</strong>  The fold of other cell numbers comparing</p></li>
<li><p><strong>cpu</strong>  number of cpus</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>pandas Dataframe of the one-vs-rest DMGs</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>dmg_table</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="ALLCools.clustering.cluster_enriched_features">
<span class="sig-name descname"><span class="pre">cluster_enriched_features</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">adata</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://anndata.readthedocs.io/en/stable/generated/anndata.AnnData.html#anndata.AnnData" title="(in anndata v0.8.0)"><span class="pre">anndata.AnnData</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">cluster_col</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.10)"><span class="pre">str</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">top_n</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">200</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">alpha</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.05</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">stat_plot</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">method</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'mc'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/ALLCools/clustering/feature_selection/feature_enrichment.html#cluster_enriched_features"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#ALLCools.clustering.cluster_enriched_features" title="Permalink to this definition"></a></dt>
<dd><p>Calculate top Cluster Enriched Features (CEF) from per-cell normalized dataset.
An post-clustering feature selection step adapted from <span id="id4">[<a class="reference internal" href="../../../project_info/citation.html#id10">Zeisel <em>et al.</em>, 2018</a>, <a class="reference internal" href="../../../project_info/citation.html#id9">LaManno <em>et al.</em>, 2021</a>]</span>
and their great [cytograph2](<a class="reference external" href="https://github.com/linnarsson-lab/cytograph2">https://github.com/linnarsson-lab/cytograph2</a>) package.
For details about CEF calculation, read the methods of <span id="id5">[<a class="reference internal" href="../../../project_info/citation.html#id10">Zeisel <em>et al.</em>, 2018</a>]</span>. Note that
in original paper, they look for cluster-specific highly expressed genes as CEFs;
for methylation, we are looking for hypo-methylation as CEFs, so the score and test is reversed.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>adata</strong>  adata containing per-cell normalized values.
For methylation fraction, the value need to be 1-centered
(1 means cells average methylation), like those produced by
<a class="reference internal" href="../mcds/mcds/index.html#ALLCools.mcds.mcds.MCDS.add_mc_frac" title="ALLCools.mcds.mcds.MCDS.add_mc_frac"><code class="xref py py-func docutils literal notranslate"><span class="pre">ALLCools.mcds.mcds.MCDS.add_mc_frac()</span></code></a> with <cite>normalize_per_cell=True</cite>.
For RNA and ATAC, you can use per cell normalized counts.
Do not log transform the data before running this function</p></li>
<li><p><strong>cluster_col</strong>  The name of categorical variable in adata.obs</p></li>
<li><p><strong>top_n</strong>  Select top N CEFs for each cluster</p></li>
<li><p><strong>alpha</strong>  FDR corrected q-value cutoff</p></li>
<li><p><strong>stat_plot</strong>  Whether making some summary plots for the CEF calculation</p></li>
<li><p><strong>method</strong>  mc for methylation CEF (look for hypo-methylation),
rna and atac for the RNA and ATAC or any count based data
(use the original cytograph algorithm, look for higher value)</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p><ul class="simple">
<li><p><em>Modify adata inplace, adding a dictionary in adata.uns called f{cluster_col}_feature_enrichment</em></p></li>
<li><p><em>The dictionary contains qvals (np.ndarray cluster-by-feature enrichment score q-value) and</em></p></li>
<li><p><em>cluster_order (cluster order of the qvals)</em></p></li>
</ul>
</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="ALLCools.clustering.filter_regions">
<span class="sig-name descname"><span class="pre">filter_regions</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">adata</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">hypo_percent</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.5</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/ALLCools/clustering/mcad.html#filter_regions"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#ALLCools.clustering.filter_regions" title="Permalink to this definition"></a></dt>
<dd><p>Filter regions based on % of cells having non-zero scores.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>adata</strong>  </p></li>
<li><p><strong>hypo_percent</strong>  min % of cells that are non-zero in this region.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="ALLCools.clustering.lsi">
<span class="sig-name descname"><span class="pre">lsi</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">adata</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">scale_factor</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">100000</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n_components</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">100</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">algorithm</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'arpack'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">obsm</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'X_pca'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">random_state</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">fit_size</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/ALLCools/clustering/mcad.html#lsi"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#ALLCools.clustering.lsi" title="Permalink to this definition"></a></dt>
<dd><p>Run TF-IDF on the binarized adata.X, followed by TruncatedSVD and then scale the components by <a href="#id6"><span class="problematic" id="id7">svd.singular_values_</span></a></p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>adata</strong>  </p></li>
<li><p><strong>scale_factor</strong>  </p></li>
<li><p><strong>n_components</strong>  </p></li>
<li><p><strong>algorithm</strong>  </p></li>
<li><p><strong>obsm</strong>  </p></li>
<li><p><strong>random_state</strong>  </p></li>
<li><p><strong>fit_size</strong>  Ratio or absolute int value, use to downsample when fitting the SVD to speed up run time.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="ALLCools.clustering.binarize_matrix">
<span class="sig-name descname"><span class="pre">binarize_matrix</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">adata</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cutoff</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.95</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/ALLCools/clustering/mcad.html#binarize_matrix"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#ALLCools.clustering.binarize_matrix" title="Permalink to this definition"></a></dt>
<dd><p>Binarize adata.X with adata.X &gt; cutoff</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>adata</strong>  AnnData object whose X is survival function matrix</p></li>
<li><p><strong>cutoff</strong>  Cutoff to binarize the survival function</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p></p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/constants.html#None" title="(in Python v3.10)">None</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="ALLCools.clustering.remove_black_list_region">
<span class="sig-name descname"><span class="pre">remove_black_list_region</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">adata</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">black_list_path</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">f</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.2</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/ALLCools/clustering/mcad.html#remove_black_list_region"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#ALLCools.clustering.remove_black_list_region" title="Permalink to this definition"></a></dt>
<dd><p>Remove regions overlap (bedtools intersect -f {f}) with regions in the black_list_path</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>adata</strong>  </p></li>
<li><p><strong>black_list_path</strong>  Path to the black list bed file</p></li>
<li><p><strong>f</strong>  Fraction of overlap when calling bedtools intersect</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p></p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/constants.html#None" title="(in Python v3.10)">None</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="ALLCools.clustering.ClusterMerge">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">ClusterMerge</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">merge_criterion</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">stop_criterion</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">stop_clusters</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">-</span> <span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n_cells</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">200</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">metric</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'euclidean'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">method</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'average'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">label_concat_str</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'::'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/ALLCools/clustering/ClusterMerging.html#ClusterMerge"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#ALLCools.clustering.ClusterMerge" title="Permalink to this definition"></a></dt>
<dd><dl class="py method">
<dt class="sig sig-object py" id="ALLCools.clustering.ClusterMerge._construct_tree">
<span class="sig-name descname"><span class="pre">_construct_tree</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">self</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#ALLCools.clustering.ClusterMerge._construct_tree" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="ALLCools.clustering.ClusterMerge._traverse">
<em class="property"><span class="pre">static</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">_traverse</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">node</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">call_back</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#ALLCools.clustering.ClusterMerge._traverse" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="ALLCools.clustering.ClusterMerge._merge_pair">
<span class="sig-name descname"><span class="pre">_merge_pair</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">self</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pair</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">concat_str</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'::'</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#ALLCools.clustering.ClusterMerge._merge_pair" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="ALLCools.clustering.ClusterMerge.fit_predict">
<span class="sig-name descname"><span class="pre">fit_predict</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">self</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">data_for_tree</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cell_to_type</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">gene_mcds</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#ALLCools.clustering.ClusterMerge.fit_predict" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="ALLCools.clustering.PairwiseDMGCriterion">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">PairwiseDMGCriterion</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">max_cell_per_group</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">100</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">top_n_markers</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">5</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">adj_p_cutoff</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.001</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">delta_rate_cutoff</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.3</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">auroc_cutoff</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.85</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">use_modality</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'either'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">random_state</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n_jobs</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">10</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">verbose</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/ALLCools/clustering/ClusterMerging.html#PairwiseDMGCriterion"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#ALLCools.clustering.PairwiseDMGCriterion" title="Permalink to this definition"></a></dt>
<dd><dl class="py method">
<dt class="sig sig-object py" id="ALLCools.clustering.PairwiseDMGCriterion.predict">
<span class="sig-name descname"><span class="pre">predict</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">self</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pair_labels</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pair_cells</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pair_cell_types</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pair_mcds</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">da_name</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'gene_da_frac'</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#ALLCools.clustering.PairwiseDMGCriterion.predict" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

</dd></dl>

</div>
</div>

    <script type="text/x-thebe-config">
    {
        requestKernel: true,
        binderOptions: {
            repo: "binder-examples/jupyter-stacks-datascience",
            ref: "master",
        },
        codeMirrorConfig: {
            theme: "abcdef",
            mode: "python"
        },
        kernelOptions: {
            kernelName: "python3",
            path: "./api/ALLCools/clustering"
        },
        predefinedOutput: true
    }
    </script>
    <script>kernelName = 'python3'</script>

              </div>
              
            
                <!-- Previous / next buttons -->
<div class='prev-next-area'> 
    <a class='left-prev' id="prev-link" href="../abc/abc_score/index.html" title="previous page">
        <i class="fas fa-angle-left"></i>
        <div class="prev-next-info">
            <p class="prev-next-subtitle">previous</p>
            <p class="prev-next-title"><code class="xref py py-mod docutils literal notranslate"><span class="pre">ALLCools.abc.abc_score</span></code></p>
        </div>
    </a>
    <a class='right-next' id="next-link" href="doublets/index.html" title="next page">
    <div class="prev-next-info">
        <p class="prev-next-subtitle">next</p>
        <p class="prev-next-title"><code class="xref py py-mod docutils literal notranslate"><span class="pre">ALLCools.clustering.doublets</span></code></p>
    </div>
    <i class="fas fa-angle-right"></i>
    </a>
</div>
            
        </div>
    </div>
    <footer class="footer">
  <p>
    
      By Hanqing Liu<br/>
    
        &copy; Copyright 2019 - 2022.<br/>
  </p>
</footer>
</main>


      </div>
    </div>
  
  <script src="../../../_static/js/index.be7d3bbb2ef33a8344ce.js"></script>

  </body>
</html>