
<!DOCTYPE html>

<html>
<head>
<meta charset="utf-8"/>
<meta content="width=device-width, initial-scale=1.0" name="viewport"/>
<title>Cell Basic Filtering</title>
<link href="../../../_static/css/theme.css" rel="stylesheet"/>
<link href="../../../_static/css/index.ff1ffe594081f20da1ef19478df9384b.css" rel="stylesheet"/>
<link href="../../../_static/vendor/fontawesome/5.13.0/css/all.min.css" rel="stylesheet"/>
<link as="font" crossorigin="" href="../../../_static/vendor/fontawesome/5.13.0/webfonts/fa-solid-900.woff2" rel="preload" type="font/woff2"/>
<link as="font" crossorigin="" href="../../../_static/vendor/fontawesome/5.13.0/webfonts/fa-brands-400.woff2" rel="preload" type="font/woff2"/>
<link href="../../../_static/pygments.css" rel="stylesheet" type="text/css">
<link href="../../../_static/sphinx-book-theme.css?digest=c3fdc42140077d1ad13ad2f1588a4309" rel="stylesheet" type="text/css">
<link href="../../../_static/togglebutton.css" rel="stylesheet" type="text/css">
<link href="../../../_static/copybutton.css" rel="stylesheet" type="text/css">
<link href="../../../_static/mystnb.css" rel="stylesheet" type="text/css">
<link href="../../../_static/sphinx-thebe.css" rel="stylesheet" type="text/css"/>
<link href="../../../_static/graphviz.css" rel="stylesheet" type="text/css"/>
<link href="../../../_static/sphinx-codeautolink.css" rel="stylesheet" type="text/css"/>
<link href="../../../_static/panels-main.c949a650a448cc0ae9fd3441c0e17fb0.css" rel="stylesheet" type="text/css"/>
<link href="../../../_static/panels-variables.06eb56fa6e07937060861dad626602ad.css" rel="stylesheet" type="text/css"/>
<link as="script" href="../../../_static/js/index.be7d3bbb2ef33a8344ce.js" rel="preload"/>
<script data-url_root="../../../" id="documentation_options" src="../../../_static/documentation_options.js"></script>
<script src="../../../_static/jquery.js"></script>
<script src="../../../_static/underscore.js"></script>
<script src="../../../_static/doctools.js"></script>
<script src="../../../_static/togglebutton.js"></script>
<script src="../../../_static/clipboard.min.js"></script>
<script src="../../../_static/copybutton.js"></script>
<script>var togglebuttonSelector = '.toggle, .admonition.dropdown, .tag_hide_input div.cell_input, .tag_hide-input div.cell_input, .tag_hide_output div.cell_output, .tag_hide-output div.cell_output, .tag_hide_cell.cell, .tag_hide-cell.cell';</script>
<script src="../../../_static/sphinx-book-theme.d59cb220de22ca1c485ebbdc042f0030.js"></script>
<script async="async" src="https://unpkg.com/thebe@0.5.1/lib/index.js"></script>
<script>
        const thebe_selector = ".thebe"
        const thebe_selector_input = "pre"
        const thebe_selector_output = ".output"
    </script>
<script async="async" src="../../../_static/sphinx-thebe.js"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
<link href="../../../_static/allcools_favicon.png" rel="shortcut icon">
<link href="../../../genindex.html" rel="index" title="Index"/>
<link href="../../../search.html" rel="search" title="Search"/>
<link href="02-DecompositionAndEmbedding.html" rel="next" title="Decomposition Using mCG-5Kb Bins"/>
<link href="intro_5kb.html" rel="prev" title="Step-by-step Clustering using mCG-5Kb Input"/>
<meta content="width=device-width, initial-scale=1" name="viewport"/>
<meta content="None" name="docsearch:language"/>
<!-- Google Analytics -->
<script async="" src="https://www.googletagmanager.com/gtag/js?id=G-4F8CRF0GK9"></script>
<script>
                    window.dataLayer = window.dataLayer || [];
                    function gtag(){ dataLayer.push(arguments); }
                    gtag('js', new Date());
                    gtag('config', 'G-4F8CRF0GK9');
                </script>
</link></link></link></link></link></link></head>
<body data-offset="80" data-spy="scroll" data-target="#bd-toc-nav">
<div class="container-fluid" id="banner"></div>
<div class="container-xl">
<div class="row">
<div class="col-12 col-md-3 bd-sidebar site-navigation show" id="site-navigation">
<div class="navbar-brand-box">
<a class="navbar-brand text-wrap" href="../../../index.html">
<!-- `logo` is deprecated in Sphinx 4.0, so remove this when we stop supporting 3 -->
<img alt="logo" class="logo" src="../../../_static/allcools_logo.png"/>
</a>
</div><form action="../../../search.html" class="bd-search d-flex align-items-center" method="get">
<i class="icon fas fa-search"></i>
<input aria-label="Search this book..." autocomplete="off" class="form-control" id="search-input" name="q" placeholder="Search this book..." type="search"/>
</form><nav aria-label="Main" class="bd-links" id="bd-docs-nav">
<div class="bd-toc-item active">
<ul class="nav bd-sidenav">
<li class="toctree-l1">
<a class="reference internal" href="../../../intro.html">
   ALLCools: ALL methyl-Cytosine tools
  </a>
</li>
</ul>
<p aria-level="2" class="caption" role="heading">
<span class="caption-text">
  GET STARTED
 </span>
</p>
<ul class="nav bd-sidenav">
<li class="toctree-l1">
<a class="reference internal" href="../../../start/installation.html">
   Installation
  </a>
</li>
<li class="toctree-l1">
<a class="reference internal" href="../../../start/input_files.html">
   Input Files
  </a>
</li>
<li class="toctree-l1">
<a class="reference internal" href="../../../start/analysis_steps.html">
   Analysis Steps
  </a>
</li>
<li class="toctree-l1">
<a class="reference internal" href="../../../start/interactive.html">
   Examples In This Document
  </a>
</li>
</ul>
<p aria-level="2" class="caption" role="heading">
<span class="caption-text">
  CELLULAR ANALYSIS
 </span>
</p>
<ul class="current nav bd-sidenav">
<li class="toctree-l1 has-children">
<a class="reference internal" href="../../basic/intro_basic_clustering.html">
   Basic Clustering Walk-through
  </a>
<input class="toctree-checkbox" id="toctree-checkbox-1" name="toctree-checkbox-1" type="checkbox">
<label for="toctree-checkbox-1">
<i class="fas fa-chevron-down">
</i>
</label>
<ul>
<li class="toctree-l2">
<a class="reference internal" href="../../basic/mch_mcg_100k_basic.html">
     Basic Cell Clustering Using 100Kb Bins
    </a>
</li>
<li class="toctree-l2">
<a class="reference internal" href="../../basic/mcg_5kb_basic.html">
     Basic Cell Clustering Using mCG-5Kb Bins
    </a>
</li>
</ul>
</input></li>
<li class="toctree-l1 current active has-children">
<a class="reference internal" href="../intro_step_by_step_clustering.html">
   Step-by-step Clustering Analysis
  </a>
<input checked="" class="toctree-checkbox" id="toctree-checkbox-2" name="toctree-checkbox-2" type="checkbox"/>
<label for="toctree-checkbox-2">
<i class="fas fa-chevron-down">
</i>
</label>
<ul class="current">
<li class="toctree-l2 has-children">
<a class="reference internal" href="../100kb/intro_100kb.html">
     Step-by-step Clustering Using 100Kb bins
    </a>
<input class="toctree-checkbox" id="toctree-checkbox-3" name="toctree-checkbox-3" type="checkbox"/>
<label for="toctree-checkbox-3">
<i class="fas fa-chevron-down">
</i>
</label>
<ul>
<li class="toctree-l3">
<a class="reference internal" href="../100kb/01-CellBasicFiltering.html">
       Cell Basic Filtering
      </a>
</li>
<li class="toctree-l3">
<a class="reference internal" href="../100kb/02-BasicFeatureFiltering.html">
       Feature Basic Filtering
      </a>
</li>
<li class="toctree-l3">
<a class="reference internal" href="../100kb/03-HighlyVariableFeatureSelection.html">
       Calculate Highly Variable Features And Get mC Fraction AnnData
      </a>
</li>
<li class="toctree-l3">
<a class="reference internal" href="../100kb/04a-PreclusteringAndClusterEnrichedFeatures-mCH.html">
       Preclustering and Cluster Enriched Features
      </a>
</li>
<li class="toctree-l3">
<a class="reference internal" href="../100kb/04b-PreclusteringAndClusterEnrichedFeatures-mCG.html">
       Preclustering and Cluster Enriched Features
      </a>
</li>
<li class="toctree-l3">
<a class="reference internal" href="../100kb/05-DecompositionAndEmbedding.html">
       Decomposition And Embedding
      </a>
</li>
<li class="toctree-l3">
<a class="reference internal" href="../100kb/06-Clustering.html">
       Consensus Clustering
      </a>
</li>
</ul>
</li>
<li class="toctree-l2 current active has-children">
<a class="reference internal" href="intro_5kb.html">
     Step-by-step Clustering using mCG-5Kb Input
    </a>
<input checked="" class="toctree-checkbox" id="toctree-checkbox-4" name="toctree-checkbox-4" type="checkbox"/>
<label for="toctree-checkbox-4">
<i class="fas fa-chevron-down">
</i>
</label>
<ul class="current">
<li class="toctree-l3 current active">
<a class="current reference internal" href="#">
       Cell Basic Filtering
      </a>
</li>
<li class="toctree-l3">
<a class="reference internal" href="02-DecompositionAndEmbedding.html">
       Decomposition Using mCG-5Kb Bins
      </a>
</li>
<li class="toctree-l3">
<a class="reference internal" href="03-Clustering.html">
       Consensus Clustering
      </a>
</li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l1 has-children">
<a class="reference internal" href="../../dmg/intro_dmg.html">
   Differential Methylated Genes Analysis
  </a>
<input class="toctree-checkbox" id="toctree-checkbox-5" name="toctree-checkbox-5" type="checkbox"/>
<label for="toctree-checkbox-5">
<i class="fas fa-chevron-down">
</i>
</label>
<ul>
<li class="toctree-l2">
<a class="reference internal" href="../../dmg/01-AddGenemCFractions.html">
     Calculate Gene mC Fractions
    </a>
</li>
<li class="toctree-l2">
<a class="reference internal" href="../../dmg/02-DifferentiallyMethylatedGenes.html">
     Cluster Differentially Methylated Genes
    </a>
</li>
<li class="toctree-l2">
<a class="reference internal" href="../../dmg/03-PlotClusterDMGs.html">
     Plot Cluster DMGs
    </a>
</li>
<li class="toctree-l2">
<a class="reference internal" href="../../dmg/04-PairwiseDMG.html">
     Differential Methylated Genes - Pairwise
    </a>
</li>
</ul>
</li>
<li class="toctree-l1 has-children">
<a class="reference internal" href="../../integration/intro_integration.html">
   Data Integration
  </a>
<input class="toctree-checkbox" id="toctree-checkbox-6" name="toctree-checkbox-6" type="checkbox"/>
<label for="toctree-checkbox-6">
<i class="fas fa-chevron-down">
</i>
</label>
<ul>
<li class="toctree-l2">
<a class="reference internal" href="../../integration/multi_modalities/Neurons/01-preprocess_10x.html">
     Preprocess 10X snRNA-seq
    </a>
</li>
<li class="toctree-l2">
<a class="reference internal" href="../../integration/multi_modalities/Neurons/02-preprocess_SMART.html">
     Preprocess SMART-seq
    </a>
</li>
<li class="toctree-l2">
<a class="reference internal" href="../../integration/multi_modalities/Neurons/03-preprocess_snATAC.html">
     Preprocess snATAC-seq
    </a>
</li>
<li class="toctree-l2">
<a class="reference internal" href="../../integration/multi_modalities/Neurons/04-integration.html">
     Perform Cross-Modality Integration
    </a>
</li>
</ul>
</li>
<li class="toctree-l1 has-children">
<a class="reference internal" href="../../doublets/intro_doublets.html">
   Doublets Identification
  </a>
<input class="toctree-checkbox" id="toctree-checkbox-7" name="toctree-checkbox-7" type="checkbox"/>
<label for="toctree-checkbox-7">
<i class="fas fa-chevron-down">
</i>
</label>
<ul>
<li class="toctree-l2">
<a class="reference internal" href="../../doublets/doublet_score_chrom100k.html">
     Calculate Cell Doublet Score
    </a>
</li>
</ul>
</li>
</ul>
<p aria-level="2" class="caption" role="heading">
<span class="caption-text">
  GENOMIC ANALYSIS
 </span>
</p>
<ul class="nav bd-sidenav">
<li class="toctree-l1 has-children">
<a class="reference internal" href="../../../cluster_level/RegionDS/intro.html">
   Post-clustering Genomic Analysis
  </a>
<input class="toctree-checkbox" id="toctree-checkbox-8" name="toctree-checkbox-8" type="checkbox"/>
<label for="toctree-checkbox-8">
<i class="fas fa-chevron-down">
</i>
</label>
<ul>
<li class="toctree-l2">
<a class="reference internal" href="../../../cluster_level/RegionDS/01a.call_dmr.html">
     DMR Calling
    </a>
</li>
<li class="toctree-l2">
<a class="reference internal" href="../../../cluster_level/RegionDS/01b.other_option_to_init_region_ds.html">
     Alternative Way to Create RegionDS
    </a>
</li>
<li class="toctree-l2">
<a class="reference internal" href="../../../cluster_level/RegionDS/02.annotation.html">
     Annotate RegionDS
    </a>
</li>
<li class="toctree-l2">
<a class="reference internal" href="../../../cluster_level/RegionDS/03.filter_dmr.html">
     Filter DMR
    </a>
</li>
</ul>
</li>
<li class="toctree-l1 has-children">
<a class="reference internal" href="../../../cluster_level/RegionDS/intro_motif.html">
   Motif Scan and Analysis
  </a>
<input class="toctree-checkbox" id="toctree-checkbox-9" name="toctree-checkbox-9" type="checkbox"/>
<label for="toctree-checkbox-9">
<i class="fas fa-chevron-down">
</i>
</label>
<ul>
<li class="toctree-l2">
<a class="reference internal" href="../../../cluster_level/RegionDS/04.motif_scan.html">
     Motif Scan
    </a>
</li>
<li class="toctree-l2">
<a class="reference internal" href="../../../cluster_level/RegionDS/05.motif_enrichment.html">
     Motif Enrichment Analysis
    </a>
</li>
</ul>
</li>
<li class="toctree-l1 has-children">
<a class="reference internal" href="../../../cluster_level/Correlation/intro_corr.html">
   Region-Region Correlation Analysis
  </a>
<input class="toctree-checkbox" id="toctree-checkbox-10" name="toctree-checkbox-10" type="checkbox"/>
<label for="toctree-checkbox-10">
<i class="fas fa-chevron-down">
</i>
</label>
<ul>
<li class="toctree-l2">
<a class="reference internal" href="../../../cluster_level/Correlation/01.mc_cluster_profile.html">
     Prepare Cluster-by-Gene mC Profile
    </a>
</li>
<li class="toctree-l2">
<a class="reference internal" href="../../../cluster_level/Correlation/02.dmr_gene_corr.html">
     DMR-Gene Correlation
    </a>
</li>
</ul>
</li>
<li class="toctree-l1 has-children">
<a class="reference internal" href="../../../cluster_level/REPTILE/intro_reptile.html">
   REPTILE
  </a>
<input class="toctree-checkbox" id="toctree-checkbox-11" name="toctree-checkbox-11" type="checkbox"/>
<label for="toctree-checkbox-11">
<i class="fas fa-chevron-down">
</i>
</label>
<ul>
<li class="toctree-l2">
<a class="reference internal" href="../../../cluster_level/REPTILE/01.prepare_input.html">
     Prepare REPTILE
    </a>
</li>
<li class="toctree-l2">
<a class="reference internal" href="../../../cluster_level/REPTILE/02.run_reptile.html">
     Predict Enhancer with REPTILE Algorithm
    </a>
</li>
</ul>
</li>
</ul>
<p aria-level="2" class="caption" role="heading">
<span class="caption-text">
  COMMAND LINE TOOLS
 </span>
</p>
<ul class="nav bd-sidenav">
<li class="toctree-l1">
<a class="reference internal" href="../../../command_line/allcools.html">
   Entry Point
  </a>
</li>
<li class="toctree-l1">
<a class="reference internal" href="../../../command_line/allcools_allc.html">
<code class="docutils literal notranslate">
<span class="pre">
     allcools
    </span>
<span class="pre">
     allc
    </span>
</code>
</a>
</li>
<li class="toctree-l1">
<a class="reference internal" href="../../../command_line/allcools_standard.html">
<code class="docutils literal notranslate">
<span class="pre">
     allcools
    </span>
<span class="pre">
     standard
    </span>
</code>
</a>
</li>
<li class="toctree-l1">
<a class="reference internal" href="../../../command_line/allcools_profile.html">
<code class="docutils literal notranslate">
<span class="pre">
     allcools
    </span>
<span class="pre">
     profile
    </span>
</code>
</a>
</li>
<li class="toctree-l1">
<a class="reference internal" href="../../../command_line/allcools_tbi.html">
<code class="docutils literal notranslate">
<span class="pre">
     allcools
    </span>
<span class="pre">
     tbi
    </span>
</code>
</a>
</li>
<li class="toctree-l1">
<a class="reference internal" href="../../../command_line/allcools_merge.html">
<code class="docutils literal notranslate">
<span class="pre">
     allcools
    </span>
<span class="pre">
     merge
    </span>
</code>
</a>
</li>
<li class="toctree-l1">
<a class="reference internal" href="../../../command_line/allcools_extract.html">
<code class="docutils literal notranslate">
<span class="pre">
     allcools
    </span>
<span class="pre">
     extract
    </span>
</code>
</a>
</li>
<li class="toctree-l1">
<a class="reference internal" href="../../../command_line/allcools_bw.html">
<code class="docutils literal notranslate">
<span class="pre">
     allcools
    </span>
<span class="pre">
     bw
    </span>
</code>
</a>
</li>
<li class="toctree-l1">
<a class="reference internal" href="../../../command_line/allcools_region.html">
<code class="docutils literal notranslate">
<span class="pre">
     allcools
    </span>
<span class="pre">
     region
    </span>
</code>
</a>
</li>
<li class="toctree-l1">
<a class="reference internal" href="../../../command_line/allcools_mcds.html">
<code class="docutils literal notranslate">
<span class="pre">
     allcools
    </span>
<span class="pre">
     mcds
    </span>
</code>
</a>
</li>
<li class="toctree-l1">
<a class="reference internal" href="../../../command_line/allcools_mcad.html">
<code class="docutils literal notranslate">
<span class="pre">
     allcools
    </span>
<span class="pre">
     mcad
    </span>
</code>
</a>
</li>
</ul>
<p aria-level="2" class="caption" role="heading">
<span class="caption-text">
  VISUALIZATION
 </span>
</p>
<ul class="nav bd-sidenav">
<li class="toctree-l1">
<a class="reference internal" href="../../../visualization/intro.html">
   Visualization
  </a>
</li>
</ul>
<p aria-level="2" class="caption" role="heading">
<span class="caption-text">
  DISCUSSION
 </span>
</p>
<ul class="nav bd-sidenav">
<li class="toctree-l1">
<a class="reference internal" href="../../../discuss/intro.html">
   Discussion
  </a>
</li>
<li class="toctree-l1">
<a class="reference internal" href="../../../discuss/dmr_benchmark.html">
   DMR Caller Benchmark
  </a>
</li>
</ul>
<p aria-level="2" class="caption" role="heading">
<span class="caption-text">
  API
 </span>
</p>
<ul class="nav bd-sidenav">
<li class="toctree-l1 has-children">
<a class="reference internal" href="../../../api/ALLCools/index.html">
<code class="xref py py-mod docutils literal notranslate">
<span class="pre">
     ALLCools
    </span>
</code>
</a>
<input class="toctree-checkbox" id="toctree-checkbox-12" name="toctree-checkbox-12" type="checkbox"/>
<label for="toctree-checkbox-12">
<i class="fas fa-chevron-down">
</i>
</label>
<ul>
<li class="toctree-l2 has-children">
<a class="reference internal" href="../../../api/ALLCools/clustering/index.html">
<code class="xref py py-mod docutils literal notranslate">
<span class="pre">
       ALLCools.clustering
      </span>
</code>
</a>
<input class="toctree-checkbox" id="toctree-checkbox-13" name="toctree-checkbox-13" type="checkbox"/>
<label for="toctree-checkbox-13">
<i class="fas fa-chevron-down">
</i>
</label>
<ul>
<li class="toctree-l3 has-children">
<a class="reference internal" href="../../../api/ALLCools/clustering/doublets/index.html">
<code class="xref py py-mod docutils literal notranslate">
<span class="pre">
         ALLCools.clustering.doublets
        </span>
</code>
</a>
<input class="toctree-checkbox" id="toctree-checkbox-14" name="toctree-checkbox-14" type="checkbox"/>
<label for="toctree-checkbox-14">
<i class="fas fa-chevron-down">
</i>
</label>
<ul>
<li class="toctree-l4">
<a class="reference internal" href="../../../api/ALLCools/clustering/doublets/coverage_doublets/index.html">
<code class="xref py py-mod docutils literal notranslate">
<span class="pre">
           ALLCools.clustering.doublets.coverage_doublets
          </span>
</code>
</a>
</li>
<li class="toctree-l4">
<a class="reference internal" href="../../../api/ALLCools/clustering/doublets/scrublet/index.html">
<code class="xref py py-mod docutils literal notranslate">
<span class="pre">
           ALLCools.clustering.doublets.scrublet
          </span>
</code>
</a>
</li>
</ul>
</li>
<li class="toctree-l3 has-children">
<a class="reference internal" href="../../../api/ALLCools/clustering/feature_selection/index.html">
<code class="xref py py-mod docutils literal notranslate">
<span class="pre">
         ALLCools.clustering.feature_selection
        </span>
</code>
</a>
<input class="toctree-checkbox" id="toctree-checkbox-15" name="toctree-checkbox-15" type="checkbox"/>
<label for="toctree-checkbox-15">
<i class="fas fa-chevron-down">
</i>
</label>
<ul>
<li class="toctree-l4">
<a class="reference internal" href="../../../api/ALLCools/clustering/feature_selection/feature_enrichment/index.html">
<code class="xref py py-mod docutils literal notranslate">
<span class="pre">
           ALLCools.clustering.feature_selection.feature_enrichment
          </span>
</code>
</a>
</li>
</ul>
</li>
<li class="toctree-l3">
<a class="reference internal" href="../../../api/ALLCools/clustering/ClusterMerging/index.html">
<code class="xref py py-mod docutils literal notranslate">
<span class="pre">
         ALLCools.clustering.ClusterMerging
        </span>
</code>
</a>
</li>
<li class="toctree-l3">
<a class="reference internal" href="../../../api/ALLCools/clustering/ConsensusClustering/index.html">
<code class="xref py py-mod docutils literal notranslate">
<span class="pre">
         ALLCools.clustering.ConsensusClustering
        </span>
</code>
</a>
</li>
<li class="toctree-l3">
<a class="reference internal" href="../../../api/ALLCools/clustering/art_of_tsne/index.html">
<code class="xref py py-mod docutils literal notranslate">
<span class="pre">
         ALLCools.clustering.art_of_tsne
        </span>
</code>
</a>
</li>
<li class="toctree-l3">
<a class="reference internal" href="../../../api/ALLCools/clustering/balanced_pca/index.html">
<code class="xref py py-mod docutils literal notranslate">
<span class="pre">
         ALLCools.clustering.balanced_pca
        </span>
</code>
</a>
</li>
<li class="toctree-l3">
<a class="reference internal" href="../../../api/ALLCools/clustering/dmg/index.html">
<code class="xref py py-mod docutils literal notranslate">
<span class="pre">
         ALLCools.clustering.dmg
        </span>
</code>
</a>
</li>
<li class="toctree-l3">
<a class="reference internal" href="../../../api/ALLCools/clustering/incremental_pca/index.html">
<code class="xref py py-mod docutils literal notranslate">
<span class="pre">
         ALLCools.clustering.incremental_pca
        </span>
</code>
</a>
</li>
<li class="toctree-l3">
<a class="reference internal" href="../../../api/ALLCools/clustering/mcad/index.html">
<code class="xref py py-mod docutils literal notranslate">
<span class="pre">
         ALLCools.clustering.mcad
        </span>
</code>
</a>
</li>
<li class="toctree-l3">
<a class="reference internal" href="../../../api/ALLCools/clustering/pvclust/index.html">
<code class="xref py py-mod docutils literal notranslate">
<span class="pre">
         ALLCools.clustering.pvclust
        </span>
</code>
</a>
</li>
</ul>
</li>
<li class="toctree-l2 has-children">
<a class="reference internal" href="../../../api/ALLCools/count_matrix/index.html">
<code class="xref py py-mod docutils literal notranslate">
<span class="pre">
       ALLCools.count_matrix
      </span>
</code>
</a>
<input class="toctree-checkbox" id="toctree-checkbox-16" name="toctree-checkbox-16" type="checkbox"/>
<label for="toctree-checkbox-16">
<i class="fas fa-chevron-down">
</i>
</label>
<ul>
<li class="toctree-l3">
<a class="reference internal" href="../../../api/ALLCools/count_matrix/atac/index.html">
<code class="xref py py-mod docutils literal notranslate">
<span class="pre">
         ALLCools.count_matrix.atac
        </span>
</code>
</a>
</li>
<li class="toctree-l3">
<a class="reference internal" href="../../../api/ALLCools/count_matrix/dataset/index.html">
<code class="xref py py-mod docutils literal notranslate">
<span class="pre">
         ALLCools.count_matrix.dataset
        </span>
</code>
</a>
</li>
<li class="toctree-l3">
<a class="reference internal" href="../../../api/ALLCools/count_matrix/h5ad/index.html">
<code class="xref py py-mod docutils literal notranslate">
<span class="pre">
         ALLCools.count_matrix.h5ad
        </span>
</code>
</a>
</li>
<li class="toctree-l3">
<a class="reference internal" href="../../../api/ALLCools/count_matrix/mcad/index.html">
<code class="xref py py-mod docutils literal notranslate">
<span class="pre">
         ALLCools.count_matrix.mcad
        </span>
</code>
</a>
</li>
<li class="toctree-l3">
<a class="reference internal" href="../../../api/ALLCools/count_matrix/mcds/index.html">
<code class="xref py py-mod docutils literal notranslate">
<span class="pre">
         ALLCools.count_matrix.mcds
        </span>
</code>
</a>
</li>
<li class="toctree-l3">
<a class="reference internal" href="../../../api/ALLCools/count_matrix/snap/index.html">
<code class="xref py py-mod docutils literal notranslate">
<span class="pre">
         ALLCools.count_matrix.snap
        </span>
</code>
</a>
</li>
<li class="toctree-l3">
<a class="reference internal" href="../../../api/ALLCools/count_matrix/zarr/index.html">
<code class="xref py py-mod docutils literal notranslate">
<span class="pre">
         ALLCools.count_matrix.zarr
        </span>
</code>
</a>
</li>
</ul>
</li>
<li class="toctree-l2 has-children">
<a class="reference internal" href="../../../api/ALLCools/dmr/index.html">
<code class="xref py py-mod docutils literal notranslate">
<span class="pre">
       ALLCools.dmr
      </span>
</code>
</a>
<input class="toctree-checkbox" id="toctree-checkbox-17" name="toctree-checkbox-17" type="checkbox"/>
<label for="toctree-checkbox-17">
<i class="fas fa-chevron-down">
</i>
</label>
<ul>
<li class="toctree-l3">
<a class="reference internal" href="../../../api/ALLCools/dmr/call_dmr/index.html">
<code class="xref py py-mod docutils literal notranslate">
<span class="pre">
         ALLCools.dmr.call_dmr
        </span>
</code>
</a>
</li>
<li class="toctree-l3">
<a class="reference internal" href="../../../api/ALLCools/dmr/call_dms/index.html">
<code class="xref py py-mod docutils literal notranslate">
<span class="pre">
         ALLCools.dmr.call_dms
        </span>
</code>
</a>
</li>
<li class="toctree-l3">
<a class="reference internal" href="../../../api/ALLCools/dmr/parse_methylpy/index.html">
<code class="xref py py-mod docutils literal notranslate">
<span class="pre">
         ALLCools.dmr.parse_methylpy
        </span>
</code>
</a>
</li>
<li class="toctree-l3">
<a class="reference internal" href="../../../api/ALLCools/dmr/rms_test/index.html">
<code class="xref py py-mod docutils literal notranslate">
<span class="pre">
         ALLCools.dmr.rms_test
        </span>
</code>
</a>
</li>
</ul>
</li>
<li class="toctree-l2 has-children">
<a class="reference internal" href="../../../api/ALLCools/integration/index.html">
<code class="xref py py-mod docutils literal notranslate">
<span class="pre">
       ALLCools.integration
      </span>
</code>
</a>
<input class="toctree-checkbox" id="toctree-checkbox-18" name="toctree-checkbox-18" type="checkbox"/>
<label for="toctree-checkbox-18">
<i class="fas fa-chevron-down">
</i>
</label>
<ul>
<li class="toctree-l3">
<a class="reference internal" href="../../../api/ALLCools/integration/cca/index.html">
<code class="xref py py-mod docutils literal notranslate">
<span class="pre">
         ALLCools.integration.cca
        </span>
</code>
</a>
</li>
<li class="toctree-l3">
<a class="reference internal" href="../../../api/ALLCools/integration/confusion/index.html">
<code class="xref py py-mod docutils literal notranslate">
<span class="pre">
         ALLCools.integration.confusion
        </span>
</code>
</a>
</li>
<li class="toctree-l3">
<a class="reference internal" href="../../../api/ALLCools/integration/harmony/index.html">
<code class="xref py py-mod docutils literal notranslate">
<span class="pre">
         ALLCools.integration.harmony
        </span>
</code>
</a>
</li>
</ul>
</li>
<li class="toctree-l2 has-children">
<a class="reference internal" href="../../../api/ALLCools/mcds/index.html">
<code class="xref py py-mod docutils literal notranslate">
<span class="pre">
       ALLCools.mcds
      </span>
</code>
</a>
<input class="toctree-checkbox" id="toctree-checkbox-19" name="toctree-checkbox-19" type="checkbox"/>
<label for="toctree-checkbox-19">
<i class="fas fa-chevron-down">
</i>
</label>
<ul>
<li class="toctree-l3">
<a class="reference internal" href="../../../api/ALLCools/mcds/correlation/index.html">
<code class="xref py py-mod docutils literal notranslate">
<span class="pre">
         ALLCools.mcds.correlation
        </span>
</code>
</a>
</li>
<li class="toctree-l3">
<a class="reference internal" href="../../../api/ALLCools/mcds/mcds/index.html">
<code class="xref py py-mod docutils literal notranslate">
<span class="pre">
         ALLCools.mcds.mcds
        </span>
</code>
</a>
</li>
<li class="toctree-l3">
<a class="reference internal" href="../../../api/ALLCools/mcds/region_ds/index.html">
<code class="xref py py-mod docutils literal notranslate">
<span class="pre">
         ALLCools.mcds.region_ds
        </span>
</code>
</a>
</li>
<li class="toctree-l3">
<a class="reference internal" href="../../../api/ALLCools/mcds/utilities/index.html">
<code class="xref py py-mod docutils literal notranslate">
<span class="pre">
         ALLCools.mcds.utilities
        </span>
</code>
</a>
</li>
</ul>
</li>
<li class="toctree-l2 has-children">
<a class="reference internal" href="../../../api/ALLCools/motif/index.html">
<code class="xref py py-mod docutils literal notranslate">
<span class="pre">
       ALLCools.motif
      </span>
</code>
</a>
<input class="toctree-checkbox" id="toctree-checkbox-20" name="toctree-checkbox-20" type="checkbox"/>
<label for="toctree-checkbox-20">
<i class="fas fa-chevron-down">
</i>
</label>
<ul>
<li class="toctree-l3">
<a class="reference internal" href="../../../api/ALLCools/motif/default_motif_set/index.html">
<code class="xref py py-mod docutils literal notranslate">
<span class="pre">
         ALLCools.motif.default_motif_set
        </span>
</code>
</a>
</li>
<li class="toctree-l3">
<a class="reference internal" href="../../../api/ALLCools/motif/motifs/index.html">
<code class="xref py py-mod docutils literal notranslate">
<span class="pre">
         ALLCools.motif.motifs
        </span>
</code>
</a>
</li>
<li class="toctree-l3">
<a class="reference internal" href="../../../api/ALLCools/motif/parse_meme/index.html">
<code class="xref py py-mod docutils literal notranslate">
<span class="pre">
         ALLCools.motif.parse_meme
        </span>
</code>
</a>
</li>
</ul>
</li>
<li class="toctree-l2 has-children">
<a class="reference internal" href="../../../api/ALLCools/plot/index.html">
<code class="xref py py-mod docutils literal notranslate">
<span class="pre">
       ALLCools.plot
      </span>
</code>
</a>
<input class="toctree-checkbox" id="toctree-checkbox-21" name="toctree-checkbox-21" type="checkbox"/>
<label for="toctree-checkbox-21">
<i class="fas fa-chevron-down">
</i>
</label>
<ul>
<li class="toctree-l3">
<a class="reference internal" href="../../../api/ALLCools/plot/categorical_scatter/index.html">
<code class="xref py py-mod docutils literal notranslate">
<span class="pre">
         ALLCools.plot.categorical_scatter
        </span>
</code>
</a>
</li>
<li class="toctree-l3">
<a class="reference internal" href="../../../api/ALLCools/plot/color/index.html">
<code class="xref py py-mod docutils literal notranslate">
<span class="pre">
         ALLCools.plot.color
        </span>
</code>
</a>
</li>
<li class="toctree-l3">
<a class="reference internal" href="../../../api/ALLCools/plot/continuous_scatter/index.html">
<code class="xref py py-mod docutils literal notranslate">
<span class="pre">
         ALLCools.plot.continuous_scatter
        </span>
</code>
</a>
</li>
<li class="toctree-l3">
<a class="reference internal" href="../../../api/ALLCools/plot/contour/index.html">
<code class="xref py py-mod docutils literal notranslate">
<span class="pre">
         ALLCools.plot.contour
        </span>
</code>
</a>
</li>
<li class="toctree-l3">
<a class="reference internal" href="../../../api/ALLCools/plot/decomposition/index.html">
<code class="xref py py-mod docutils literal notranslate">
<span class="pre">
         ALLCools.plot.decomposition
        </span>
</code>
</a>
</li>
<li class="toctree-l3">
<a class="reference internal" href="../../../api/ALLCools/plot/dendro/index.html">
<code class="xref py py-mod docutils literal notranslate">
<span class="pre">
         ALLCools.plot.dendro
        </span>
</code>
</a>
</li>
<li class="toctree-l3">
<a class="reference internal" href="../../../api/ALLCools/plot/interactive_scatter/index.html">
<code class="xref py py-mod docutils literal notranslate">
<span class="pre">
         ALLCools.plot.interactive_scatter
        </span>
</code>
</a>
</li>
<li class="toctree-l3">
<a class="reference internal" href="../../../api/ALLCools/plot/qc_plots/index.html">
<code class="xref py py-mod docutils literal notranslate">
<span class="pre">
         ALLCools.plot.qc_plots
        </span>
</code>
</a>
</li>
<li class="toctree-l3">
<a class="reference internal" href="../../../api/ALLCools/plot/size/index.html">
<code class="xref py py-mod docutils literal notranslate">
<span class="pre">
         ALLCools.plot.size
        </span>
</code>
</a>
</li>
<li class="toctree-l3">
<a class="reference internal" href="../../../api/ALLCools/plot/sunburst/index.html">
<code class="xref py py-mod docutils literal notranslate">
<span class="pre">
         ALLCools.plot.sunburst
        </span>
</code>
</a>
</li>
<li class="toctree-l3">
<a class="reference internal" href="../../../api/ALLCools/plot/text_anno_scatter/index.html">
<code class="xref py py-mod docutils literal notranslate">
<span class="pre">
         ALLCools.plot.text_anno_scatter
        </span>
</code>
</a>
</li>
<li class="toctree-l3">
<a class="reference internal" href="../../../api/ALLCools/plot/utilities/index.html">
<code class="xref py py-mod docutils literal notranslate">
<span class="pre">
         ALLCools.plot.utilities
        </span>
</code>
</a>
</li>
</ul>
</li>
<li class="toctree-l2 has-children">
<a class="reference internal" href="../../../api/ALLCools/pseudo_cell/index.html">
<code class="xref py py-mod docutils literal notranslate">
<span class="pre">
       ALLCools.pseudo_cell
      </span>
</code>
</a>
<input class="toctree-checkbox" id="toctree-checkbox-22" name="toctree-checkbox-22" type="checkbox"/>
<label for="toctree-checkbox-22">
<i class="fas fa-chevron-down">
</i>
</label>
<ul>
<li class="toctree-l3">
<a class="reference internal" href="../../../api/ALLCools/pseudo_cell/pseudo_cell_kmeans/index.html">
<code class="xref py py-mod docutils literal notranslate">
<span class="pre">
         ALLCools.pseudo_cell.pseudo_cell_kmeans
        </span>
</code>
</a>
</li>
<li class="toctree-l3">
<a class="reference internal" href="../../../api/ALLCools/pseudo_cell/pseudo_cell_knn/index.html">
<code class="xref py py-mod docutils literal notranslate">
<span class="pre">
         ALLCools.pseudo_cell.pseudo_cell_knn
        </span>
</code>
</a>
</li>
</ul>
</li>
<li class="toctree-l2 has-children">
<a class="reference internal" href="../../../api/ALLCools/reptile/index.html">
<code class="xref py py-mod docutils literal notranslate">
<span class="pre">
       ALLCools.reptile
      </span>
</code>
</a>
<input class="toctree-checkbox" id="toctree-checkbox-23" name="toctree-checkbox-23" type="checkbox"/>
<label for="toctree-checkbox-23">
<i class="fas fa-chevron-down">
</i>
</label>
<ul>
<li class="toctree-l3">
<a class="reference internal" href="../../../api/ALLCools/reptile/reptile/index.html">
<code class="xref py py-mod docutils literal notranslate">
<span class="pre">
         ALLCools.reptile.reptile
        </span>
</code>
</a>
</li>
</ul>
</li>
<li class="toctree-l2 has-children">
<a class="reference internal" href="../../../api/ALLCools/sandbox/index.html">
<code class="xref py py-mod docutils literal notranslate">
<span class="pre">
       ALLCools.sandbox
      </span>
</code>
</a>
<input class="toctree-checkbox" id="toctree-checkbox-24" name="toctree-checkbox-24" type="checkbox"/>
<label for="toctree-checkbox-24">
<i class="fas fa-chevron-down">
</i>
</label>
<ul>
<li class="toctree-l3 has-children">
<a class="reference internal" href="../../../api/ALLCools/sandbox/igv/index.html">
<code class="xref py py-mod docutils literal notranslate">
<span class="pre">
         ALLCools.sandbox.igv
        </span>
</code>
</a>
<input class="toctree-checkbox" id="toctree-checkbox-25" name="toctree-checkbox-25" type="checkbox"/>
<label for="toctree-checkbox-25">
<i class="fas fa-chevron-down">
</i>
</label>
<ul>
<li class="toctree-l4">
<a class="reference internal" href="../../../api/ALLCools/sandbox/igv/BaseClass/index.html">
<code class="xref py py-mod docutils literal notranslate">
<span class="pre">
           ALLCools.sandbox.igv.BaseClass
          </span>
</code>
</a>
</li>
<li class="toctree-l4">
<a class="reference internal" href="../../../api/ALLCools/sandbox/igv/Session/index.html">
<code class="xref py py-mod docutils literal notranslate">
<span class="pre">
           ALLCools.sandbox.igv.Session
          </span>
</code>
</a>
</li>
<li class="toctree-l4">
<a class="reference internal" href="../../../api/ALLCools/sandbox/igv/Track/index.html">
<code class="xref py py-mod docutils literal notranslate">
<span class="pre">
           ALLCools.sandbox.igv.Track
          </span>
</code>
</a>
</li>
<li class="toctree-l4">
<a class="reference internal" href="../../../api/ALLCools/sandbox/igv/defaults/index.html">
<code class="xref py py-mod docutils literal notranslate">
<span class="pre">
           ALLCools.sandbox.igv.defaults
          </span>
</code>
</a>
</li>
<li class="toctree-l4">
<a class="reference internal" href="../../../api/ALLCools/sandbox/igv/make_session/index.html">
<code class="xref py py-mod docutils literal notranslate">
<span class="pre">
           ALLCools.sandbox.igv.make_session
          </span>
</code>
</a>
</li>
</ul>
</li>
<li class="toctree-l3 has-children">
<a class="reference internal" href="../../../api/ALLCools/sandbox/motif/index.html">
<code class="xref py py-mod docutils literal notranslate">
<span class="pre">
         ALLCools.sandbox.motif
        </span>
</code>
</a>
<input class="toctree-checkbox" id="toctree-checkbox-26" name="toctree-checkbox-26" type="checkbox"/>
<label for="toctree-checkbox-26">
<i class="fas fa-chevron-down">
</i>
</label>
<ul>
<li class="toctree-l4">
<a class="reference internal" href="../../../api/ALLCools/sandbox/motif/ame/index.html">
<code class="xref py py-mod docutils literal notranslate">
<span class="pre">
           ALLCools.sandbox.motif.ame
          </span>
</code>
</a>
</li>
<li class="toctree-l4">
<a class="reference internal" href="../../../api/ALLCools/sandbox/motif/cmotif/index.html">
<code class="xref py py-mod docutils literal notranslate">
<span class="pre">
           ALLCools.sandbox.motif.cmotif
          </span>
</code>
</a>
</li>
<li class="toctree-l4">
<a class="reference internal" href="../../../api/ALLCools/sandbox/motif/fimo/index.html">
<code class="xref py py-mod docutils literal notranslate">
<span class="pre">
           ALLCools.sandbox.motif.fimo
          </span>
</code>
</a>
</li>
<li class="toctree-l4">
<a class="reference internal" href="../../../api/ALLCools/sandbox/motif/motif_scan/index.html">
<code class="xref py py-mod docutils literal notranslate">
<span class="pre">
           ALLCools.sandbox.motif.motif_scan
          </span>
</code>
</a>
</li>
<li class="toctree-l4">
<a class="reference internal" href="../../../api/ALLCools/sandbox/motif/utilities/index.html">
<code class="xref py py-mod docutils literal notranslate">
<span class="pre">
           ALLCools.sandbox.motif.utilities
          </span>
</code>
</a>
</li>
</ul>
</li>
<li class="toctree-l3">
<a class="reference internal" href="../../../api/ALLCools/sandbox/balanced_knn/index.html">
<code class="xref py py-mod docutils literal notranslate">
<span class="pre">
         ALLCools.sandbox.balanced_knn
        </span>
</code>
</a>
</li>
<li class="toctree-l3">
<a class="reference internal" href="../../../api/ALLCools/sandbox/scanorama/index.html">
<code class="xref py py-mod docutils literal notranslate">
<span class="pre">
         ALLCools.sandbox.scanorama
        </span>
</code>
</a>
</li>
<li class="toctree-l3">
<a class="reference internal" href="../../../api/ALLCools/sandbox/sccaf/index.html">
<code class="xref py py-mod docutils literal notranslate">
<span class="pre">
         ALLCools.sandbox.sccaf
        </span>
</code>
</a>
</li>
</ul>
</li>
<li class="toctree-l2 has-children">
<a class="reference internal" href="../../../api/ALLCools/schema/index.html">
<code class="xref py py-mod docutils literal notranslate">
<span class="pre">
       ALLCools.schema
      </span>
</code>
</a>
<input class="toctree-checkbox" id="toctree-checkbox-27" name="toctree-checkbox-27" type="checkbox"/>
<label for="toctree-checkbox-27">
<i class="fas fa-chevron-down">
</i>
</label>
<ul>
<li class="toctree-l3">
<a class="reference internal" href="../../../api/ALLCools/schema/mcds_schema/index.html">
<code class="xref py py-mod docutils literal notranslate">
<span class="pre">
         ALLCools.schema.mcds_schema
        </span>
</code>
</a>
</li>
</ul>
</li>
<li class="toctree-l2">
<a class="reference internal" href="../../../api/ALLCools/__main__/index.html">
<code class="xref py py-mod docutils literal notranslate">
<span class="pre">
       ALLCools.__main__
      </span>
</code>
</a>
</li>
<li class="toctree-l2">
<a class="reference internal" href="../../../api/ALLCools/_allc_to_bigwig/index.html">
<code class="xref py py-mod docutils literal notranslate">
<span class="pre">
       ALLCools._allc_to_bigwig
      </span>
</code>
</a>
</li>
<li class="toctree-l2">
<a class="reference internal" href="../../../api/ALLCools/_allc_to_bigwig_new/index.html">
<code class="xref py py-mod docutils literal notranslate">
<span class="pre">
       ALLCools._allc_to_bigwig_new
      </span>
</code>
</a>
</li>
<li class="toctree-l2">
<a class="reference internal" href="../../../api/ALLCools/_allc_to_region_count/index.html">
<code class="xref py py-mod docutils literal notranslate">
<span class="pre">
       ALLCools._allc_to_region_count
      </span>
</code>
</a>
</li>
<li class="toctree-l2">
<a class="reference internal" href="../../../api/ALLCools/_bam_to_allc/index.html">
<code class="xref py py-mod docutils literal notranslate">
<span class="pre">
       ALLCools._bam_to_allc
      </span>
</code>
</a>
</li>
<li class="toctree-l2">
<a class="reference internal" href="../../../api/ALLCools/_doc/index.html">
<code class="xref py py-mod docutils literal notranslate">
<span class="pre">
       ALLCools._doc
      </span>
</code>
</a>
</li>
<li class="toctree-l2">
<a class="reference internal" href="../../../api/ALLCools/_extract_allc/index.html">
<code class="xref py py-mod docutils literal notranslate">
<span class="pre">
       ALLCools._extract_allc
      </span>
</code>
</a>
</li>
<li class="toctree-l2">
<a class="reference internal" href="../../../api/ALLCools/_merge_allc/index.html">
<code class="xref py py-mod docutils literal notranslate">
<span class="pre">
       ALLCools._merge_allc
      </span>
</code>
</a>
</li>
<li class="toctree-l2">
<a class="reference internal" href="../../../api/ALLCools/_open/index.html">
<code class="xref py py-mod docutils literal notranslate">
<span class="pre">
       ALLCools._open
      </span>
</code>
</a>
</li>
<li class="toctree-l2">
<a class="reference internal" href="../../../api/ALLCools/_version/index.html">
<code class="xref py py-mod docutils literal notranslate">
<span class="pre">
       ALLCools._version
      </span>
</code>
</a>
</li>
<li class="toctree-l2">
<a class="reference internal" href="../../../api/ALLCools/api/index.html">
<code class="xref py py-mod docutils literal notranslate">
<span class="pre">
       ALLCools.api
      </span>
</code>
</a>
</li>
<li class="toctree-l2">
<a class="reference internal" href="../../../api/ALLCools/table_to_allc/index.html">
<code class="xref py py-mod docutils literal notranslate">
<span class="pre">
       ALLCools.table_to_allc
      </span>
</code>
</a>
</li>
<li class="toctree-l2">
<a class="reference internal" href="../../../api/ALLCools/utilities/index.html">
<code class="xref py py-mod docutils literal notranslate">
<span class="pre">
       ALLCools.utilities
      </span>
</code>
</a>
</li>
</ul>
</li>
</ul>
<p aria-level="2" class="caption" role="heading">
<span class="caption-text">
  PROJECT INFO
 </span>
</p>
<ul class="nav bd-sidenav">
<li class="toctree-l1">
<a class="reference internal" href="../../../project_info/citation.html">
   Citation And Reference
  </a>
</li>
<li class="toctree-l1">
<a class="reference external" href="https://github.com/lhqing/ALLCools/releases">
   Changelog
  </a>
</li>
</ul>
</div>
</nav> <!-- To handle the deprecated key -->
<div class="navbar_extra_footer">
  Powered by <a href="https://jupyterbook.org">Jupyter Book</a>
</div>
</div>
<main class="col py-md-3 pl-md-4 bd-content overflow-auto" role="main">
<div class="topbar container-xl fixed-top">
<div class="topbar-contents row">
<div class="col-12 col-md-3 bd-topbar-whitespace site-navigation show"></div>
<div class="col pl-md-4 topbar-main">
<button aria-controls="site-navigation" aria-expanded="true" aria-label="Toggle navigation" class="navbar-toggler ml-0" data-placement="left" data-target=".site-navigation" data-toggle="tooltip" id="navbar-toggler" title="Toggle navigation" type="button">
<i class="fas fa-bars"></i>
<i class="fas fa-arrow-left"></i>
<i class="fas fa-arrow-up"></i>
</button>
<div class="dropdown-buttons-trigger">
<button aria-label="Download this page" class="btn btn-secondary topbarbtn" id="dropdown-buttons-trigger"><i class="fas fa-download"></i></button>
<div class="dropdown-buttons">
<!-- ipynb file if we had a myst markdown file -->
<!-- Download raw file -->
<a class="dropdown-buttons" href="../../../_sources/cell_level/step_by_step/5kb/01-CellBasicFiltering.ipynb"><button class="btn btn-secondary topbarbtn" data-placement="left" data-toggle="tooltip" title="Download source file" type="button">.ipynb</button></a>
<!-- Download PDF via print -->
<button class="btn btn-secondary topbarbtn" data-placement="left" data-toggle="tooltip" id="download-print" onclick="printPdf(this)" title="Print to PDF" type="button">.pdf</button>
</div>
</div>
<!-- Source interaction buttons -->
<div class="dropdown-buttons-trigger">
<button aria-label="Connect with source repository" class="btn btn-secondary topbarbtn" id="dropdown-buttons-trigger"><i class="fab fa-github"></i></button>
<div class="dropdown-buttons sourcebuttons">
<a class="repository-button" href="https://github.com/lhqing/ALLCools"><button class="btn btn-secondary topbarbtn" data-placement="left" data-toggle="tooltip" title="Source repository" type="button"><i class="fab fa-github"></i>repository</button></a>
<a class="issues-button" href="https://github.com/lhqing/ALLCools/issues/new?title=Issue%20on%20page%20%2Fcell_level/step_by_step/5kb/01-CellBasicFiltering.html&amp;body=Your%20issue%20content%20here."><button class="btn btn-secondary topbarbtn" data-placement="left" data-toggle="tooltip" title="Open an issue" type="button"><i class="fas fa-lightbulb"></i>open issue</button></a>
<a class="edit-button" href="https://github.com/lhqing/ALLCools/edit/master/docs/allcools/cell_level/step_by_step/5kb/01-CellBasicFiltering.ipynb"><button class="btn btn-secondary topbarbtn" data-placement="left" data-toggle="tooltip" title="Edit this page" type="button"><i class="fas fa-pencil-alt"></i>suggest edit</button></a>
</div>
</div>
<!-- Full screen (wrap in <a> to have style consistency -->
<a class="full-screen-button"><button aria-label="Fullscreen mode" class="btn btn-secondary topbarbtn" data-placement="bottom" data-toggle="tooltip" onclick="toggleFullScreen()" title="Fullscreen mode" type="button"><i class="fas fa-expand"></i></button></a>
<!-- Launch buttons -->
</div>
<!-- Table of contents -->
<div class="d-none d-md-block col-md-2 bd-toc show noprint">
<div class="tocsection onthispage pt-5 pb-3">
<i class="fas fa-list"></i> Contents
            </div>
<nav aria-label="Page" id="bd-toc-nav">
<ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry">
<a class="reference internal nav-link" href="#content">
   Content
  </a>
</li>
<li class="toc-h2 nav-item toc-entry">
<a class="reference internal nav-link" href="#input">
   Input
  </a>
</li>
<li class="toc-h2 nav-item toc-entry">
<a class="reference internal nav-link" href="#output">
   Output
  </a>
</li>
<li class="toc-h2 nav-item toc-entry">
<a class="reference internal nav-link" href="#about-cell-mapping-metrics">
   About Cell Mapping Metrics
  </a>
</li>
<li class="toc-h2 nav-item toc-entry">
<a class="reference internal nav-link" href="#import">
   Import
  </a>
</li>
<li class="toc-h2 nav-item toc-entry">
<a class="reference internal nav-link" href="#parameters">
   Parameters
  </a>
</li>
<li class="toc-h2 nav-item toc-entry">
<a class="reference internal nav-link" href="#load-metadata">
   Load metadata
  </a>
</li>
<li class="toc-h2 nav-item toc-entry">
<a class="reference internal nav-link" href="#filter-by-key-mapping-metrics">
   Filter by key mapping metrics
  </a>
<ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry">
<a class="reference internal nav-link" href="#bismark-mapping-rate">
     Bismark Mapping Rate
    </a>
</li>
<li class="toc-h3 nav-item toc-entry">
<a class="reference internal nav-link" href="#final-reads">
     Final Reads
    </a>
</li>
<li class="toc-h3 nav-item toc-entry">
<a class="reference internal nav-link" href="#mccc-ccc">
     mCCC / CCC
    </a>
</li>
<li class="toc-h3 nav-item toc-entry">
<a class="reference internal nav-link" href="#mch-ch">
     mCH / CH
    </a>
</li>
<li class="toc-h3 nav-item toc-entry">
<a class="reference internal nav-link" href="#mcg">
     mCG
    </a>
</li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry">
<a class="reference internal nav-link" href="#combine-filters">
   Combine filters
  </a>
</li>
<li class="toc-h2 nav-item toc-entry">
<a class="reference internal nav-link" href="#sanity-test">
   Sanity Test
  </a>
</li>
<li class="toc-h2 nav-item toc-entry">
<a class="reference internal nav-link" href="#save-filtered-metadata">
   Save filtered metadata
  </a>
</li>
</ul>
</nav>
</div>
</div>
</div>
<div class="row" id="main-content">
<div class="col-12 col-md-9 pl-md-3 pr-md-0">
<!-- Table of contents that is only displayed when printing the page -->
<div class="onlyprint" id="jb-print-docs-body">
<h1>Cell Basic Filtering</h1>
<!-- Table of contents -->
<div id="print-main-content">
<div id="jb-print-toc">
<div>
<h2> Contents </h2>
</div>
<nav aria-label="Page">
<ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry">
<a class="reference internal nav-link" href="#content">
   Content
  </a>
</li>
<li class="toc-h2 nav-item toc-entry">
<a class="reference internal nav-link" href="#input">
   Input
  </a>
</li>
<li class="toc-h2 nav-item toc-entry">
<a class="reference internal nav-link" href="#output">
   Output
  </a>
</li>
<li class="toc-h2 nav-item toc-entry">
<a class="reference internal nav-link" href="#about-cell-mapping-metrics">
   About Cell Mapping Metrics
  </a>
</li>
<li class="toc-h2 nav-item toc-entry">
<a class="reference internal nav-link" href="#import">
   Import
  </a>
</li>
<li class="toc-h2 nav-item toc-entry">
<a class="reference internal nav-link" href="#parameters">
   Parameters
  </a>
</li>
<li class="toc-h2 nav-item toc-entry">
<a class="reference internal nav-link" href="#load-metadata">
   Load metadata
  </a>
</li>
<li class="toc-h2 nav-item toc-entry">
<a class="reference internal nav-link" href="#filter-by-key-mapping-metrics">
   Filter by key mapping metrics
  </a>
<ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry">
<a class="reference internal nav-link" href="#bismark-mapping-rate">
     Bismark Mapping Rate
    </a>
</li>
<li class="toc-h3 nav-item toc-entry">
<a class="reference internal nav-link" href="#final-reads">
     Final Reads
    </a>
</li>
<li class="toc-h3 nav-item toc-entry">
<a class="reference internal nav-link" href="#mccc-ccc">
     mCCC / CCC
    </a>
</li>
<li class="toc-h3 nav-item toc-entry">
<a class="reference internal nav-link" href="#mch-ch">
     mCH / CH
    </a>
</li>
<li class="toc-h3 nav-item toc-entry">
<a class="reference internal nav-link" href="#mcg">
     mCG
    </a>
</li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry">
<a class="reference internal nav-link" href="#combine-filters">
   Combine filters
  </a>
</li>
<li class="toc-h2 nav-item toc-entry">
<a class="reference internal nav-link" href="#sanity-test">
   Sanity Test
  </a>
</li>
<li class="toc-h2 nav-item toc-entry">
<a class="reference internal nav-link" href="#save-filtered-metadata">
   Save filtered metadata
  </a>
</li>
</ul>
</nav>
</div>
</div>
</div>
<div>
<div class="tex2jax_ignore mathjax_ignore section" id="cell-basic-filtering">
<h1>Cell Basic Filtering<a class="headerlink" href="#cell-basic-filtering" title="Permalink to this headline"></a></h1>
<div class="section" id="content">
<h2>Content<a class="headerlink" href="#content" title="Permalink to this headline"></a></h2>
<p>The purpose of this step is to get rid of cells having <strong>obvious</strong> issues, including the cells with low mapping rate (potentially contaminated), low final reads (empty well or lost a large amount of DNA during library prep.), or abnormal methylation fractions (failed in bisulfite conversion or contaminated).</p>
<p>We have two principles when applying these filters:</p>
<ol class="simple">
<li><p><strong>We set the cutoff based on the distribution of the whole dataset</strong>, where we assume the input dataset is largely successful (mostly &gt; 80-90% cells will pass QC). The cutoffs below are typical values we used in brain methylome analysis. Still, you may need to adjust cutoffs based on different data quality or sample source.</p></li>
<li><p><strong>The cutoff is intended to be loose.</strong> We do not use stringent cutoffs here to prevent potential data loss. Abormal cells may remain after basic filtering, and will likely be identified in the analysis based filtering (see later notebooks about doublet score and outliers in clustering)</p></li>
</ol>
</div>
<div class="section" id="input">
<h2>Input<a class="headerlink" href="#input" title="Permalink to this headline"></a></h2>
<ul class="simple">
<li><p>Cell metadata table that contains mapping metric for basic QC filtering.</p></li>
</ul>
</div>
<div class="section" id="output">
<h2>Output<a class="headerlink" href="#output" title="Permalink to this headline"></a></h2>
<ul class="simple">
<li><p>Filtered cell metadata table that contains only cells passed QC.</p></li>
</ul>
</div>
<div class="section" id="about-cell-mapping-metrics">
<h2>About Cell Mapping Metrics<a class="headerlink" href="#about-cell-mapping-metrics" title="Permalink to this headline"></a></h2>
<p>We usually gather many mapping metrics from each processing step, but not all of the metrics are relevant to the cell filtering. Below are the most relevant metrics that we use to filter cells. The name of these metrics might be different in your dataset. Change it according to the file you have.</p>
<p>If you use <a class="reference external" href="https://hq-1.gitbook.io/mc">YAP</a> to do mapping, you can find up-to-date mapping metrics documentation for <a class="reference external" href="https://hq-1.gitbook.io/mc/mapping-metrics/key-mapping-metrics">key metrics</a> and <a class="reference external" href="https://hq-1.gitbook.io/mc/mapping-metrics/all-mapping-metrics">all metrics</a> in YAP doc.</p>
</div>
<div class="section" id="import">
<h2>Import<a class="headerlink" href="#import" title="Permalink to this headline"></a></h2>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <a class="sphinx-codeautolink-a" href="https://pandas.pydata.org/pandas-docs/stable/index.html#module-pandas" title="pandas"><span class="nn">pandas</span></a> <span class="k">as</span> <span class="nn">pd</span>
<span class="kn">import</span> <span class="nn">seaborn</span> <span class="k">as</span> <span class="nn">sns</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">sns</span><span class="o">.</span><span class="n">set_context</span><span class="p">(</span><span class="n">context</span><span class="o">=</span><span class="s1">'notebook'</span><span class="p">,</span> <span class="n">font_scale</span><span class="o">=</span><span class="mf">1.3</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
</div>
<div class="section" id="parameters">
<h2>Parameters<a class="headerlink" href="#parameters" title="Permalink to this headline"></a></h2>
<div class="cell tag_parameters docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># change this to the path to your metadata</span>
<span class="n">metadata_path</span> <span class="o">=</span> <span class="s1">'../../../data/PIT/PIT.CellMetadata.csv.gz'</span>

<span class="c1"># Basic filtering parameters</span>
<span class="n">mapping_rate_cutoff</span> <span class="o">=</span> <span class="mf">0.5</span>
<span class="n">mapping_rate_col_name</span> <span class="o">=</span> <span class="s1">'MappingRate'</span>  <span class="c1"># Name may change</span>
<span class="n">final_reads_cutoff</span> <span class="o">=</span> <span class="mi">500000</span>
<span class="n">final_reads_col_name</span> <span class="o">=</span> <span class="s1">'FinalmCReads'</span>  <span class="c1"># Name may change</span>
<span class="n">mccc_cutoff</span> <span class="o">=</span> <span class="mf">0.03</span>
<span class="n">mccc_col_name</span> <span class="o">=</span> <span class="s1">'mCCCFrac'</span>  <span class="c1"># Name may change</span>
<span class="n">mch_cutoff</span> <span class="o">=</span> <span class="mf">0.2</span>
<span class="n">mch_col_name</span> <span class="o">=</span> <span class="s1">'mCHFrac'</span>  <span class="c1"># Name may change</span>
<span class="n">mcg_cutoff</span> <span class="o">=</span> <span class="mf">0.5</span>
<span class="n">mcg_col_name</span> <span class="o">=</span> <span class="s1">'mCGFrac'</span>  <span class="c1"># Name may change</span>
</pre></div>
</div>
</div>
</div>
</div>
<div class="section" id="load-metadata">
<h2>Load metadata<a class="headerlink" href="#load-metadata" title="Permalink to this headline"></a></h2>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">metadata</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">metadata_path</span><span class="p">,</span> <span class="n">index_col</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
<span class="n">total_cells</span> <span class="o">=</span> <span class="n">metadata</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s1">'Metadata of </span><span class="si">{</span><span class="n">total_cells</span><span class="si">}</span><span class="s1"> cells'</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Metadata of 2756 cells
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">metadata</span><span class="o">.</span><span class="n">head</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div>
<style scoped="">
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
<thead>
<tr style="text-align: right;">
<th></th>
<th>CellInputReadPairs</th>
<th>MappingRate</th>
<th>FinalmCReads</th>
<th>mCCCFrac</th>
<th>mCGFrac</th>
<th>mCHFrac</th>
<th>Plate</th>
<th>Col384</th>
<th>Row384</th>
<th>CellTypeAnno</th>
</tr>
<tr>
<th>index</th>
<th></th>
<th></th>
<th></th>
<th></th>
<th></th>
<th></th>
<th></th>
<th></th>
<th></th>
<th></th>
</tr>
</thead>
<tbody>
<tr>
<th>PIT_P1-PIT_P2-A1-AD001</th>
<td>1858622.0</td>
<td>0.685139</td>
<td>1612023.0</td>
<td>0.003644</td>
<td>0.679811</td>
<td>0.005782</td>
<td>PIT_P1</td>
<td>0</td>
<td>0</td>
<td>Outlier</td>
</tr>
<tr>
<th>PIT_P1-PIT_P2-A1-AD004</th>
<td>1599190.0</td>
<td>0.686342</td>
<td>1367004.0</td>
<td>0.004046</td>
<td>0.746012</td>
<td>0.008154</td>
<td>PIT_P1</td>
<td>1</td>
<td>0</td>
<td>Gonadotropes</td>
</tr>
<tr>
<th>PIT_P1-PIT_P2-A1-AD006</th>
<td>1932242.0</td>
<td>0.669654</td>
<td>1580990.0</td>
<td>0.003958</td>
<td>0.683584</td>
<td>0.005689</td>
<td>PIT_P1</td>
<td>1</td>
<td>1</td>
<td>Somatotropes</td>
</tr>
<tr>
<th>PIT_P1-PIT_P2-A1-AD007</th>
<td>1588505.0</td>
<td>0.664612</td>
<td>1292770.0</td>
<td>0.003622</td>
<td>0.735217</td>
<td>0.005460</td>
<td>PIT_P2</td>
<td>0</td>
<td>0</td>
<td>Rbpms+</td>
</tr>
<tr>
<th>PIT_P1-PIT_P2-A1-AD010</th>
<td>1738409.0</td>
<td>0.703835</td>
<td>1539676.0</td>
<td>0.003769</td>
<td>0.744640</td>
<td>0.006679</td>
<td>PIT_P2</td>
<td>1</td>
<td>0</td>
<td>Rbpms+</td>
</tr>
</tbody>
</table>
</div></div></div>
</div>
</div>
<div class="section" id="filter-by-key-mapping-metrics">
<h2>Filter by key mapping metrics<a class="headerlink" href="#filter-by-key-mapping-metrics" title="Permalink to this headline"></a></h2>
<div class="section" id="bismark-mapping-rate">
<h3>Bismark Mapping Rate<a class="headerlink" href="#bismark-mapping-rate" title="Permalink to this headline"></a></h3>
<ul class="simple">
<li><p>Low mapping rate indicates potential contamination.</p></li>
<li><p>Usually R1 mapping rate is 8-10% higher than R2 mapping rate for snmC based technologies, but they should be highly correlated. Here I am using the combined mapping rate. If you are using the R1MappingRate or R2MappingRate, change the cutoff accordingly.</p></li>
<li><p>Usually there is a peak on the left, which corresponding to the empty wells.</p></li>
</ul>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">_cutoff</span> <span class="o">=</span> <span class="n">mapping_rate_cutoff</span>
<span class="n">_col_name</span> <span class="o">=</span> <span class="n">mapping_rate_col_name</span>

<span class="c1"># plot distribution to make sure cutoff is appropriate</span>
<span class="n">g</span> <span class="o">=</span> <span class="n">sns</span><span class="o">.</span><span class="n">displot</span><span class="p">(</span><span class="n">metadata</span><span class="p">[</span><span class="n">_col_name</span><span class="p">],</span> <span class="n">binrange</span><span class="o">=</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">))</span>
<span class="n">g</span><span class="o">.</span><span class="n">ax</span><span class="o">.</span><span class="n">plot</span><span class="p">((</span><span class="n">_cutoff</span><span class="p">,</span> <span class="n">_cutoff</span><span class="p">),</span> <span class="n">g</span><span class="o">.</span><span class="n">ax</span><span class="o">.</span><span class="n">get_ylim</span><span class="p">(),</span> <span class="n">c</span><span class="o">=</span><span class="s1">'r'</span><span class="p">,</span> <span class="n">linestyle</span><span class="o">=</span><span class="s1">'--'</span><span class="p">)</span>

<span class="n">mapping_rate_judge</span> <span class="o">=</span> <span class="n">metadata</span><span class="p">[</span><span class="n">_col_name</span><span class="p">]</span> <span class="o">&gt;</span> <span class="n">_cutoff</span>
<span class="n">_passed_cells</span> <span class="o">=</span> <span class="n">mapping_rate_judge</span><span class="o">.</span><span class="n">sum</span><span class="p">()</span>
<span class="nb">print</span><span class="p">(</span>
    <span class="sa">f</span><span class="s1">'</span><span class="si">{</span><span class="n">_passed_cells</span><span class="si">}</span><span class="s1"> / </span><span class="si">{</span><span class="n">total_cells</span><span class="si">}</span><span class="s1"> cells (</span><span class="si">{</span><span class="n">_passed_cells</span> <span class="o">/</span> <span class="n">total_cells</span> <span class="o">*</span> <span class="mi">100</span><span class="si">:</span><span class="s1">.1f</span><span class="si">}</span><span class="s1">%) '</span>
    <span class="sa">f</span><span class="s1">'passed the </span><span class="si">{</span><span class="n">_col_name</span><span class="si">}</span><span class="s1"> cutoff </span><span class="si">{</span><span class="n">_cutoff</span><span class="si">}</span><span class="s1">.'</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>2756 / 2756 cells (100.0%) passed the MappingRate cutoff 0.5.
</pre></div>
</div>
<img alt="../../../_images/01-CellBasicFiltering_11_11.png" src="../../../_images/01-CellBasicFiltering_11_11.png">
</img></div>
</div>
</div>
<div class="section" id="final-reads">
<h3>Final Reads<a class="headerlink" href="#final-reads" title="Permalink to this headline"></a></h3>
<ul class="simple">
<li><p>The cutoff may change depending on how deep the library has been sequenced.</p></li>
<li><p>Usually there is a peak on the left, which corresponding to the empty wells.</p></li>
<li><p>There are also some cells having small number of reads, these wells may lost most of the DNA during library prep. Cells having too less reads can be hard to classify, since the methylome sequencing is an untargeted whole-genome sequencing.</p></li>
</ul>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">_cutoff</span> <span class="o">=</span> <span class="n">final_reads_cutoff</span>
<span class="n">_col_name</span> <span class="o">=</span> <span class="n">final_reads_col_name</span>

<span class="c1"># plot distribution to make sure cutoff is appropriate</span>
<span class="n">g</span> <span class="o">=</span> <span class="n">sns</span><span class="o">.</span><span class="n">displot</span><span class="p">(</span><span class="n">metadata</span><span class="p">[</span><span class="n">_col_name</span><span class="p">],</span> <span class="n">binrange</span><span class="o">=</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="mf">5e6</span><span class="p">))</span>
<span class="n">g</span><span class="o">.</span><span class="n">ax</span><span class="o">.</span><span class="n">plot</span><span class="p">((</span><span class="n">_cutoff</span><span class="p">,</span> <span class="n">_cutoff</span><span class="p">),</span> <span class="n">g</span><span class="o">.</span><span class="n">ax</span><span class="o">.</span><span class="n">get_ylim</span><span class="p">(),</span> <span class="n">c</span><span class="o">=</span><span class="s1">'r'</span><span class="p">,</span> <span class="n">linestyle</span><span class="o">=</span><span class="s1">'--'</span><span class="p">)</span>

<span class="n">final_reads_judge</span> <span class="o">=</span> <span class="n">metadata</span><span class="p">[</span><span class="n">_col_name</span><span class="p">]</span> <span class="o">&gt;</span> <span class="n">_cutoff</span>
<span class="n">_passed_cells</span> <span class="o">=</span> <span class="n">final_reads_judge</span><span class="o">.</span><span class="n">sum</span><span class="p">()</span>
<span class="nb">print</span><span class="p">(</span>
    <span class="sa">f</span><span class="s1">'</span><span class="si">{</span><span class="n">_passed_cells</span><span class="si">}</span><span class="s1"> / </span><span class="si">{</span><span class="n">total_cells</span><span class="si">}</span><span class="s1"> cells (</span><span class="si">{</span><span class="n">_passed_cells</span> <span class="o">/</span> <span class="n">total_cells</span> <span class="o">*</span> <span class="mi">100</span><span class="si">:</span><span class="s1">.1f</span><span class="si">}</span><span class="s1">%) '</span>
    <span class="sa">f</span><span class="s1">'passed the </span><span class="si">{</span><span class="n">_col_name</span><span class="si">}</span><span class="s1"> cutoff </span><span class="si">{</span><span class="n">_cutoff</span><span class="si">}</span><span class="s1">.'</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>2756 / 2756 cells (100.0%) passed the FinalmCReads cutoff 500000.
</pre></div>
</div>
<img alt="../../../_images/01-CellBasicFiltering_13_11.png" src="../../../_images/01-CellBasicFiltering_13_11.png"/>
</div>
</div>
</div>
<div class="section" id="mccc-ccc">
<h3>mCCC / CCC<a class="headerlink" href="#mccc-ccc" title="Permalink to this headline"></a></h3>
<ul class="simple">
<li><p>The mCCC fraction is used as the proxy of the upper bound of the non-conversion rate for cell-level QC. The methylation level at CCC sites is the lowest among all of the different 3 base-contexts (CNN), and, in fact, it is very close to the unmethylated lambda mC fraction.</p></li>
<li><p>However, mCCC fraction is correlated with mCH (especially in brain data), so you can see a similar shape of distribution of mCCC and mCH, but the range is different.</p></li>
</ul>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">_cutoff</span> <span class="o">=</span> <span class="n">mccc_cutoff</span>
<span class="n">_col_name</span> <span class="o">=</span> <span class="n">mccc_col_name</span>

<span class="c1"># plot distribution to make sure cutoff is appropriate</span>
<span class="n">g</span> <span class="o">=</span> <span class="n">sns</span><span class="o">.</span><span class="n">displot</span><span class="p">(</span><span class="n">metadata</span><span class="p">[</span><span class="n">_col_name</span><span class="p">],</span> <span class="n">binrange</span><span class="o">=</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="mf">0.05</span><span class="p">))</span>
<span class="n">g</span><span class="o">.</span><span class="n">ax</span><span class="o">.</span><span class="n">plot</span><span class="p">((</span><span class="n">_cutoff</span><span class="p">,</span> <span class="n">_cutoff</span><span class="p">),</span> <span class="n">g</span><span class="o">.</span><span class="n">ax</span><span class="o">.</span><span class="n">get_ylim</span><span class="p">(),</span> <span class="n">c</span><span class="o">=</span><span class="s1">'r'</span><span class="p">,</span> <span class="n">linestyle</span><span class="o">=</span><span class="s1">'--'</span><span class="p">)</span>

<span class="n">mccc_judge</span> <span class="o">=</span> <span class="n">metadata</span><span class="p">[</span><span class="n">_col_name</span><span class="p">]</span> <span class="o">&lt;</span> <span class="n">_cutoff</span>
<span class="n">_passed_cells</span> <span class="o">=</span> <span class="n">mccc_judge</span><span class="o">.</span><span class="n">sum</span><span class="p">()</span>
<span class="nb">print</span><span class="p">(</span>
    <span class="sa">f</span><span class="s1">'</span><span class="si">{</span><span class="n">_passed_cells</span><span class="si">}</span><span class="s1"> / </span><span class="si">{</span><span class="n">total_cells</span><span class="si">}</span><span class="s1"> cells (</span><span class="si">{</span><span class="n">_passed_cells</span> <span class="o">/</span> <span class="n">total_cells</span> <span class="o">*</span> <span class="mi">100</span><span class="si">:</span><span class="s1">.1f</span><span class="si">}</span><span class="s1">%) '</span>
    <span class="sa">f</span><span class="s1">'passed the </span><span class="si">{</span><span class="n">_col_name</span><span class="si">}</span><span class="s1"> cutoff </span><span class="si">{</span><span class="n">_cutoff</span><span class="si">}</span><span class="s1">.'</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>2756 / 2756 cells (100.0%) passed the mCCCFrac cutoff 0.03.
</pre></div>
</div>
<img alt="../../../_images/01-CellBasicFiltering_15_11.png" src="../../../_images/01-CellBasicFiltering_15_11.png"/>
</div>
</div>
</div>
<div class="section" id="mch-ch">
<h3>mCH / CH<a class="headerlink" href="#mch-ch" title="Permalink to this headline"></a></h3>
<ul class="simple">
<li><p>Usually failed cells (empty well or contaminated) tend to have abormal methylation level as well.</p></li>
</ul>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">_cutoff</span> <span class="o">=</span> <span class="n">mch_cutoff</span>
<span class="n">_col_name</span> <span class="o">=</span> <span class="n">mch_col_name</span>

<span class="c1"># plot distribution to make sure cutoff is appropriate</span>
<span class="n">g</span> <span class="o">=</span> <span class="n">sns</span><span class="o">.</span><span class="n">displot</span><span class="p">(</span><span class="n">metadata</span><span class="p">[</span><span class="n">_col_name</span><span class="p">],</span> <span class="n">binrange</span><span class="o">=</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="mf">0.3</span><span class="p">))</span>
<span class="n">g</span><span class="o">.</span><span class="n">ax</span><span class="o">.</span><span class="n">plot</span><span class="p">((</span><span class="n">_cutoff</span><span class="p">,</span> <span class="n">_cutoff</span><span class="p">),</span> <span class="n">g</span><span class="o">.</span><span class="n">ax</span><span class="o">.</span><span class="n">get_ylim</span><span class="p">(),</span> <span class="n">c</span><span class="o">=</span><span class="s1">'r'</span><span class="p">,</span> <span class="n">linestyle</span><span class="o">=</span><span class="s1">'--'</span><span class="p">)</span>

<span class="n">mch_judge</span> <span class="o">=</span> <span class="n">metadata</span><span class="p">[</span><span class="n">_col_name</span><span class="p">]</span> <span class="o">&lt;</span> <span class="n">_cutoff</span>
<span class="n">_passed_cells</span> <span class="o">=</span> <span class="n">mch_judge</span><span class="o">.</span><span class="n">sum</span><span class="p">()</span>
<span class="nb">print</span><span class="p">(</span>
    <span class="sa">f</span><span class="s1">'</span><span class="si">{</span><span class="n">_passed_cells</span><span class="si">}</span><span class="s1"> / </span><span class="si">{</span><span class="n">total_cells</span><span class="si">}</span><span class="s1"> cells (</span><span class="si">{</span><span class="n">_passed_cells</span> <span class="o">/</span> <span class="n">total_cells</span> <span class="o">*</span> <span class="mi">100</span><span class="si">:</span><span class="s1">.1f</span><span class="si">}</span><span class="s1">%) '</span>
    <span class="sa">f</span><span class="s1">'passed the </span><span class="si">{</span><span class="n">_col_name</span><span class="si">}</span><span class="s1"> cutoff </span><span class="si">{</span><span class="n">_cutoff</span><span class="si">}</span><span class="s1">.'</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>2756 / 2756 cells (100.0%) passed the mCHFrac cutoff 0.2.
</pre></div>
</div>
<img alt="../../../_images/01-CellBasicFiltering_17_11.png" src="../../../_images/01-CellBasicFiltering_17_11.png"/>
</div>
</div>
</div>
<div class="section" id="mcg">
<h3>mCG<a class="headerlink" href="#mcg" title="Permalink to this headline"></a></h3>
<ul class="simple">
<li><p>Usually failed cells (empty well or contaminated) tend to have abormal methylation level as well.</p></li>
</ul>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">_cutoff</span> <span class="o">=</span> <span class="n">mcg_cutoff</span>
<span class="n">_col_name</span> <span class="o">=</span> <span class="n">mcg_col_name</span>

<span class="c1"># plot distribution to make sure cutoff is appropriate</span>
<span class="n">g</span> <span class="o">=</span> <span class="n">sns</span><span class="o">.</span><span class="n">displot</span><span class="p">(</span><span class="n">metadata</span><span class="p">[</span><span class="n">_col_name</span><span class="p">],</span> <span class="n">binrange</span><span class="o">=</span><span class="p">(</span><span class="mf">0.3</span><span class="p">,</span> <span class="mi">1</span><span class="p">))</span>
<span class="n">g</span><span class="o">.</span><span class="n">ax</span><span class="o">.</span><span class="n">plot</span><span class="p">((</span><span class="n">_cutoff</span><span class="p">,</span> <span class="n">_cutoff</span><span class="p">),</span> <span class="n">g</span><span class="o">.</span><span class="n">ax</span><span class="o">.</span><span class="n">get_ylim</span><span class="p">(),</span> <span class="n">c</span><span class="o">=</span><span class="s1">'r'</span><span class="p">,</span> <span class="n">linestyle</span><span class="o">=</span><span class="s1">'--'</span><span class="p">)</span>

<span class="n">mcg_judge</span> <span class="o">=</span> <span class="n">metadata</span><span class="p">[</span><span class="n">_col_name</span><span class="p">]</span> <span class="o">&gt;</span> <span class="n">_cutoff</span>
<span class="n">_passed_cells</span> <span class="o">=</span> <span class="n">mcg_judge</span><span class="o">.</span><span class="n">sum</span><span class="p">()</span>
<span class="nb">print</span><span class="p">(</span>
    <span class="sa">f</span><span class="s1">'</span><span class="si">{</span><span class="n">_passed_cells</span><span class="si">}</span><span class="s1"> / </span><span class="si">{</span><span class="n">total_cells</span><span class="si">}</span><span class="s1"> cells (</span><span class="si">{</span><span class="n">_passed_cells</span> <span class="o">/</span> <span class="n">total_cells</span> <span class="o">*</span> <span class="mi">100</span><span class="si">:</span><span class="s1">.1f</span><span class="si">}</span><span class="s1">%) '</span>
    <span class="sa">f</span><span class="s1">'passed the </span><span class="si">{</span><span class="n">_col_name</span><span class="si">}</span><span class="s1"> cutoff </span><span class="si">{</span><span class="n">_cutoff</span><span class="si">}</span><span class="s1">.'</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>2756 / 2756 cells (100.0%) passed the mCGFrac cutoff 0.5.
</pre></div>
</div>
<img alt="../../../_images/01-CellBasicFiltering_19_11.png" src="../../../_images/01-CellBasicFiltering_19_11.png"/>
</div>
</div>
</div>
</div>
<div class="section" id="combine-filters">
<h2>Combine filters<a class="headerlink" href="#combine-filters" title="Permalink to this headline"></a></h2>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">judge</span> <span class="o">=</span> <span class="n">mapping_rate_judge</span> <span class="o">&amp;</span> <span class="n">final_reads_judge</span> <span class="o">&amp;</span> <span class="n">mccc_judge</span> <span class="o">&amp;</span> <span class="n">mch_judge</span> <span class="o">&amp;</span> <span class="n">mcg_judge</span>
<span class="n">passed_cells</span> <span class="o">=</span> <span class="n">judge</span><span class="o">.</span><span class="n">sum</span><span class="p">()</span>
<span class="nb">print</span><span class="p">(</span>
    <span class="sa">f</span><span class="s1">'</span><span class="si">{</span><span class="n">passed_cells</span><span class="si">}</span><span class="s1"> / </span><span class="si">{</span><span class="n">total_cells</span><span class="si">}</span><span class="s1"> cells (</span><span class="si">{</span><span class="n">passed_cells</span> <span class="o">/</span> <span class="n">total_cells</span> <span class="o">*</span> <span class="mi">100</span><span class="si">:</span><span class="s1">.1f</span><span class="si">}</span><span class="s1">%) '</span>
    <span class="sa">f</span><span class="s1">'passed all the filters.'</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>2756 / 2756 cells (100.0%) passed all the filters.
</pre></div>
</div>
</div>
</div>
</div>
<div class="section" id="sanity-test">
<h2>Sanity Test<a class="headerlink" href="#sanity-test" title="Permalink to this headline"></a></h2>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="k">try</span><span class="p">:</span>
    <span class="k">assert</span> <span class="p">(</span><span class="n">passed_cells</span> <span class="o">/</span> <span class="n">total_cells</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mf">0.6</span>
<span class="k">except</span> <span class="ne">AssertionError</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
    <span class="n">e</span><span class="o">.</span><span class="n">args</span> <span class="o">+=</span> <span class="p">(</span>
        <span class="s1">'A large amount of the cells do not pass filter, check your cutoffs or overall dataset quality.'</span><span class="p">,</span>
    <span class="p">)</span>
    <span class="k">raise</span> <span class="n">e</span>

<span class="k">try</span><span class="p">:</span>
    <span class="k">assert</span> <span class="n">passed_cells</span> <span class="o">&gt;</span> <span class="mi">0</span>
<span class="k">except</span> <span class="ne">AssertionError</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
    <span class="n">e</span><span class="o">.</span><span class="n">args</span> <span class="o">+=</span> <span class="p">(</span><span class="s1">'No cell remained after all the filters.'</span><span class="p">,</span> <span class="p">)</span>
    <span class="k">raise</span> <span class="n">e</span>

<span class="nb">print</span><span class="p">(</span><span class="s1">'Feel good'</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Feel good
</pre></div>
</div>
</div>
</div>
</div>
<div class="section" id="save-filtered-metadata">
<h2>Save filtered metadata<a class="headerlink" href="#save-filtered-metadata" title="Permalink to this headline"></a></h2>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">metadata_filtered</span> <span class="o">=</span> <span class="n">metadata</span><span class="p">[</span><span class="n">judge</span><span class="p">]</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>

<span class="n">metadata_filtered</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="s1">'CellMetadata.PassQC.csv.gz'</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
</div>
</div>
<script type="text/x-thebe-config">
    {
        requestKernel: true,
        binderOptions: {
            repo: "binder-examples/jupyter-stacks-datascience",
            ref: "master",
        },
        codeMirrorConfig: {
            theme: "abcdef",
            mode: "python"
        },
        kernelOptions: {
            kernelName: "python3",
            path: "./cell_level/step_by_step/5kb"
        },
        predefinedOutput: true
    }
    </script>
<script>kernelName = 'python3'</script>
</div>
<!-- Previous / next buttons -->
<div class="prev-next-area">
<a class="left-prev" href="intro_5kb.html" id="prev-link" title="previous page">
<i class="fas fa-angle-left"></i>
<div class="prev-next-info">
<p class="prev-next-subtitle">previous</p>
<p class="prev-next-title">Step-by-step Clustering using mCG-5Kb Input</p>
</div>
</a>
<a class="right-next" href="02-DecompositionAndEmbedding.html" id="next-link" title="next page">
<div class="prev-next-info">
<p class="prev-next-subtitle">next</p>
<p class="prev-next-title">Decomposition Using mCG-5Kb Bins</p>
</div>
<i class="fas fa-angle-right"></i>
</a>
</div>
</div>
</div>
<footer class="footer">
<p>
    
      By Hanqing Liu<br/>
    
         Copyright 2019 - 2022.<br/>
</p>
</footer>
</main>
</div>
</div>
<script src="../../../_static/js/index.be7d3bbb2ef33a8344ce.js"></script>
</body>
</html>