
<!DOCTYPE html>

<html>
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>Input Files</title>
    
  <!-- Loaded before other Sphinx assets -->
  <link href="../_static/styles/theme.css?digest=1999514e3f237ded88cf" rel="stylesheet">
<link href="../_static/styles/pydata-sphinx-theme.css?digest=1999514e3f237ded88cf" rel="stylesheet">

    
  <link rel="stylesheet"
    href="../_static/vendor/fontawesome/5.13.0/css/all.min.css">
  <link rel="preload" as="font" type="font/woff2" crossorigin
    href="../_static/vendor/fontawesome/5.13.0/webfonts/fa-solid-900.woff2">
  <link rel="preload" as="font" type="font/woff2" crossorigin
    href="../_static/vendor/fontawesome/5.13.0/webfonts/fa-brands-400.woff2">

    <link rel="stylesheet" type="text/css" href="../_static/pygments.css" />
    <link rel="stylesheet" href="../_static/styles/sphinx-book-theme.css?digest=5115cc725059bd94278eecd172e13a965bf8f5a9" type="text/css" />
    <link rel="stylesheet" type="text/css" href="../_static/togglebutton.css" />
    <link rel="stylesheet" type="text/css" href="../_static/copybutton.css" />
    <link rel="stylesheet" type="text/css" href="../_static/mystnb.css" />
    <link rel="stylesheet" type="text/css" href="../_static/sphinx-thebe.css" />
    <link rel="stylesheet" type="text/css" href="../_static/graphviz.css" />
    <link rel="stylesheet" type="text/css" href="../_static/sphinx-codeautolink.css" />
    <link rel="stylesheet" type="text/css" href="../_static/design-style.b7bb847fb20b106c3d81b95245e65545.min.css" />
    
  <!-- Pre-loaded scripts that we'll load fully later -->
  <link rel="preload" as="script" href="../_static/scripts/pydata-sphinx-theme.js?digest=1999514e3f237ded88cf">

    <script data-url_root="../" id="documentation_options" src="../_static/documentation_options.js"></script>
    <script src="../_static/jquery.js"></script>
    <script src="../_static/underscore.js"></script>
    <script src="../_static/doctools.js"></script>
    <script src="../_static/clipboard.min.js"></script>
    <script src="../_static/copybutton.js"></script>
    <script src="../_static/scripts/sphinx-book-theme.js?digest=9c920249402e914e316237a7dbc6769907cce411"></script>
    <script>let toggleHintShow = 'Click to show';</script>
    <script>let toggleHintHide = 'Click to hide';</script>
    <script>let toggleOpenOnPrint = 'true';</script>
    <script src="../_static/togglebutton.js"></script>
    <script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    <script>var togglebuttonSelector = '.toggle, .admonition.dropdown, .tag_hide_input div.cell_input, .tag_hide-input div.cell_input, .tag_hide_output div.cell_output, .tag_hide-output div.cell_output, .tag_hide_cell.cell, .tag_hide-cell.cell';</script>
    <script src="../_static/design-tabs.js"></script>
    <script>const THEBE_JS_URL = "https://unpkg.com/thebe@0.8.2/lib/index.js"
const thebe_selector = ".thebe,.cell"
const thebe_selector_input = "pre"
const thebe_selector_output = ".output, .cell_output"
</script>
    <script async="async" src="../_static/sphinx-thebe.js"></script>
    <link rel="shortcut icon" href="../_static/allcools_favicon.png"/>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="Analysis Steps" href="analysis_steps.html" />
    <link rel="prev" title="Installation" href="installation.html" />
    <meta name="viewport" content="width=device-width, initial-scale=1" />
    <meta name="docsearch:language" content="None">
    

    <!-- Google Analytics -->
    
<script async="" src="https://www.googletagmanager.com/gtag/js?id=G-4F8CRF0GK9"></script>
<script>
                    window.dataLayer = window.dataLayer || [];
                    function gtag(){ dataLayer.push(arguments); }
                    gtag('js', new Date());
                    gtag('config', 'G-4F8CRF0GK9');
                </script>

  </head>
  <body data-spy="scroll" data-target="#bd-toc-nav" data-offset="60">
<!-- Checkboxes to toggle the left sidebar -->
<input type="checkbox" class="sidebar-toggle" name="__navigation" id="__navigation" aria-label="Toggle navigation sidebar">
<label class="overlay overlay-navbar" for="__navigation">
    <div class="visually-hidden">Toggle navigation sidebar</div>
</label>
<!-- Checkboxes to toggle the in-page toc -->
<input type="checkbox" class="sidebar-toggle" name="__page-toc" id="__page-toc" aria-label="Toggle in-page Table of Contents">
<label class="overlay overlay-pagetoc" for="__page-toc">
    <div class="visually-hidden">Toggle in-page Table of Contents</div>
</label>
<!-- Headers at the top -->
<div class="announcement header-item noprint"></div>
<div class="header header-item noprint"></div>

    
    <div class="container-fluid" id="banner"></div>

    

    <div class="container-xl">
      <div class="row">
          
<!-- Sidebar -->
<div class="bd-sidebar noprint" id="site-navigation">
    <div class="bd-sidebar__content">
        <div class="bd-sidebar__top"><div class="navbar-brand-box">
    <a class="navbar-brand text-wrap" href="../index.html">
      
        <!-- `logo` is deprecated in Sphinx 4.0, so remove this when we stop supporting 3 -->
        
      
      
      <img src="../_static/allcools_logo.png" class="logo" alt="logo">
      
      
    </a>
</div><form class="bd-search d-flex align-items-center" action="../search.html" method="get">
  <i class="icon fas fa-search"></i>
  <input type="search" class="form-control" name="q" id="search-input" placeholder="Search this book..." aria-label="Search this book..." autocomplete="off" >
</form><nav class="bd-links" id="bd-docs-nav" aria-label="Main">
    <div class="bd-toc-item active">
        
        <ul class="nav bd-sidenav bd-sidenav__home-link">
            <li class="toctree-l1">
                <a class="reference internal" href="../intro.html">
                    ALLCools: ALL methyl-Cytosine tools
                </a>
            </li>
        </ul>
        <p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  GET STARTED
 </span>
</p>
<ul class="current nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="installation.html">
   Installation
  </a>
 </li>
 <li class="toctree-l1 current active">
  <a class="current reference internal" href="#">
   Input Files
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="analysis_steps.html">
   Analysis Steps
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="interactive.html">
   Examples In This Document
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  CELLULAR ANALYSIS
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../cell_level/basic/intro_basic_clustering.html">
   Basic Clustering Walk-through
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-1" name="toctree-checkbox-1" type="checkbox"/>
  <label for="toctree-checkbox-1">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../cell_level/basic/mch_mcg_100k_basic.html">
     Basic Cell Clustering Using 100Kb Bins
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../cell_level/basic/mcg_5kb_basic.html">
     Basic Cell Clustering Using mCG-5Kb Bins
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../cell_level/step_by_step/intro_step_by_step_clustering.html">
   Step-by-step Clustering Analysis
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-2" name="toctree-checkbox-2" type="checkbox"/>
  <label for="toctree-checkbox-2">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2 has-children">
    <a class="reference internal" href="../cell_level/step_by_step/100kb/intro_100kb.html">
     Step-by-step Clustering Using 100Kb bins
    </a>
    <input class="toctree-checkbox" id="toctree-checkbox-3" name="toctree-checkbox-3" type="checkbox"/>
    <label for="toctree-checkbox-3">
     <i class="fas fa-chevron-down">
     </i>
    </label>
    <ul>
     <li class="toctree-l3">
      <a class="reference internal" href="../cell_level/step_by_step/100kb/01-CellBasicFiltering.html">
       Cell Basic Filtering
      </a>
     </li>
     <li class="toctree-l3">
      <a class="reference internal" href="../cell_level/step_by_step/100kb/02-BasicFeatureFiltering.html">
       Feature Basic Filtering
      </a>
     </li>
     <li class="toctree-l3">
      <a class="reference internal" href="../cell_level/step_by_step/100kb/03-HighlyVariableFeatureSelection.html">
       Calculate Highly Variable Features And Get mC Fraction AnnData
      </a>
     </li>
     <li class="toctree-l3">
      <a class="reference internal" href="../cell_level/step_by_step/100kb/04a-PreclusteringAndClusterEnrichedFeatures-mCH.html">
       Preclustering and Cluster Enriched Features
      </a>
     </li>
     <li class="toctree-l3">
      <a class="reference internal" href="../cell_level/step_by_step/100kb/04b-PreclusteringAndClusterEnrichedFeatures-mCG.html">
       Preclustering and Cluster Enriched Features
      </a>
     </li>
     <li class="toctree-l3">
      <a class="reference internal" href="../cell_level/step_by_step/100kb/05-DecompositionAndEmbedding.html">
       Decomposition And Embedding
      </a>
     </li>
     <li class="toctree-l3">
      <a class="reference internal" href="../cell_level/step_by_step/100kb/06-Clustering.html">
       Consensus Clustering
      </a>
     </li>
    </ul>
   </li>
   <li class="toctree-l2 has-children">
    <a class="reference internal" href="../cell_level/step_by_step/5kb/intro_5kb.html">
     Step-by-step Clustering using mCG-5Kb Input
    </a>
    <input class="toctree-checkbox" id="toctree-checkbox-4" name="toctree-checkbox-4" type="checkbox"/>
    <label for="toctree-checkbox-4">
     <i class="fas fa-chevron-down">
     </i>
    </label>
    <ul>
     <li class="toctree-l3">
      <a class="reference internal" href="../cell_level/step_by_step/5kb/02-DecompositionAndEmbedding.html">
       Decomposition Using mCG-5Kb Bins
      </a>
     </li>
     <li class="toctree-l3">
      <a class="reference internal" href="../cell_level/step_by_step/5kb/03-Clustering.html">
       Consensus Clustering
      </a>
     </li>
    </ul>
   </li>
  </ul>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../cell_level/dmg/intro_dmg.html">
   Differential Methylated Genes Analysis
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-5" name="toctree-checkbox-5" type="checkbox"/>
  <label for="toctree-checkbox-5">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../cell_level/dmg/01-AddGenemCFractions.html">
     Calculate Gene mC Fractions
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../cell_level/dmg/02-DifferentiallyMethylatedGenes.html">
     Cluster Differentially Methylated Genes
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../cell_level/dmg/03-PlotClusterDMGs.html">
     Plot Cluster DMGs
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../cell_level/dmg/04-PairwiseDMG.html">
     Differential Methylated Genes - Pairwise
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../cell_level/integration/intro_integration.html">
   Data Integration
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-6" name="toctree-checkbox-6" type="checkbox"/>
  <label for="toctree-checkbox-6">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../cell_level/integration/multi_modalities/Neurons/01-preprocess_10x.html">
     Preprocess 10X snRNA-seq
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../cell_level/integration/multi_modalities/Neurons/02-preprocess_SMART.html">
     Preprocess SMART-seq
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../cell_level/integration/multi_modalities/Neurons/03-preprocess_snATAC.html">
     Preprocess snATAC-seq
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../cell_level/integration/multi_modalities/Neurons/04-integration.html">
     Perform Cross-Modality Integration
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../cell_level/doublets/intro_doublets.html">
   Doublets Identification
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-7" name="toctree-checkbox-7" type="checkbox"/>
  <label for="toctree-checkbox-7">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../cell_level/doublets/doublet_score_chrom100k.html">
     Calculate Cell Doublet Score
    </a>
   </li>
  </ul>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  GENOMIC ANALYSIS
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../cluster_level/intro.html">
   Post-clustering Genomic Analysis
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-8" name="toctree-checkbox-8" type="checkbox"/>
  <label for="toctree-checkbox-8">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../cluster_level/RegionDS/01a.call_dmr.html">
     DMR Calling
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../cluster_level/RegionDS/01b.other_option_to_init_region_ds.html">
     Alternative Way to Create RegionDS
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../cluster_level/RegionDS/02.annotation.html">
     Annotate RegionDS
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../cluster_level/RegionDS/03.filter_dmr.html">
     Filter DMR
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../cluster_level/RegionDS/intro_motif.html">
   Motif Scan and Analysis
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-9" name="toctree-checkbox-9" type="checkbox"/>
  <label for="toctree-checkbox-9">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../cluster_level/RegionDS/04.motif_scan.html">
     Motif Scan
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../cluster_level/RegionDS/05.motif_enrichment.html">
     Motif Enrichment Analysis
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../cluster_level/Correlation/intro_corr.html">
   Region-Region Correlation Analysis
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-10" name="toctree-checkbox-10" type="checkbox"/>
  <label for="toctree-checkbox-10">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../cluster_level/Correlation/01.mc_cluster_profile.html">
     Prepare Cluster-by-Gene mC Profile
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../cluster_level/Correlation/02.dmr_gene_corr.html">
     DMR-Gene Correlation
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../cluster_level/REPTILE/intro_reptile.html">
   REPTILE
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-11" name="toctree-checkbox-11" type="checkbox"/>
  <label for="toctree-checkbox-11">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../cluster_level/REPTILE/01.prepare_input.html">
     Prepare REPTILE
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../cluster_level/REPTILE/02.run_reptile.html">
     Predict Enhancer with REPTILE Algorithm
    </a>
   </li>
  </ul>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  COMMAND LINE TOOLS
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../command_line/allcools.html">
   Entry Point
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../command_line/allcools_allc.html">
   <code class="docutils literal notranslate">
    <span class="pre">
     allcools
    </span>
    <span class="pre">
     allc
    </span>
   </code>
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../command_line/allcools_standard.html">
   <code class="docutils literal notranslate">
    <span class="pre">
     allcools
    </span>
    <span class="pre">
     standard
    </span>
   </code>
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../command_line/allcools_profile.html">
   <code class="docutils literal notranslate">
    <span class="pre">
     allcools
    </span>
    <span class="pre">
     profile
    </span>
   </code>
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../command_line/allcools_tbi.html">
   <code class="docutils literal notranslate">
    <span class="pre">
     allcools
    </span>
    <span class="pre">
     tbi
    </span>
   </code>
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../command_line/allcools_merge.html">
   <code class="docutils literal notranslate">
    <span class="pre">
     allcools
    </span>
    <span class="pre">
     merge
    </span>
   </code>
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../command_line/allcools_extract.html">
   <code class="docutils literal notranslate">
    <span class="pre">
     allcools
    </span>
    <span class="pre">
     extract
    </span>
   </code>
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../command_line/allcools_bw.html">
   <code class="docutils literal notranslate">
    <span class="pre">
     allcools
    </span>
    <span class="pre">
     bw
    </span>
   </code>
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../command_line/allcools_region.html">
   <code class="docutils literal notranslate">
    <span class="pre">
     allcools
    </span>
    <span class="pre">
     region
    </span>
   </code>
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../command_line/allcools_dataset.html">
   <code class="docutils literal notranslate">
    <span class="pre">
     allcools
    </span>
    <span class="pre">
     generate-dataset
    </span>
   </code>
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../command_line/allcools_mcds.html">
   <code class="docutils literal notranslate">
    <span class="pre">
     allcools
    </span>
    <span class="pre">
     mcds
    </span>
   </code>
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../command_line/allcools_mcad.html">
   <code class="docutils literal notranslate">
    <span class="pre">
     allcools
    </span>
    <span class="pre">
     mcad
    </span>
   </code>
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../command_line/allcools_table_to_allc.html">
   <code class="docutils literal notranslate">
    <span class="pre">
     allcools
    </span>
    <span class="pre">
     table-to-allc
    </span>
   </code>
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  API
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../api/ALLCools/index.html">
   <code class="xref py py-mod docutils literal notranslate">
    <span class="pre">
     ALLCools
    </span>
   </code>
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-12" name="toctree-checkbox-12" type="checkbox"/>
  <label for="toctree-checkbox-12">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2 has-children">
    <a class="reference internal" href="../api/ALLCools/abc/index.html">
     <code class="xref py py-mod docutils literal notranslate">
      <span class="pre">
       ALLCools.abc
      </span>
     </code>
    </a>
    <input class="toctree-checkbox" id="toctree-checkbox-13" name="toctree-checkbox-13" type="checkbox"/>
    <label for="toctree-checkbox-13">
     <i class="fas fa-chevron-down">
     </i>
    </label>
    <ul>
     <li class="toctree-l3">
      <a class="reference internal" href="../api/ALLCools/abc/abc_score/index.html">
       <code class="xref py py-mod docutils literal notranslate">
        <span class="pre">
         ALLCools.abc.abc_score
        </span>
       </code>
      </a>
     </li>
    </ul>
   </li>
   <li class="toctree-l2 has-children">
    <a class="reference internal" href="../api/ALLCools/clustering/index.html">
     <code class="xref py py-mod docutils literal notranslate">
      <span class="pre">
       ALLCools.clustering
      </span>
     </code>
    </a>
    <input class="toctree-checkbox" id="toctree-checkbox-14" name="toctree-checkbox-14" type="checkbox"/>
    <label for="toctree-checkbox-14">
     <i class="fas fa-chevron-down">
     </i>
    </label>
    <ul>
     <li class="toctree-l3 has-children">
      <a class="reference internal" href="../api/ALLCools/clustering/doublets/index.html">
       <code class="xref py py-mod docutils literal notranslate">
        <span class="pre">
         ALLCools.clustering.doublets
        </span>
       </code>
      </a>
      <input class="toctree-checkbox" id="toctree-checkbox-15" name="toctree-checkbox-15" type="checkbox"/>
      <label for="toctree-checkbox-15">
       <i class="fas fa-chevron-down">
       </i>
      </label>
      <ul>
       <li class="toctree-l4">
        <a class="reference internal" href="../api/ALLCools/clustering/doublets/coverage_doublets/index.html">
         <code class="xref py py-mod docutils literal notranslate">
          <span class="pre">
           ALLCools.clustering.doublets.coverage_doublets
          </span>
         </code>
        </a>
       </li>
       <li class="toctree-l4">
        <a class="reference internal" href="../api/ALLCools/clustering/doublets/scrublet/index.html">
         <code class="xref py py-mod docutils literal notranslate">
          <span class="pre">
           ALLCools.clustering.doublets.scrublet
          </span>
         </code>
        </a>
       </li>
      </ul>
     </li>
     <li class="toctree-l3 has-children">
      <a class="reference internal" href="../api/ALLCools/clustering/feature_selection/index.html">
       <code class="xref py py-mod docutils literal notranslate">
        <span class="pre">
         ALLCools.clustering.feature_selection
        </span>
       </code>
      </a>
      <input class="toctree-checkbox" id="toctree-checkbox-16" name="toctree-checkbox-16" type="checkbox"/>
      <label for="toctree-checkbox-16">
       <i class="fas fa-chevron-down">
       </i>
      </label>
      <ul>
       <li class="toctree-l4">
        <a class="reference internal" href="../api/ALLCools/clustering/feature_selection/feature_enrichment/index.html">
         <code class="xref py py-mod docutils literal notranslate">
          <span class="pre">
           ALLCools.clustering.feature_selection.feature_enrichment
          </span>
         </code>
        </a>
       </li>
       <li class="toctree-l4">
        <a class="reference internal" href="../api/ALLCools/clustering/feature_selection/gene_panel_design/index.html">
         <code class="xref py py-mod docutils literal notranslate">
          <span class="pre">
           ALLCools.clustering.feature_selection.gene_panel_design
          </span>
         </code>
        </a>
       </li>
      </ul>
     </li>
     <li class="toctree-l3">
      <a class="reference internal" href="../api/ALLCools/clustering/ClusterMerging/index.html">
       <code class="xref py py-mod docutils literal notranslate">
        <span class="pre">
         ALLCools.clustering.ClusterMerging
        </span>
       </code>
      </a>
     </li>
     <li class="toctree-l3">
      <a class="reference internal" href="../api/ALLCools/clustering/ConsensusClustering/index.html">
       <code class="xref py py-mod docutils literal notranslate">
        <span class="pre">
         ALLCools.clustering.ConsensusClustering
        </span>
       </code>
      </a>
     </li>
     <li class="toctree-l3">
      <a class="reference internal" href="../api/ALLCools/clustering/art_of_tsne/index.html">
       <code class="xref py py-mod docutils literal notranslate">
        <span class="pre">
         ALLCools.clustering.art_of_tsne
        </span>
       </code>
      </a>
     </li>
     <li class="toctree-l3">
      <a class="reference internal" href="../api/ALLCools/clustering/balanced_pca/index.html">
       <code class="xref py py-mod docutils literal notranslate">
        <span class="pre">
         ALLCools.clustering.balanced_pca
        </span>
       </code>
      </a>
     </li>
     <li class="toctree-l3">
      <a class="reference internal" href="../api/ALLCools/clustering/chromatin_conformation/index.html">
       <code class="xref py py-mod docutils literal notranslate">
        <span class="pre">
         ALLCools.clustering.chromatin_conformation
        </span>
       </code>
      </a>
     </li>
     <li class="toctree-l3">
      <a class="reference internal" href="../api/ALLCools/clustering/dmg/index.html">
       <code class="xref py py-mod docutils literal notranslate">
        <span class="pre">
         ALLCools.clustering.dmg
        </span>
       </code>
      </a>
     </li>
     <li class="toctree-l3">
      <a class="reference internal" href="../api/ALLCools/clustering/incremental_pca/index.html">
       <code class="xref py py-mod docutils literal notranslate">
        <span class="pre">
         ALLCools.clustering.incremental_pca
        </span>
       </code>
      </a>
     </li>
     <li class="toctree-l3">
      <a class="reference internal" href="../api/ALLCools/clustering/lsi/index.html">
       <code class="xref py py-mod docutils literal notranslate">
        <span class="pre">
         ALLCools.clustering.lsi
        </span>
       </code>
      </a>
     </li>
     <li class="toctree-l3">
      <a class="reference internal" href="../api/ALLCools/clustering/mcad/index.html">
       <code class="xref py py-mod docutils literal notranslate">
        <span class="pre">
         ALLCools.clustering.mcad
        </span>
       </code>
      </a>
     </li>
     <li class="toctree-l3">
      <a class="reference internal" href="../api/ALLCools/clustering/pvclust/index.html">
       <code class="xref py py-mod docutils literal notranslate">
        <span class="pre">
         ALLCools.clustering.pvclust
        </span>
       </code>
      </a>
     </li>
     <li class="toctree-l3">
      <a class="reference internal" href="../api/ALLCools/clustering/rutilities/index.html">
       <code class="xref py py-mod docutils literal notranslate">
        <span class="pre">
         ALLCools.clustering.rutilities
        </span>
       </code>
      </a>
     </li>
    </ul>
   </li>
   <li class="toctree-l2 has-children">
    <a class="reference internal" href="../api/ALLCools/count_matrix/index.html">
     <code class="xref py py-mod docutils literal notranslate">
      <span class="pre">
       ALLCools.count_matrix
      </span>
     </code>
    </a>
    <input class="toctree-checkbox" id="toctree-checkbox-17" name="toctree-checkbox-17" type="checkbox"/>
    <label for="toctree-checkbox-17">
     <i class="fas fa-chevron-down">
     </i>
    </label>
    <ul>
     <li class="toctree-l3">
      <a class="reference internal" href="../api/ALLCools/count_matrix/atac/index.html">
       <code class="xref py py-mod docutils literal notranslate">
        <span class="pre">
         ALLCools.count_matrix.atac
        </span>
       </code>
      </a>
     </li>
     <li class="toctree-l3">
      <a class="reference internal" href="../api/ALLCools/count_matrix/dataset/index.html">
       <code class="xref py py-mod docutils literal notranslate">
        <span class="pre">
         ALLCools.count_matrix.dataset
        </span>
       </code>
      </a>
     </li>
     <li class="toctree-l3">
      <a class="reference internal" href="../api/ALLCools/count_matrix/h5ad/index.html">
       <code class="xref py py-mod docutils literal notranslate">
        <span class="pre">
         ALLCools.count_matrix.h5ad
        </span>
       </code>
      </a>
     </li>
     <li class="toctree-l3">
      <a class="reference internal" href="../api/ALLCools/count_matrix/mcad/index.html">
       <code class="xref py py-mod docutils literal notranslate">
        <span class="pre">
         ALLCools.count_matrix.mcad
        </span>
       </code>
      </a>
     </li>
     <li class="toctree-l3">
      <a class="reference internal" href="../api/ALLCools/count_matrix/mcds/index.html">
       <code class="xref py py-mod docutils literal notranslate">
        <span class="pre">
         ALLCools.count_matrix.mcds
        </span>
       </code>
      </a>
     </li>
     <li class="toctree-l3">
      <a class="reference internal" href="../api/ALLCools/count_matrix/snap/index.html">
       <code class="xref py py-mod docutils literal notranslate">
        <span class="pre">
         ALLCools.count_matrix.snap
        </span>
       </code>
      </a>
     </li>
     <li class="toctree-l3">
      <a class="reference internal" href="../api/ALLCools/count_matrix/zarr/index.html">
       <code class="xref py py-mod docutils literal notranslate">
        <span class="pre">
         ALLCools.count_matrix.zarr
        </span>
       </code>
      </a>
     </li>
    </ul>
   </li>
   <li class="toctree-l2 has-children">
    <a class="reference internal" href="../api/ALLCools/dataset/index.html">
     <code class="xref py py-mod docutils literal notranslate">
      <span class="pre">
       ALLCools.dataset
      </span>
     </code>
    </a>
    <input class="toctree-checkbox" id="toctree-checkbox-18" name="toctree-checkbox-18" type="checkbox"/>
    <label for="toctree-checkbox-18">
     <i class="fas fa-chevron-down">
     </i>
    </label>
    <ul>
     <li class="toctree-l3">
      <a class="reference internal" href="../api/ALLCools/dataset/directory_structure/index.html">
       <code class="xref py py-mod docutils literal notranslate">
        <span class="pre">
         ALLCools.dataset.directory_structure
        </span>
       </code>
      </a>
     </li>
     <li class="toctree-l3">
      <a class="reference internal" href="../api/ALLCools/dataset/entry_point/index.html">
       <code class="xref py py-mod docutils literal notranslate">
        <span class="pre">
         ALLCools.dataset.entry_point
        </span>
       </code>
      </a>
     </li>
    </ul>
   </li>
   <li class="toctree-l2 has-children">
    <a class="reference internal" href="../api/ALLCools/dmr/index.html">
     <code class="xref py py-mod docutils literal notranslate">
      <span class="pre">
       ALLCools.dmr
      </span>
     </code>
    </a>
    <input class="toctree-checkbox" id="toctree-checkbox-19" name="toctree-checkbox-19" type="checkbox"/>
    <label for="toctree-checkbox-19">
     <i class="fas fa-chevron-down">
     </i>
    </label>
    <ul>
     <li class="toctree-l3">
      <a class="reference internal" href="../api/ALLCools/dmr/call_dmr/index.html">
       <code class="xref py py-mod docutils literal notranslate">
        <span class="pre">
         ALLCools.dmr.call_dmr
        </span>
       </code>
      </a>
     </li>
     <li class="toctree-l3">
      <a class="reference internal" href="../api/ALLCools/dmr/call_dms/index.html">
       <code class="xref py py-mod docutils literal notranslate">
        <span class="pre">
         ALLCools.dmr.call_dms
        </span>
       </code>
      </a>
     </li>
     <li class="toctree-l3">
      <a class="reference internal" href="../api/ALLCools/dmr/parse_methylpy/index.html">
       <code class="xref py py-mod docutils literal notranslate">
        <span class="pre">
         ALLCools.dmr.parse_methylpy
        </span>
       </code>
      </a>
     </li>
     <li class="toctree-l3">
      <a class="reference internal" href="../api/ALLCools/dmr/rms_test/index.html">
       <code class="xref py py-mod docutils literal notranslate">
        <span class="pre">
         ALLCools.dmr.rms_test
        </span>
       </code>
      </a>
     </li>
    </ul>
   </li>
   <li class="toctree-l2 has-children">
    <a class="reference internal" href="../api/ALLCools/gtf/index.html">
     <code class="xref py py-mod docutils literal notranslate">
      <span class="pre">
       ALLCools.gtf
      </span>
     </code>
    </a>
    <input class="toctree-checkbox" id="toctree-checkbox-20" name="toctree-checkbox-20" type="checkbox"/>
    <label for="toctree-checkbox-20">
     <i class="fas fa-chevron-down">
     </i>
    </label>
    <ul>
     <li class="toctree-l3">
      <a class="reference internal" href="../api/ALLCools/gtf/Gtf/index.html">
       <code class="xref py py-mod docutils literal notranslate">
        <span class="pre">
         ALLCools.gtf.Gtf
        </span>
       </code>
      </a>
     </li>
     <li class="toctree-l3">
      <a class="reference internal" href="../api/ALLCools/gtf/utilities/index.html">
       <code class="xref py py-mod docutils literal notranslate">
        <span class="pre">
         ALLCools.gtf.utilities
        </span>
       </code>
      </a>
     </li>
    </ul>
   </li>
   <li class="toctree-l2 has-children">
    <a class="reference internal" href="../api/ALLCools/integration/index.html">
     <code class="xref py py-mod docutils literal notranslate">
      <span class="pre">
       ALLCools.integration
      </span>
     </code>
    </a>
    <input class="toctree-checkbox" id="toctree-checkbox-21" name="toctree-checkbox-21" type="checkbox"/>
    <label for="toctree-checkbox-21">
     <i class="fas fa-chevron-down">
     </i>
    </label>
    <ul>
     <li class="toctree-l3">
      <a class="reference internal" href="../api/ALLCools/integration/cca/index.html">
       <code class="xref py py-mod docutils literal notranslate">
        <span class="pre">
         ALLCools.integration.cca
        </span>
       </code>
      </a>
     </li>
     <li class="toctree-l3">
      <a class="reference internal" href="../api/ALLCools/integration/confusion/index.html">
       <code class="xref py py-mod docutils literal notranslate">
        <span class="pre">
         ALLCools.integration.confusion
        </span>
       </code>
      </a>
     </li>
     <li class="toctree-l3">
      <a class="reference internal" href="../api/ALLCools/integration/metric/index.html">
       <code class="xref py py-mod docutils literal notranslate">
        <span class="pre">
         ALLCools.integration.metric
        </span>
       </code>
      </a>
     </li>
     <li class="toctree-l3">
      <a class="reference internal" href="../api/ALLCools/integration/seurat_class/index.html">
       <code class="xref py py-mod docutils literal notranslate">
        <span class="pre">
         ALLCools.integration.seurat_class
        </span>
       </code>
      </a>
     </li>
    </ul>
   </li>
   <li class="toctree-l2 has-children">
    <a class="reference internal" href="../api/ALLCools/mcds/index.html">
     <code class="xref py py-mod docutils literal notranslate">
      <span class="pre">
       ALLCools.mcds
      </span>
     </code>
    </a>
    <input class="toctree-checkbox" id="toctree-checkbox-22" name="toctree-checkbox-22" type="checkbox"/>
    <label for="toctree-checkbox-22">
     <i class="fas fa-chevron-down">
     </i>
    </label>
    <ul>
     <li class="toctree-l3">
      <a class="reference internal" href="../api/ALLCools/mcds/correlation/index.html">
       <code class="xref py py-mod docutils literal notranslate">
        <span class="pre">
         ALLCools.mcds.correlation
        </span>
       </code>
      </a>
     </li>
     <li class="toctree-l3">
      <a class="reference internal" href="../api/ALLCools/mcds/mcds/index.html">
       <code class="xref py py-mod docutils literal notranslate">
        <span class="pre">
         ALLCools.mcds.mcds
        </span>
       </code>
      </a>
     </li>
     <li class="toctree-l3">
      <a class="reference internal" href="../api/ALLCools/mcds/region_ds/index.html">
       <code class="xref py py-mod docutils literal notranslate">
        <span class="pre">
         ALLCools.mcds.region_ds
        </span>
       </code>
      </a>
     </li>
     <li class="toctree-l3">
      <a class="reference internal" href="../api/ALLCools/mcds/utilities/index.html">
       <code class="xref py py-mod docutils literal notranslate">
        <span class="pre">
         ALLCools.mcds.utilities
        </span>
       </code>
      </a>
     </li>
    </ul>
   </li>
   <li class="toctree-l2 has-children">
    <a class="reference internal" href="../api/ALLCools/motif/index.html">
     <code class="xref py py-mod docutils literal notranslate">
      <span class="pre">
       ALLCools.motif
      </span>
     </code>
    </a>
    <input class="toctree-checkbox" id="toctree-checkbox-23" name="toctree-checkbox-23" type="checkbox"/>
    <label for="toctree-checkbox-23">
     <i class="fas fa-chevron-down">
     </i>
    </label>
    <ul>
     <li class="toctree-l3">
      <a class="reference internal" href="../api/ALLCools/motif/default_motif_set/index.html">
       <code class="xref py py-mod docutils literal notranslate">
        <span class="pre">
         ALLCools.motif.default_motif_set
        </span>
       </code>
      </a>
     </li>
     <li class="toctree-l3">
      <a class="reference internal" href="../api/ALLCools/motif/motifs/index.html">
       <code class="xref py py-mod docutils literal notranslate">
        <span class="pre">
         ALLCools.motif.motifs
        </span>
       </code>
      </a>
     </li>
     <li class="toctree-l3">
      <a class="reference internal" href="../api/ALLCools/motif/parse_meme/index.html">
       <code class="xref py py-mod docutils literal notranslate">
        <span class="pre">
         ALLCools.motif.parse_meme
        </span>
       </code>
      </a>
     </li>
     <li class="toctree-l3">
      <a class="reference internal" href="../api/ALLCools/motif/snakemake/index.html">
       <code class="xref py py-mod docutils literal notranslate">
        <span class="pre">
         ALLCools.motif.snakemake
        </span>
       </code>
      </a>
     </li>
    </ul>
   </li>
   <li class="toctree-l2 has-children">
    <a class="reference internal" href="../api/ALLCools/plot/index.html">
     <code class="xref py py-mod docutils literal notranslate">
      <span class="pre">
       ALLCools.plot
      </span>
     </code>
    </a>
    <input class="toctree-checkbox" id="toctree-checkbox-24" name="toctree-checkbox-24" type="checkbox"/>
    <label for="toctree-checkbox-24">
     <i class="fas fa-chevron-down">
     </i>
    </label>
    <ul>
     <li class="toctree-l3 has-children">
      <a class="reference internal" href="../api/ALLCools/plot/genome_track/index.html">
       <code class="xref py py-mod docutils literal notranslate">
        <span class="pre">
         ALLCools.plot.genome_track
        </span>
       </code>
      </a>
      <input class="toctree-checkbox" id="toctree-checkbox-25" name="toctree-checkbox-25" type="checkbox"/>
      <label for="toctree-checkbox-25">
       <i class="fas fa-chevron-down">
       </i>
      </label>
      <ul>
       <li class="toctree-l4">
        <a class="reference internal" href="../api/ALLCools/plot/genome_track/GtfTrack/index.html">
         <code class="xref py py-mod docutils literal notranslate">
          <span class="pre">
           ALLCools.plot.genome_track.GtfTrack
          </span>
         </code>
        </a>
       </li>
       <li class="toctree-l4">
        <a class="reference internal" href="../api/ALLCools/plot/genome_track/HiCMatrixCoolTrack/index.html">
         <code class="xref py py-mod docutils literal notranslate">
          <span class="pre">
           ALLCools.plot.genome_track.HiCMatrixCoolTrack
          </span>
         </code>
        </a>
       </li>
       <li class="toctree-l4">
        <a class="reference internal" href="../api/ALLCools/plot/genome_track/_pygenometrack/index.html">
         <code class="xref py py-mod docutils literal notranslate">
          <span class="pre">
           ALLCools.plot.genome_track._pygenometrack
          </span>
         </code>
        </a>
       </li>
      </ul>
     </li>
     <li class="toctree-l3">
      <a class="reference internal" href="../api/ALLCools/plot/categorical_scatter/index.html">
       <code class="xref py py-mod docutils literal notranslate">
        <span class="pre">
         ALLCools.plot.categorical_scatter
        </span>
       </code>
      </a>
     </li>
     <li class="toctree-l3">
      <a class="reference internal" href="../api/ALLCools/plot/color/index.html">
       <code class="xref py py-mod docutils literal notranslate">
        <span class="pre">
         ALLCools.plot.color
        </span>
       </code>
      </a>
     </li>
     <li class="toctree-l3">
      <a class="reference internal" href="../api/ALLCools/plot/continuous_scatter/index.html">
       <code class="xref py py-mod docutils literal notranslate">
        <span class="pre">
         ALLCools.plot.continuous_scatter
        </span>
       </code>
      </a>
     </li>
     <li class="toctree-l3">
      <a class="reference internal" href="../api/ALLCools/plot/contour/index.html">
       <code class="xref py py-mod docutils literal notranslate">
        <span class="pre">
         ALLCools.plot.contour
        </span>
       </code>
      </a>
     </li>
     <li class="toctree-l3">
      <a class="reference internal" href="../api/ALLCools/plot/decomposition/index.html">
       <code class="xref py py-mod docutils literal notranslate">
        <span class="pre">
         ALLCools.plot.decomposition
        </span>
       </code>
      </a>
     </li>
     <li class="toctree-l3">
      <a class="reference internal" href="../api/ALLCools/plot/dendro/index.html">
       <code class="xref py py-mod docutils literal notranslate">
        <span class="pre">
         ALLCools.plot.dendro
        </span>
       </code>
      </a>
     </li>
     <li class="toctree-l3">
      <a class="reference internal" href="../api/ALLCools/plot/interactive_scatter/index.html">
       <code class="xref py py-mod docutils literal notranslate">
        <span class="pre">
         ALLCools.plot.interactive_scatter
        </span>
       </code>
      </a>
     </li>
     <li class="toctree-l3">
      <a class="reference internal" href="../api/ALLCools/plot/qc_plots/index.html">
       <code class="xref py py-mod docutils literal notranslate">
        <span class="pre">
         ALLCools.plot.qc_plots
        </span>
       </code>
      </a>
     </li>
     <li class="toctree-l3">
      <a class="reference internal" href="../api/ALLCools/plot/size/index.html">
       <code class="xref py py-mod docutils literal notranslate">
        <span class="pre">
         ALLCools.plot.size
        </span>
       </code>
      </a>
     </li>
     <li class="toctree-l3">
      <a class="reference internal" href="../api/ALLCools/plot/sunburst/index.html">
       <code class="xref py py-mod docutils literal notranslate">
        <span class="pre">
         ALLCools.plot.sunburst
        </span>
       </code>
      </a>
     </li>
     <li class="toctree-l3">
      <a class="reference internal" href="../api/ALLCools/plot/text_anno_scatter/index.html">
       <code class="xref py py-mod docutils literal notranslate">
        <span class="pre">
         ALLCools.plot.text_anno_scatter
        </span>
       </code>
      </a>
     </li>
     <li class="toctree-l3">
      <a class="reference internal" href="../api/ALLCools/plot/utilities/index.html">
       <code class="xref py py-mod docutils literal notranslate">
        <span class="pre">
         ALLCools.plot.utilities
        </span>
       </code>
      </a>
     </li>
    </ul>
   </li>
   <li class="toctree-l2 has-children">
    <a class="reference internal" href="../api/ALLCools/pseudo_cell/index.html">
     <code class="xref py py-mod docutils literal notranslate">
      <span class="pre">
       ALLCools.pseudo_cell
      </span>
     </code>
    </a>
    <input class="toctree-checkbox" id="toctree-checkbox-26" name="toctree-checkbox-26" type="checkbox"/>
    <label for="toctree-checkbox-26">
     <i class="fas fa-chevron-down">
     </i>
    </label>
    <ul>
     <li class="toctree-l3">
      <a class="reference internal" href="../api/ALLCools/pseudo_cell/pseudo_cell_kmeans/index.html">
       <code class="xref py py-mod docutils literal notranslate">
        <span class="pre">
         ALLCools.pseudo_cell.pseudo_cell_kmeans
        </span>
       </code>
      </a>
     </li>
     <li class="toctree-l3">
      <a class="reference internal" href="../api/ALLCools/pseudo_cell/pseudo_cell_knn/index.html">
       <code class="xref py py-mod docutils literal notranslate">
        <span class="pre">
         ALLCools.pseudo_cell.pseudo_cell_knn
        </span>
       </code>
      </a>
     </li>
    </ul>
   </li>
   <li class="toctree-l2 has-children">
    <a class="reference internal" href="../api/ALLCools/reptile/index.html">
     <code class="xref py py-mod docutils literal notranslate">
      <span class="pre">
       ALLCools.reptile
      </span>
     </code>
    </a>
    <input class="toctree-checkbox" id="toctree-checkbox-27" name="toctree-checkbox-27" type="checkbox"/>
    <label for="toctree-checkbox-27">
     <i class="fas fa-chevron-down">
     </i>
    </label>
    <ul>
     <li class="toctree-l3">
      <a class="reference internal" href="../api/ALLCools/reptile/reptile/index.html">
       <code class="xref py py-mod docutils literal notranslate">
        <span class="pre">
         ALLCools.reptile.reptile
        </span>
       </code>
      </a>
     </li>
    </ul>
   </li>
   <li class="toctree-l2 has-children">
    <a class="reference internal" href="../api/ALLCools/schema/index.html">
     <code class="xref py py-mod docutils literal notranslate">
      <span class="pre">
       ALLCools.schema
      </span>
     </code>
    </a>
    <input class="toctree-checkbox" id="toctree-checkbox-28" name="toctree-checkbox-28" type="checkbox"/>
    <label for="toctree-checkbox-28">
     <i class="fas fa-chevron-down">
     </i>
    </label>
    <ul>
     <li class="toctree-l3">
      <a class="reference internal" href="../api/ALLCools/schema/mcds_schema/index.html">
       <code class="xref py py-mod docutils literal notranslate">
        <span class="pre">
         ALLCools.schema.mcds_schema
        </span>
       </code>
      </a>
     </li>
    </ul>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../api/ALLCools/__main__/index.html">
     <code class="xref py py-mod docutils literal notranslate">
      <span class="pre">
       ALLCools.__main__
      </span>
     </code>
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../api/ALLCools/_allc_to_bigwig/index.html">
     <code class="xref py py-mod docutils literal notranslate">
      <span class="pre">
       ALLCools._allc_to_bigwig
      </span>
     </code>
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../api/ALLCools/_allc_to_bigwig_new/index.html">
     <code class="xref py py-mod docutils literal notranslate">
      <span class="pre">
       ALLCools._allc_to_bigwig_new
      </span>
     </code>
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../api/ALLCools/_allc_to_region_count/index.html">
     <code class="xref py py-mod docutils literal notranslate">
      <span class="pre">
       ALLCools._allc_to_region_count
      </span>
     </code>
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../api/ALLCools/_bam_to_allc/index.html">
     <code class="xref py py-mod docutils literal notranslate">
      <span class="pre">
       ALLCools._bam_to_allc
      </span>
     </code>
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../api/ALLCools/_doc/index.html">
     <code class="xref py py-mod docutils literal notranslate">
      <span class="pre">
       ALLCools._doc
      </span>
     </code>
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../api/ALLCools/_extract_allc/index.html">
     <code class="xref py py-mod docutils literal notranslate">
      <span class="pre">
       ALLCools._extract_allc
      </span>
     </code>
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../api/ALLCools/_merge_allc/index.html">
     <code class="xref py py-mod docutils literal notranslate">
      <span class="pre">
       ALLCools._merge_allc
      </span>
     </code>
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../api/ALLCools/_open/index.html">
     <code class="xref py py-mod docutils literal notranslate">
      <span class="pre">
       ALLCools._open
      </span>
     </code>
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../api/ALLCools/api/index.html">
     <code class="xref py py-mod docutils literal notranslate">
      <span class="pre">
       ALLCools.api
      </span>
     </code>
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../api/ALLCools/table_to_allc/index.html">
     <code class="xref py py-mod docutils literal notranslate">
      <span class="pre">
       ALLCools.table_to_allc
      </span>
     </code>
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../api/ALLCools/utilities/index.html">
     <code class="xref py py-mod docutils literal notranslate">
      <span class="pre">
       ALLCools.utilities
      </span>
     </code>
    </a>
   </li>
  </ul>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  PROJECT INFO
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../project_info/citation.html">
   Citation And Reference
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference external" href="https://github.com/lhqing/ALLCools/releases">
   Changelog
  </a>
 </li>
</ul>

    </div>
</nav></div>
        <div class="bd-sidebar__bottom">
             <!-- To handle the deprecated key -->
            
            <div class="navbar_extra_footer">
            Powered by <a href="https://jupyterbook.org">Jupyter Book</a>
            </div>
            
        </div>
    </div>
    <div id="rtd-footer-container"></div>
</div>


          


          
<!-- A tiny helper pixel to detect if we've scrolled -->
<div class="sbt-scroll-pixel-helper"></div>
<!-- Main content -->
<div class="col py-0 content-container">
    
    <div class="header-article row sticky-top noprint">
        



<div class="col py-1 d-flex header-article-main">
    <div class="header-article__left">
        
        <label for="__navigation"
  class="headerbtn"
  data-toggle="tooltip"
data-placement="right"
title="Toggle navigation"
>
  

<span class="headerbtn__icon-container">
  <i class="fas fa-bars"></i>
  </span>

</label>

        
    </div>
    <div class="header-article__right">
<button onclick="toggleFullScreen()"
  class="headerbtn"
  data-toggle="tooltip"
data-placement="bottom"
title="Fullscreen mode"
>
  

<span class="headerbtn__icon-container">
  <i class="fas fa-expand"></i>
  </span>

</button>

<div class="menu-dropdown menu-dropdown-repository-buttons">
  <button class="headerbtn menu-dropdown__trigger"
      aria-label="Source repositories">
      <i class="fab fa-github"></i>
  </button>
  <div class="menu-dropdown__content">
    <ul>
      <li>
        <a href="https://github.com/lhqing/ALLCools"
   class="headerbtn"
   data-toggle="tooltip"
data-placement="left"
title="Source repository"
>
  

<span class="headerbtn__icon-container">
  <i class="fab fa-github"></i>
  </span>
<span class="headerbtn__text-container">repository</span>
</a>

      </li>
      
      <li>
        <a href="https://github.com/lhqing/ALLCools/issues/new?title=Issue%20on%20page%20%2Fstart/input_files.html&body=Your%20issue%20content%20here."
   class="headerbtn"
   data-toggle="tooltip"
data-placement="left"
title="Open an issue"
>
  

<span class="headerbtn__icon-container">
  <i class="fas fa-lightbulb"></i>
  </span>
<span class="headerbtn__text-container">open issue</span>
</a>

      </li>
      
      <li>
        <a href="https://github.com/lhqing/ALLCools/edit/master/docs/allcools/start/input_files.md"
   class="headerbtn"
   data-toggle="tooltip"
data-placement="left"
title="Edit this page"
>
  

<span class="headerbtn__icon-container">
  <i class="fas fa-pencil-alt"></i>
  </span>
<span class="headerbtn__text-container">suggest edit</span>
</a>

      </li>
      
    </ul>
  </div>
</div>

<div class="menu-dropdown menu-dropdown-download-buttons">
  <button class="headerbtn menu-dropdown__trigger"
      aria-label="Download this page">
      <i class="fas fa-download"></i>
  </button>
  <div class="menu-dropdown__content">
    <ul>
      <li>
        <a href="../_sources/start/input_files.md"
   class="headerbtn"
   data-toggle="tooltip"
data-placement="left"
title="Download source file"
>
  

<span class="headerbtn__icon-container">
  <i class="fas fa-file"></i>
  </span>
<span class="headerbtn__text-container">.md</span>
</a>

      </li>
      
      <li>
        
<button onclick="printPdf(this)"
  class="headerbtn"
  data-toggle="tooltip"
data-placement="left"
title="Print to PDF"
>
  

<span class="headerbtn__icon-container">
  <i class="fas fa-file-pdf"></i>
  </span>
<span class="headerbtn__text-container">.pdf</span>
</button>

      </li>
      
    </ul>
  </div>
</div>
<label for="__page-toc"
  class="headerbtn headerbtn-page-toc"
  
>
  

<span class="headerbtn__icon-container">
  <i class="fas fa-list"></i>
  </span>

</label>

    </div>
</div>

<!-- Table of contents -->
<div class="col-md-3 bd-toc show noprint">
    <div class="tocsection onthispage pt-5 pb-3">
        <i class="fas fa-list"></i> Contents
    </div>
    <nav id="bd-toc-nav" aria-label="Page">
        <ul class="visible nav section-nav flex-column">
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#allcools-starts-from-mapped-data">
   ALLCools Starts From Mapped Data
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#data-analysis-starting-from-core-file-formats">
   Data Analysis Starting From Core File Formats
  </a>
  <ul class="nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#relationships-between-three-data-file-formats">
     Relationships between three data file formats
    </a>
   </li>
  </ul>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#base-resolution-count-table">
   Base Resolution Count Table
  </a>
  <ul class="nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#allc-file">
     ALLC File
    </a>
    <ul class="nav section-nav flex-column">
     <li class="toc-h4 nav-item toc-entry">
      <a class="reference internal nav-link" href="#columns-in-allc-file">
       Columns in ALLC file
      </a>
     </li>
     <li class="toc-h4 nav-item toc-entry">
      <a class="reference internal nav-link" href="#compression-and-indexing-allc">
       Compression and Indexing ALLC
      </a>
     </li>
    </ul>
   </li>
  </ul>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#cell-by-feature-methylome-matrix">
   Cell-by-feature Methylome Matrix
  </a>
  <ul class="nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#mcds-file-n-d-dense-matrix-for-raw-counts">
     MCDS File: N-D dense matrix for raw counts
    </a>
    <ul class="nav section-nav flex-column">
     <li class="toc-h4 nav-item toc-entry">
      <a class="reference internal nav-link" href="#raw-count-matrix-in-mcds">
       Raw count matrix in MCDS
      </a>
     </li>
     <li class="toc-h4 nav-item toc-entry">
      <a class="reference internal nav-link" href="#mcds-store-multiple-region-sets-together">
       MCDS store multiple region sets together
      </a>
     </li>
     <li class="toc-h4 nav-item toc-entry">
      <a class="reference internal nav-link" href="#mcds-enables-on-disk-data-handling">
       MCDS enables on-disk data handling
      </a>
     </li>
     <li class="toc-h4 nav-item toc-entry">
      <a class="reference internal nav-link" href="#when-quantifying-small-size-features">
       When quantifying small-size features
      </a>
     </li>
     <li class="toc-h4 nav-item toc-entry">
      <a class="reference internal nav-link" href="#hypo-methylation-and-hyper-methylation-score">
       Hypo-methylation and Hyper-methylation Score
      </a>
     </li>
    </ul>
   </li>
  </ul>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#annotated-genome-region-matrix">
   Annotated Genome Region Matrix
  </a>
  <ul class="nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#regionds-file-n-d-dense-matrix-for-genome-region-with-various-annotations">
     RegionDS file: N-D dense matrix for Genome Region with various annotations
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#relationships-between-regionds-and-mcds">
     Relationships between RegionDS and MCDS
    </a>
   </li>
  </ul>
 </li>
</ul>

    </nav>
</div>
    </div>
    <div class="article row">
        <div class="col pl-md-3 pl-lg-5 content-container">
            <!-- Table of contents that is only displayed when printing the page -->
            <div id="jb-print-docs-body" class="onlyprint">
                <h1>Input Files</h1>
                <!-- Table of contents -->
                <div id="print-main-content">
                    <div id="jb-print-toc">
                        
                        <div>
                            <h2> Contents </h2>
                        </div>
                        <nav aria-label="Page">
                            <ul class="visible nav section-nav flex-column">
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#allcools-starts-from-mapped-data">
   ALLCools Starts From Mapped Data
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#data-analysis-starting-from-core-file-formats">
   Data Analysis Starting From Core File Formats
  </a>
  <ul class="nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#relationships-between-three-data-file-formats">
     Relationships between three data file formats
    </a>
   </li>
  </ul>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#base-resolution-count-table">
   Base Resolution Count Table
  </a>
  <ul class="nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#allc-file">
     ALLC File
    </a>
    <ul class="nav section-nav flex-column">
     <li class="toc-h4 nav-item toc-entry">
      <a class="reference internal nav-link" href="#columns-in-allc-file">
       Columns in ALLC file
      </a>
     </li>
     <li class="toc-h4 nav-item toc-entry">
      <a class="reference internal nav-link" href="#compression-and-indexing-allc">
       Compression and Indexing ALLC
      </a>
     </li>
    </ul>
   </li>
  </ul>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#cell-by-feature-methylome-matrix">
   Cell-by-feature Methylome Matrix
  </a>
  <ul class="nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#mcds-file-n-d-dense-matrix-for-raw-counts">
     MCDS File: N-D dense matrix for raw counts
    </a>
    <ul class="nav section-nav flex-column">
     <li class="toc-h4 nav-item toc-entry">
      <a class="reference internal nav-link" href="#raw-count-matrix-in-mcds">
       Raw count matrix in MCDS
      </a>
     </li>
     <li class="toc-h4 nav-item toc-entry">
      <a class="reference internal nav-link" href="#mcds-store-multiple-region-sets-together">
       MCDS store multiple region sets together
      </a>
     </li>
     <li class="toc-h4 nav-item toc-entry">
      <a class="reference internal nav-link" href="#mcds-enables-on-disk-data-handling">
       MCDS enables on-disk data handling
      </a>
     </li>
     <li class="toc-h4 nav-item toc-entry">
      <a class="reference internal nav-link" href="#when-quantifying-small-size-features">
       When quantifying small-size features
      </a>
     </li>
     <li class="toc-h4 nav-item toc-entry">
      <a class="reference internal nav-link" href="#hypo-methylation-and-hyper-methylation-score">
       Hypo-methylation and Hyper-methylation Score
      </a>
     </li>
    </ul>
   </li>
  </ul>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#annotated-genome-region-matrix">
   Annotated Genome Region Matrix
  </a>
  <ul class="nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#regionds-file-n-d-dense-matrix-for-genome-region-with-various-annotations">
     RegionDS file: N-D dense matrix for Genome Region with various annotations
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#relationships-between-regionds-and-mcds">
     Relationships between RegionDS and MCDS
    </a>
   </li>
  </ul>
 </li>
</ul>

                        </nav>
                    </div>
                </div>
            </div>
            <main id="main-content" role="main">
                
              <div>
                
  <div class="tex2jax_ignore mathjax_ignore section" id="input-files">
<h1>Input Files<a class="headerlink" href="#input-files" title="Permalink to this headline">#</a></h1>
<div class="section" id="allcools-starts-from-mapped-data">
<h2>ALLCools Starts From Mapped Data<a class="headerlink" href="#allcools-starts-from-mapped-data" title="Permalink to this headline">#</a></h2>
<p>ALLCools is designed for post-mapping data preprocessing data analysis. Therefore, our tool starts from the mapped and demultiplexed data. Please refer to the data provider or corresponding methods/pipelines for demultiplexing single-cell reads and mapping them on to the reference genome.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>For all the data generated by snmC-seq and related single-cell multi-omics technologies in Ecker Lab, we use the <a class="reference external" href="https://hq-1.gitbook.io/mc/">YAP</a> pipeline to demultiplex and map the reads. YAP uses bismark to map reads and generate single-cell ALLC files from each filtered BAM files.</p>
</div>
<p>After demultiplexing and mapping, you will get one BAM file per cell containing sorted and uniquely mapped reads. From each BAM file, you can create one ALLC file using <a class="reference internal" href="../command_line/allcools.html"><span class="doc std std-doc"><code class="docutils literal notranslate"><span class="pre">allcools</span></code> command line tools</span></a>.</p>
</div>
<div class="section" id="data-analysis-starting-from-core-file-formats">
<h2>Data Analysis Starting From Core File Formats<a class="headerlink" href="#data-analysis-starting-from-core-file-formats" title="Permalink to this headline">#</a></h2>
<p>We use three file formats to store cellular analysis and genomic analysis data.</p>
<div class="docutils">
<table class="colwidths-auto table">
<thead>
<tr class="row-odd"><th class="head"><p><strong>File Format</strong></p></th>
<th class="head"><p><strong>Analysis Level</strong></p></th>
<th class="head"><p><strong>Genome Resolution</strong></p></th>
<th class="head"><p><strong>Data Type</strong></p></th>
<th class="head"><p><strong>mC Context</strong></p></th>
<th class="head"><p><strong>Matrix Form</strong></p></th>
<th class="head"><p><strong>Based On</strong></p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>ALLC</p></td>
<td><p>Cellular</p></td>
<td><p>single base pair</p></td>
<td><p>raw counts</p></td>
<td><p>all mC context</p></td>
<td><p>N.A.</p></td>
<td><p>tab-separated table, compressed by <a class="reference external" href="http://www.htslib.org/doc/bgzip.html">bgzip</a>, indexed by <a class="reference external" href="http://www.htslib.org/doc/tabix.html">tabix</a></p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="../api/ALLCools/mcds/mcds/index.html#ALLCools.mcds.mcds.MCDS" title="ALLCools.mcds.mcds.MCDS"><code class="xref py py-class docutils literal notranslate"><span class="pre">MCDS</span></code></a></p></td>
<td><p>Cellular</p></td>
<td><p>genome regions</p></td>
<td><p>raw counts &amp; </br>miscellaneous</p></td>
<td><p>multiple user-defined mC context</p></td>
<td><p>N-D Dense Matrix</p></td>
<td><p><a class="reference external" href="https://xarray.pydata.org/en/stable/generated/xarray.Dataset.html#xarray.Dataset" title="(in xarray v2022.6.1.dev0)"><code class="xref py py-class docutils literal notranslate"><span class="pre">xarray.Dataset</span></code></a></p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="../api/ALLCools/mcds/region_ds/index.html#ALLCools.mcds.region_ds.RegionDS" title="ALLCools.mcds.region_ds.RegionDS"><code class="xref py py-class docutils literal notranslate"><span class="pre">RegionDS</span></code></a></p></td>
<td><p>Genomic</p></td>
<td><p>single base pair to large genome regions</p></td>
<td><p>raw counts &amp; </br>miscellaneous</p></td>
<td><p>miscellaneous</p></td>
<td><p>N-D Dense Matrix</p></td>
<td><p><a class="reference external" href="https://xarray.pydata.org/en/stable/generated/xarray.Dataset.html#xarray.Dataset" title="(in xarray v2022.6.1.dev0)"><code class="xref py py-class docutils literal notranslate"><span class="pre">xarray.Dataset</span></code></a></p></td>
</tr>
</tbody>
</table>
</div>
<div class="section" id="relationships-between-three-data-file-formats">
<h3>Relationships between three data file formats<a class="headerlink" href="#relationships-between-three-data-file-formats" title="Permalink to this headline">#</a></h3>
<p>The ALLC files are generated from the raw sequencing data, containing the base-level cytosine counts (raw data). Subsequently, the <a class="reference internal" href="../api/ALLCools/mcds/mcds/index.html#ALLCools.mcds.mcds.MCDS" title="ALLCools.mcds.mcds.MCDS"><code class="xref py py-class docutils literal notranslate"><span class="pre">MCDS</span></code></a> file is generated from a set of ALLC files. The MCDS file is the starting point of cellular analysis (cell-by-region matrix). The RegionDS file is the starting point of genomic analysis (region-by-annotation matrix).</p>
<div class="figure align-default" id="data-model-fig">
<a class="reference internal image-reference" href="../_images/data_model.png"><img alt="../_images/data_model.png" src="../_images/data_model.png" style="height: 300px;" /></a>
<p class="caption"><span class="caption-number">Fig. 3 </span><span class="caption-text">ALLCools data model for handling single-cell methylomes and performing cellular and genomic analysis.</span><a class="headerlink" href="#data-model-fig" title="Permalink to this image">#</a></p>
</div>
</div>
</div>
<div class="section" id="base-resolution-count-table">
<h2>Base Resolution Count Table<a class="headerlink" href="#base-resolution-count-table" title="Permalink to this headline">#</a></h2>
<div class="section" id="allc-file">
<h3>ALLC File<a class="headerlink" href="#allc-file" title="Permalink to this headline">#</a></h3>
<p>The ALLC (ALL Cytosine) format is a tab-separated table containing <strong>base level methylation and coverage counts</strong>. This format is originally defined by <a class="reference external" href="https://github.com/yupenghe/methylpy"><code class="docutils literal notranslate"><span class="pre">methylpy</span></code></a>, a python package developed in <a class="reference external" href="https://ecker.salk.edu/">the Ecker lab</a> for bulk WGBS-seq data analysis. Each row in an ALLC file corresponds to one cytosine in the genome. An ALLC file contains 7 mandatory columns with no header. ALLCools assumes all the ALLC files are compressed and indexed by <a class="reference external" href="http://www.htslib.org/doc/bgzip.html"><code class="docutils literal notranslate"><span class="pre">bgzip</span></code></a> and <a class="reference external" href="http://www.htslib.org/doc/tabix.html"><code class="docutils literal notranslate"><span class="pre">tabix</span></code></a> from the <a class="reference external" href="http://www.htslib.org/">htslib</a> <span id="id1">[<a class="reference internal" href="../project_info/citation.html#id13" title="Heng Li. Tabix: fast retrieval of sequence features from generic TAB-delimited files. Bioinformatics, 27(5):718719, March 2011.">Li, 2011</a>]</span>. The ALLC file <a class="reference internal" href="../command_line/allcools_allc.html"><span class="doc std std-doc">generated from a single-cell BAM file</span></a> only contains information from a single cell. Multiple ALLC files can also be <a class="reference internal" href="../command_line/allcools_merge.html"><span class="doc std std-doc">merged</span></a> into an ALLC file (e.g., merge by cell cluster) as a pseudo-bulk-level methylation table.</p>
<p>For more details on generating and handling ALLC files, please read the introduction of <a class="reference internal" href="../command_line/allcools.html"><span class="doc std std-doc"><code class="docutils literal notranslate"><span class="pre">allcools</span></code> command line tools</span></a>.</p>
<div class="section" id="columns-in-allc-file">
<h4>Columns in ALLC file<a class="headerlink" href="#columns-in-allc-file" title="Permalink to this headline">#</a></h4>
<table class="colwidths-auto table">
<thead>
<tr class="row-odd"><th class="head"><p><strong>index</strong></p></th>
<th class="head"><p><strong>column name</strong></p></th>
<th class="head"><p><strong>example</strong></p></th>
<th class="head"><p><strong>note</strong></p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>1</p></td>
<td><p>chromosome</p></td>
<td><p>chr12</p></td>
<td><p>chromosome names need to be the same as genome FASTA</p></td>
</tr>
<tr class="row-odd"><td><p>2</p></td>
<td><p>position</p></td>
<td><p>18283342</p></td>
<td><p>1-based</p></td>
</tr>
<tr class="row-even"><td><p>3</p></td>
<td><p>strand</p></td>
<td><p>+</p></td>
<td><p>either + or -</p></td>
</tr>
<tr class="row-odd"><td><p>4</p></td>
<td><p>sequence context</p></td>
<td><p>CGT</p></td>
<td><p>can be more than 3 bases</p></td>
</tr>
<tr class="row-even"><td><p>5</p></td>
<td><p>mc</p></td>
<td><p>1</p></td>
<td><p>count of reads supporting methylation</p></td>
</tr>
<tr class="row-odd"><td><p>6</p></td>
<td><p>cov</p></td>
<td><p>2</p></td>
<td><p>read coverage</p></td>
</tr>
<tr class="row-even"><td><p>7</p></td>
<td><p>methylated</p></td>
<td><p>1</p></td>
<td><p>indicator of significant methylation (1 if no test is performed)</p></td>
</tr>
</tbody>
</table>
</div>
<div class="section" id="compression-and-indexing-allc">
<h4>Compression and Indexing ALLC<a class="headerlink" href="#compression-and-indexing-allc" title="Permalink to this headline">#</a></h4>
<p>All the ALLC files generated by ALLCools are compressed by <a class="reference external" href="http://www.htslib.org/doc/bgzip.html"><code class="docutils literal notranslate"><span class="pre">bgzip</span></code></a>
and indexed by <a class="reference external" href="http://www.htslib.org/doc/tabix.html"><code class="docutils literal notranslate"><span class="pre">tabix</span></code></a>. The <code class="docutils literal notranslate"><span class="pre">tabix</span></code> enable us to query ALLC files with
specific genomic regions. To check whether your ALLC files meet the standard, you can use
<a class="reference internal" href="../command_line/allcools_standard.html"><span class="doc std std-doc"><code class="docutils literal notranslate"><span class="pre">allcools</span> <span class="pre">standard</span></code></span></a> to validate the compression and indexing.
You can also use <a class="reference internal" href="../command_line/allcools_tbi.html"><span class="doc std std-doc"><code class="docutils literal notranslate"><span class="pre">allcools</span> <span class="pre">tbi</span></code></span></a> to index ALLC files using <code class="docutils literal notranslate"><span class="pre">tabix</span></code>.</p>
<div class="admonition tip">
<p class="admonition-title">Tip</p>
<p><code class="docutils literal notranslate"><span class="pre">bgzip</span></code> and <code class="docutils literal notranslate"><span class="pre">tabix</span></code> <span id="id2">[<a class="reference internal" href="../project_info/citation.html#id13" title="Heng Li. Tabix: fast retrieval of sequence features from generic TAB-delimited files. Bioinformatics, 27(5):718719, March 2011.">Li, 2011</a>]</span> are wonderful tools to compress tab separated tables with genome coordinates. Because it allows accessing any parts of the compressed file by genome location or region, which allows allcools to perform many region level parallelization in its file handling commands.</p>
</div>
</div>
</div>
</div>
<div class="section" id="cell-by-feature-methylome-matrix">
<h2>Cell-by-feature Methylome Matrix<a class="headerlink" href="#cell-by-feature-methylome-matrix" title="Permalink to this headline">#</a></h2>
<p>For cellular analysis like clustering, we need to first aggregate the base-level methylation information into
the region-level and gather thousands of cells together into a data matrix. To do so, we provide MCDS format to store all kinds of quantification outlined below.</p>
<div class="section" id="mcds-file-n-d-dense-matrix-for-raw-counts">
<h3>MCDS File: N-D dense matrix for raw counts<a class="headerlink" href="#mcds-file-n-d-dense-matrix-for-raw-counts" title="Permalink to this headline">#</a></h3>
<div class="figure align-default" id="mcds-fig">
<a class="reference internal image-reference" href="../_images/mcds.png"><img alt="../_images/mcds.png" src="../_images/mcds.png" style="height: 250px;" /></a>
<p class="caption"><span class="caption-number">Fig. 4 </span><span class="caption-text">Schematic of one MCDS file.</span><a class="headerlink" href="#mcds-fig" title="Permalink to this image">#</a></p>
</div>
<p>Here are some design principles for MCDS format:</p>
<div class="section" id="raw-count-matrix-in-mcds">
<h4>Raw count matrix in MCDS<a class="headerlink" href="#raw-count-matrix-in-mcds" title="Permalink to this headline">#</a></h4>
<p>The <a class="reference internal" href="../api/ALLCools/mcds/mcds/index.html#ALLCools.mcds.mcds.MCDS" title="ALLCools.mcds.mcds.MCDS"><code class="xref py py-class docutils literal notranslate"><span class="pre">MCDS</span></code></a> is a versatile data matrix format for storing region-level methylation values from a set of cells. It can handle all sizes of regions (from hundreds of BPs to MBs), such as genomic bins or gene body regions. Because the high coverage of single-cell methylome data (for snmC-seq2, 1.6M reads per cell), the methylation matrix is usually not sparse (~80-90% non-zero values when counting gene bodies). Therefore, we store raw count matrix as a dense N-dimensional matrix, with <strong>four common dimensions</strong> representing the cells, regions, methylation context (e.g., CH or CG), and count type (e.g., methylated base count (mc in ALLC) or total base count (cov in ALLC)).</p>
</div>
<div class="section" id="mcds-store-multiple-region-sets-together">
<h4>MCDS store multiple region sets together<a class="headerlink" href="#mcds-store-multiple-region-sets-together" title="Permalink to this headline">#</a></h4>
<p>It is also necessary to count the same cell using different region sets. For example, in the brain cell clustering analysis, we will <a class="reference internal" href="../cell_level/basic/mch_mcg_100k_basic.html"><span class="doc std std-doc">use 100Kb genomic bins in clustering</span></a> and <a class="reference internal" href="../cell_level/dmg/intro_dmg.html"><span class="doc std std-doc">use the gene body regions to identify DMGs</span></a>. One MCDS can save multiple region sets together, making it easy to switch between different region sets.</p>
</div>
<div class="section" id="mcds-enables-on-disk-data-handling">
<h4>MCDS enables on-disk data handling<a class="headerlink" href="#mcds-enables-on-disk-data-handling" title="Permalink to this headline">#</a></h4>
<p>Single-cell methylome dataset is essentially hundreds of thousands of whole-genome sequencing profiles. Due to this high coverage (million reads per cell) in nature, cell-level analysis, especially the preprocessing steps before dimensionality reduction, can be quite memory intensive. One of the most important reasons that we chose <code class="docutils literal notranslate"><span class="pre">xarray.Dataset</span></code> to model MCDS is because <a class="reference external" href="http://xarray.pydata.org/en/stable/dask.html">it allows on-disk data manipulation and lazy computation</a>. This feature enabled by python package <code class="docutils literal notranslate"><span class="pre">xarray</span></code> and <code class="docutils literal notranslate"><span class="pre">dask</span></code> allow us to handle hundreds of thousands of single-cell methylomes that dont fit into the memory of a single computational node. Still, memory consumption need to be carefully monitored during the single-cell methylome data analysis, especially before dimensionality reduction.</p>
</div>
<div class="section" id="when-quantifying-small-size-features">
<h4>When quantifying small-size features<a class="headerlink" href="#when-quantifying-small-size-features" title="Permalink to this headline">#</a></h4>
<p>In some analysis, one may need to count the single-cell methylomes with small genomic regions (e.g., 5Kb genomic
bins or hundreds of bp DMRs), since smaller regions might capture methylation diversity at cis-regulatory elements better during cell-level analysis.</p>
<p>The caveats of using small regions are that</p>
<ol class="simple">
<li><p>the matrix sparsity increases when the region is small;</p></li>
<li><p>the number of total regions need to be considered also goes up quickly.
Therefore, instead of saving the raw counts, we provide an alternative way of quantification, by using the probability score from binomial distribution, to turn the methylation level of each feature into a 0-1 value. Depending on the direction, this score is called hypo-methylation or hyper-methylation score.</p></li>
</ol>
<p>In further analysis steps, the hypo- or hyper-methylation score matrix can be treated as a boolean matrix, using a clustering strategy similar to snATAC-seq cell clustering. For more details, see our <a class="reference internal" href="../cell_level/basic/mcg_5kb_basic.html"><span class="doc std std-doc">mCG-5Kb bins clustering methods</span></a> tutorial.</p>
</div>
<div class="section" id="hypo-methylation-and-hyper-methylation-score">
<span id="hypo-and-hyper-score"></span><h4>Hypo-methylation and Hyper-methylation Score<a class="headerlink" href="#hypo-methylation-and-hyper-methylation-score" title="Permalink to this headline">#</a></h4>
<div class="admonition note">
<p class="admonition-title">Note</p>
<ul class="simple">
<li><p>The hypo-methylation score ranges from 0 to 1. The higher the score, the more hypo-methylated the region is in the cell.</p></li>
<li><p>The hyper-methylation score ranges from 0 to 1. The higher the score, the more hyper-methylated the region is in the cell.</p></li>
</ul>
</div>
<p>The hypo-methylation score is calculated by performing binomial test on each individual base count pair and then using the survival function (<code class="docutils literal notranslate"><span class="pre">1</span> <span class="pre">-</span> <span class="pre">p</span></code>) as the value. This score ranges from 0 to 1, the higher, the more hypo-methylated (potentially, more open and active). A cutoff value is applied so that only significant values are stored. During <a class="reference internal" href="../cell_level/basic/mcg_5kb_basic.html"><span class="doc std std-doc">analysis using this matrix</span></a>, we will apply more filters and then binarize it to apply the algorithms adapted from snATAC-seq analysis.</p>
<p>The hyper-methylation score is for the reverse trend as the hypo-methylation score (using <code class="docutils literal notranslate"><span class="pre">p</span></code>, instead of <code class="docutils literal notranslate"><span class="pre">1</span> <span class="pre">-</span> <span class="pre">p</span></code>). It is used for cases where hyper-methylation is of the interests. For example, the methylation level of GpC context when the assay contains NOMe treatment (hyper-methylation means open region).</p>
</div>
</div>
</div>
<div class="section" id="annotated-genome-region-matrix">
<h2>Annotated Genome Region Matrix<a class="headerlink" href="#annotated-genome-region-matrix" title="Permalink to this headline">#</a></h2>
<div class="section" id="regionds-file-n-d-dense-matrix-for-genome-region-with-various-annotations">
<h3>RegionDS file: N-D dense matrix for Genome Region with various annotations<a class="headerlink" href="#regionds-file-n-d-dense-matrix-for-genome-region-with-various-annotations" title="Permalink to this headline">#</a></h3>
<p>A <a class="reference internal" href="../api/ALLCools/mcds/region_ds/index.html#ALLCools.mcds.region_ds.RegionDS" title="ALLCools.mcds.region_ds.RegionDS"><code class="xref py py-class docutils literal notranslate"><span class="pre">RegionDS</span></code></a> can be created from <a class="reference internal" href="../cluster_level/RegionDS/01a.call_dmr.html"><span class="doc std std-doc">DMR analysis</span></a> or <a class="reference internal" href="../cluster_level/RegionDS/01b.other_option_to_init_region_ds.html"><span class="doc std std-doc">a set of genomic regions in a BED file</span></a>. An initial RegionDS only contains genome coordinates of the regions, it can then be annotated with other genomic features, epigenomic profiles, TF binding motifs, etc. Other genomic analysis such as Region-Region correlation and enhancer prediction can be performed after a rich annotated RegionDS. RegionDS is ready to handle millions of regions with the wonderful <a class="reference external" href="https://xarray.pydata.org/en/stable/index.html">xarray</a> and <a class="reference external" href="https://zarr.readthedocs.io/en/stable/">zarr</a> package as backends.</p>
</div>
<div class="section" id="relationships-between-regionds-and-mcds">
<h3>Relationships between RegionDS and MCDS<a class="headerlink" href="#relationships-between-regionds-and-mcds" title="Permalink to this headline">#</a></h3>
<p><a class="reference internal" href="../api/ALLCools/mcds/mcds/index.html#ALLCools.mcds.mcds.MCDS" title="ALLCools.mcds.mcds.MCDS"><code class="xref py py-class docutils literal notranslate"><span class="pre">MCDS</span></code></a> and <a class="reference internal" href="../api/ALLCools/mcds/region_ds/index.html#ALLCools.mcds.region_ds.RegionDS" title="ALLCools.mcds.region_ds.RegionDS"><code class="xref py py-class docutils literal notranslate"><span class="pre">RegionDS</span></code></a> are very similar data structures designed for different purpose (cellular and genomic analysis). Two key ideas for understanding their relationship includes:</p>
<ol class="simple">
<li><p>The MCDS and RegionDS are both inherited from the <a class="reference external" href="https://xarray.pydata.org/en/stable/generated/xarray.Dataset.html#xarray.Dataset" title="(in xarray v2022.6.1.dev0)"><code class="xref py py-class docutils literal notranslate"><span class="pre">xarray.Dataset</span></code></a> class, which internally uses the <a class="reference external" href="https://zarr.readthedocs.io/en/stable/">zarr</a> format for <a class="reference external" href="https://xarray.pydata.org/en/stable/user-guide/io.html#zarr">data storage and lazy loading</a> as well as the <a class="reference external" href="https://dask.org/">dask</a> package for parallel computation.</p></li>
<li><p>The methods in MCDS treat cells as observations and genomic regions as features; the goal is to do cellular analysis. The methods in RegionDS treat genomic regions as observations and cell or cell clusters as features; the goal is to do genomic analysis.</p></li>
</ol>
<div class="figure align-default" id="design-principle">
<a class="reference internal image-reference" href="../_images/design_principle.png"><img alt="../_images/design_principle.png" src="../_images/design_principle.png" style="height: 450px;" /></a>
<p class="caption"><span class="caption-number">Fig. 5 </span><span class="caption-text">A. The bases are the same and the methods are different. B. MCDS is cell-by-region matrix for cellular analysis; RegionDS is region-by-cell matrix for genomic analysis.</span><a class="headerlink" href="#design-principle" title="Permalink to this image">#</a></p>
</div>
</div>
</div>
</div>

    <script type="text/x-thebe-config">
    {
        requestKernel: true,
        binderOptions: {
            repo: "binder-examples/jupyter-stacks-datascience",
            ref: "master",
        },
        codeMirrorConfig: {
            theme: "abcdef",
            mode: "python"
        },
        kernelOptions: {
            kernelName: "python3",
            path: "./start"
        },
        predefinedOutput: true
    }
    </script>
    <script>kernelName = 'python3'</script>

              </div>
              
            </main>
            <footer class="footer-article noprint">
                
    <!-- Previous / next buttons -->
<div class='prev-next-area'>
    <a class='left-prev' id="prev-link" href="installation.html" title="previous page">
        <i class="fas fa-angle-left"></i>
        <div class="prev-next-info">
            <p class="prev-next-subtitle">previous</p>
            <p class="prev-next-title">Installation</p>
        </div>
    </a>
    <a class='right-next' id="next-link" href="analysis_steps.html" title="next page">
    <div class="prev-next-info">
        <p class="prev-next-subtitle">next</p>
        <p class="prev-next-title">Analysis Steps</p>
    </div>
    <i class="fas fa-angle-right"></i>
    </a>
</div>
            </footer>
        </div>
    </div>
    <div class="footer-content row">
        <footer class="col footer"><p>
  
    By Hanqing Liu<br/>
  
      &copy; Copyright 2019 - 2022.<br/>
</p>
        </footer>
    </div>
    
</div>


      </div>
    </div>
  
  <!-- Scripts loaded after <body> so the DOM is not blocked -->
  <script src="../_static/scripts/pydata-sphinx-theme.js?digest=1999514e3f237ded88cf"></script>


  </body>
</html>